// @medusajs/core-flows@2.6.1/dist/product/workflows/create-products.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProductsWorkflow = exports.createProductsWorkflowId = exports.validateProductInputStep = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const sales_channel_1 = require("../../sales-channel");
const create_products_1 = require("../steps/create-products");
const create_product_variants_1 = require("./create-product-variants");
const validateProductInputStepId = "validate-product-input";
/**
 * This step validates that all provided products have options.
 * If a product is missing options or a shipping profile, an error is thrown.
 *
 * @example
 * const data = validateProductInputStep({
 *   products: [
 *     {
 *       title: "Shirt",
 *       options: [
 *         {
 *           title: "Size",
 *           values: ["S", "M", "L"]
 *         }
 *       ],
 *       variants: [
 *         {
 *           title: "Small Shirt",
 *           sku: "SMALLSHIRT",
 *           options: {
 *             Size: "S"
 *           },
 *           prices: [
 *             {
 *               amount: 10,
 *               currency_code: "usd"
 *             }
 *           ],
 *           manage_inventory: true,
 *         },
 *       ]
 *     }
 *   ]
 * })
 */
exports.validateProductInputStep = (0, workflows_sdk_1.createStep)(validateProductInputStepId, async (data) => {
    const { products } = data;
    const missingOptionsProductTitles = products
        .filter((product) => !product.options?.length)
        .map((product) => product.title);
    if (missingOptionsProductTitles.length) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Product options are not provided for: [${missingOptionsProductTitles.join(", ")}].`);
    }
});
exports.createProductsWorkflowId = "create-products";
/**
 * This workflow creates one or more products. It's used by the [Create Product Admin API Route](https://docs.medusajs.com/api/admin#products_postproducts).
 * It can also be useful to you when creating [seed scripts](https://docs.medusajs.com/learn/fundamentals/custom-cli-scripts/seed-data), for example.
 *
 * This workflow has a hook that allows you to perform custom actions on the created products. You can see an example in [this guide](https://docs.medusajs.com/resources/commerce-modules/product/extend).
 *
 * You can also use this workflow within your customizations or your own custom workflows, allowing you to wrap custom logic around product creation.
 *
 * @example
 * const { result } = await createProductsWorkflow(container)
 * .run({
 *   input: {
 *     products: [
 *       {
 *         title: "Shirt",
 *         options: [
 *           {
 *             title: "Size",
 *             values: ["S", "M", "L"]
 *           }
 *         ],
 *         variants: [
 *           {
 *             title: "Small Shirt",
 *             sku: "SMALLSHIRT",
 *             options: {
 *               Size: "S"
 *             },
 *             prices: [
 *               {
 *                 amount: 10,
 *                 currency_code: "usd"
 *               }
 *             ],
 *             manage_inventory: true,
 *           },
 *         ]
 *       }
 *     ],
 *     shipping_profile_id: "sp_123",
 *     additional_data: {
 *       erp_id: "123"
 *     }
 *   }
 * })
 *
 * @summary
 *
 * Create one or more products with options and variants.
 *
 * @property hooks.productsCreated - This hook is executed after the products are created. You can consume this hook to perform custom actions on the created products.
 */
exports.createProductsWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.createProductsWorkflowId, (input) => {
    // Passing prices to the product module will fail, we want to keep them for after the product is created.
    const { products: productWithoutExternalRelations } = (0, workflows_sdk_1.transform)({ input }, (data) => {
        const productsData = data.input.products.map((p) => {
            return {
                ...p,
                sales_channels: undefined,
                shipping_profile_id: undefined,
                variants: undefined,
            };
        });
        return { products: productsData };
    });
    (0, exports.validateProductInputStep)({ products: input.products });
    const createdProducts = (0, create_products_1.createProductsStep)(productWithoutExternalRelations);
    const salesChannelLinks = (0, workflows_sdk_1.transform)({ input, createdProducts }, (data) => {
        return data.createdProducts
            .map((createdProduct, i) => {
            const inputProduct = data.input.products[i];
            return (inputProduct.sales_channels?.map((salesChannel) => ({
                sales_channel_id: salesChannel.id,
                product_id: createdProduct.id,
            })) ?? []);
        })
            .flat();
    });
    (0, sales_channel_1.associateProductsWithSalesChannelsStep)({ links: salesChannelLinks });
    const shippingProfileLinks = (0, workflows_sdk_1.transform)({ input, createdProducts }, (data) => {
        return data.createdProducts
            .map((createdProduct, i) => {
            return {
                [utils_1.Modules.PRODUCT]: {
                    product_id: createdProduct.id,
                },
                [utils_1.Modules.FULFILLMENT]: {
                    shipping_profile_id: data.input.products[i].shipping_profile_id,
                },
            };
        })
            .filter((link) => !!link[utils_1.Modules.FULFILLMENT].shipping_profile_id);
    });
    (0, common_1.createRemoteLinkStep)(shippingProfileLinks);
    const variantsInput = (0, workflows_sdk_1.transform)({ input, createdProducts }, (data) => {
        // TODO: Move this to a unified place for all product workflow types
        const productVariants = [];
        data.createdProducts.forEach((product, i) => {
            const inputProduct = data.input.products[i];
            for (const inputVariant of inputProduct.variants || []) {
                (0, utils_1.isPresent)(inputVariant) &&
                    productVariants.push({
                        product_id: product.id,
                        ...inputVariant,
                    });
            }
        });
        return {
            input: { product_variants: productVariants },
        };
    });
    const createdVariants = create_product_variants_1.createProductVariantsWorkflow.runAsStep(variantsInput);
    const response = (0, workflows_sdk_1.transform)({ createdVariants, input, createdProducts }, (data) => {
        const variantMap = {};
        for (const variant of data.createdVariants) {
            const array = variantMap[variant.product_id] || [];
            array.push(variant);
            variantMap[variant.product_id] = array;
        }
        for (const product of data.createdProducts) {
            product.variants = variantMap[product.id] || [];
        }
        return data.createdProducts;
    });
    const productIdEvents = (0, workflows_sdk_1.transform)({ response }, ({ response }) => {
        return response.map((v) => {
            return { id: v.id };
        });
    });
    (0, common_1.emitEventStep)({
        eventName: utils_1.ProductWorkflowEvents.CREATED,
        data: productIdEvents,
    });
    const productsCreated = (0, workflows_sdk_1.createHook)("productsCreated", {
        products: response,
        additional_data: input.additional_data,
    });
    return new workflows_sdk_1.WorkflowResponse(response, {
        hooks: [productsCreated],
    });
});
