// @medusajs/core-flows@2.6.1/dist/product/workflows/delete-product-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteProductTypesWorkflow = exports.deleteProductTypesWorkflowId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const emit_event_1 = require("../../common/steps/emit-event");
const steps_1 = require("../steps");
exports.deleteProductTypesWorkflowId = "delete-product-types";
/**
 * This workflow deletes one or more product types. It's used by the
 * [Delete Product Types Admin API Route](https://docs.medusajs.com/api/admin#product-types_deleteproducttypesid).
 *
 * This workflow has a hook that allows you to perform custom actions after the product types are deleted. For example,
 * you can delete custom records linked to the product types.
 *
 * You can also use this workflow within your own custom workflows, allowing you to wrap custom logic around product-type deletion.
 *
 * @example
 * const { result } = await deleteProductTypesWorkflow(container)
 * .run({
 *   input: {
 *     ids: ["ptyp_123"],
 *   }
 * })
 *
 * @summary
 *
 * Delete one or more product types.
 *
 * @property hooks.productTypesDeleted - This hook is executed after the types are deleted. You can consume this hook to perform custom actions on the deleted types.
 */
exports.deleteProductTypesWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.deleteProductTypesWorkflowId, (input) => {
    const deletedProductTypes = (0, steps_1.deleteProductTypesStep)(input.ids);
    const productTypesDeleted = (0, workflows_sdk_1.createHook)("productTypesDeleted", {
        ids: input.ids,
    });
    const typeIdEvents = (0, workflows_sdk_1.transform)({ input }, ({ input }) => {
        return input.ids?.map((id) => {
            return { id };
        });
    });
    (0, emit_event_1.emitEventStep)({
        eventName: utils_1.ProductTypeWorkflowEvents.DELETED,
        data: typeIdEvents,
    });
    return new workflows_sdk_1.WorkflowResponse(deletedProductTypes, {
        hooks: [productTypesDeleted],
    });
});
