// @medusajs/core-flows@2.6.1/dist/product/workflows/update-products.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateProductsWorkflow = exports.updateProductsWorkflowId = void 0;
const update_products_1 = require("../steps/update-products");
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const upsert_variant_prices_1 = require("./upsert-variant-prices");
function prepareUpdateProductInput({ input, }) {
    if ("products" in input) {
        if (!input.products.length) {
            return { products: [] };
        }
        return {
            products: input.products.map((p) => ({
                ...p,
                sales_channels: undefined,
                shipping_profile_id: undefined,
                variants: p.variants?.map((v) => ({
                    ...v,
                    prices: undefined,
                })),
            })),
        };
    }
    return {
        selector: input.selector,
        update: {
            ...input.update,
            sales_channels: undefined,
            shipping_profile_id: undefined,
            variants: input.update?.variants?.map((v) => ({
                ...v,
                prices: undefined,
            })),
        },
    };
}
// This helper finds the IDs of products that have associated sales channels.
function findProductsWithSalesChannels({ updatedProducts, input, }) {
    let productIds = updatedProducts.map((p) => p.id);
    if ("products" in input) {
        const discardedProductIds = input.products
            .filter((p) => !p.sales_channels)
            .map((p) => p.id);
        return (0, utils_1.arrayDifference)(productIds, discardedProductIds);
    }
    return !input.update?.sales_channels ? [] : productIds;
}
function findProductsWithShippingProfiles({ updatedProducts, input, }) {
    let productIds = updatedProducts.map((p) => p.id);
    if ("products" in input) {
        const discardedProductIds = input.products
            .filter((p) => !(0, utils_1.isDefined)(p.shipping_profile_id))
            .map((p) => p.id);
        return (0, utils_1.arrayDifference)(productIds, discardedProductIds);
    }
    return !(0, utils_1.isDefined)(input.update?.shipping_profile_id) ? [] : productIds;
}
function prepareSalesChannelLinks({ input, updatedProducts, }) {
    if ("products" in input) {
        if (!input.products.length) {
            return [];
        }
        return input.products
            .filter((p) => p.sales_channels)
            .flatMap((p) => p.sales_channels.map((sc) => ({
            [utils_1.Modules.PRODUCT]: {
                product_id: p.id,
            },
            [utils_1.Modules.SALES_CHANNEL]: {
                sales_channel_id: sc.id,
            },
        })));
    }
    if (input.selector && input.update?.sales_channels?.length) {
        return updatedProducts.flatMap((p) => input.update.sales_channels.map((channel) => ({
            [utils_1.Modules.PRODUCT]: {
                product_id: p.id,
            },
            [utils_1.Modules.SALES_CHANNEL]: {
                sales_channel_id: channel.id,
            },
        })));
    }
    return [];
}
function prepareShippingProfileLinks({ input, updatedProducts, }) {
    if ("products" in input) {
        if (!input.products.length) {
            return [];
        }
        return input.products
            .filter((p) => typeof p.shipping_profile_id === "string")
            .map((p) => ({
            [utils_1.Modules.PRODUCT]: {
                product_id: p.id,
            },
            [utils_1.Modules.FULFILLMENT]: {
                shipping_profile_id: p.shipping_profile_id,
            },
        }));
    }
    if (input.selector && typeof input.update?.shipping_profile_id === "string") {
        return updatedProducts.map((p) => ({
            [utils_1.Modules.PRODUCT]: {
                product_id: p.id,
            },
            [utils_1.Modules.FULFILLMENT]: {
                shipping_profile_id: input.update.shipping_profile_id,
            },
        }));
    }
    return [];
}
function prepareVariantPrices({ input, updatedProducts, }) {
    if ("products" in input) {
        if (!input.products.length) {
            return [];
        }
        // Note: We rely on the ordering of input and update here.
        return input.products.flatMap((product, i) => {
            if (!product.variants?.length) {
                return [];
            }
            const updatedProduct = updatedProducts[i];
            return product.variants.map((variant, j) => {
                const updatedVariant = updatedProduct.variants[j];
                return {
                    product_id: updatedProduct.id,
                    variant_id: updatedVariant.id,
                    prices: variant.prices,
                };
            });
        });
    }
    if (input.selector && input.update?.variants?.length) {
        return updatedProducts.flatMap((p) => {
            return input.update.variants.map((variant, i) => ({
                product_id: p.id,
                variant_id: p.variants[i].id,
                prices: variant.prices,
            }));
        });
    }
    return [];
}
function prepareToDeleteSalesChannelLinks({ currentSalesChannelLinks, }) {
    if (!currentSalesChannelLinks.length) {
        return [];
    }
    return currentSalesChannelLinks.map(({ product_id, sales_channel_id }) => ({
        [utils_1.Modules.PRODUCT]: {
            product_id,
        },
        [utils_1.Modules.SALES_CHANNEL]: {
            sales_channel_id,
        },
    }));
}
function prepareToDeleteShippingProfileLinks({ currentShippingProfileLinks, }) {
    if (!currentShippingProfileLinks.length) {
        return [];
    }
    return currentShippingProfileLinks.map(({ product_id, shipping_profile_id }) => ({
        [utils_1.Modules.PRODUCT]: {
            product_id,
        },
        [utils_1.Modules.FULFILLMENT]: {
            shipping_profile_id,
        },
    }));
}
exports.updateProductsWorkflowId = "update-products";
/**
 * This workflow updates one or more products. It's used by the [Update Product Admin API Route](https://docs.medusajs.com/api/admin#products_postproductsid).
 *
 * This workflow has a hook that allows you to perform custom actions on the updated products. For example, you can pass under `additional_data` custom data that
 * allows you to update custom data models linked to the products.
 *
 * You can also use this workflow within your customizations or your own custom workflows, allowing you to wrap custom logic around product update.
 *
 * @example
 * To update products by their IDs:
 *
 * ```ts
 * const { result } = await updateProductsWorkflow(container)
 * .run({
 *   input: {
 *     products: [
 *       {
 *         id: "prod_123",
 *         title: "Shirts"
 *       },
 *       {
 *         id: "prod_321",
 *         variants: [
 *           {
 *             id: "variant_123",
 *             options: {
 *               Size: "S"
 *             }
 *           }
 *         ]
 *       }
 *     ],
 *     additional_data: {
 *       erp_id: "erp_123"
 *     }
 *   }
 * })
 * ```
 *
 * You can also update products by a selector:
 *
 * ```ts
 * const { result } = await updateProductsWorkflow(container)
 * .run({
 *   input: {
 *     selector: {
 *       type_id: ["ptyp_123"]
 *     },
 *     update: {
 *       description: "This is a shirt product"
 *     },
 *     additional_data: {
 *       erp_id: "erp_123"
 *     }
 *   }
 * })
 * ```
 *
 * @summary
 *
 * Update one or more products with options and variants.
 *
 * @property hooks.productsUpdated - This hook is executed after the products are updated. You can consume this hook to perform custom actions on the updated products.
 */
exports.updateProductsWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.updateProductsWorkflowId, (input) => {
    // We only get the variant ids of products that are updating the variants and prices.
    const variantIdsSelector = (0, workflows_sdk_1.transform)({ input }, (data) => {
        if ("products" in data.input) {
            return {
                filters: {
                    id: data.input.products
                        .filter((p) => !!p.variants)
                        .map((p) => p.id),
                },
            };
        }
        return {
            filters: data.input.update?.variants ? data.input.selector : { id: [] },
        };
    });
    const previousProductsWithVariants = (0, common_1.useRemoteQueryStep)({
        entry_point: "product",
        fields: ["variants.id"],
        variables: variantIdsSelector,
    }).config({ name: "get-previous-products-variants-step" });
    const previousVariantIds = (0, workflows_sdk_1.transform)({ previousProductsWithVariants }, (data) => {
        return data.previousProductsWithVariants.flatMap((p) => p.variants?.map((v) => v.id));
    });
    const toUpdateInput = (0, workflows_sdk_1.transform)({ input }, prepareUpdateProductInput);
    const updatedProducts = (0, update_products_1.updateProductsStep)(toUpdateInput);
    const salesChannelLinks = (0, workflows_sdk_1.transform)({ input, updatedProducts }, prepareSalesChannelLinks);
    const shippingProfileLinks = (0, workflows_sdk_1.transform)({ input, updatedProducts }, prepareShippingProfileLinks);
    const variantPrices = (0, workflows_sdk_1.transform)({ input, updatedProducts }, prepareVariantPrices);
    const productsWithSalesChannels = (0, workflows_sdk_1.transform)({ updatedProducts, input }, findProductsWithSalesChannels);
    const productsWithShippingProfiles = (0, workflows_sdk_1.transform)({ updatedProducts, input }, findProductsWithShippingProfiles);
    const currentSalesChannelLinks = (0, common_1.useRemoteQueryStep)({
        entry_point: "product_sales_channel",
        fields: ["product_id", "sales_channel_id"],
        variables: { filters: { product_id: productsWithSalesChannels } },
    }).config({ name: "get-current-sales-channel-links-step" });
    const currentShippingProfileLinks = (0, common_1.useRemoteQueryStep)({
        entry_point: "product_shipping_profile",
        fields: ["product_id", "shipping_profile_id"],
        variables: { filters: { product_id: productsWithShippingProfiles } },
    }).config({ name: "get-current-shipping-profile-links-step" });
    const toDeleteSalesChannelLinks = (0, workflows_sdk_1.transform)({ currentSalesChannelLinks }, prepareToDeleteSalesChannelLinks);
    const toDeleteShippingProfileLinks = (0, workflows_sdk_1.transform)({ currentShippingProfileLinks }, prepareToDeleteShippingProfileLinks);
    upsert_variant_prices_1.upsertVariantPricesWorkflow.runAsStep({
        input: { variantPrices, previousVariantIds },
    });
    (0, common_1.dismissRemoteLinkStep)(toDeleteSalesChannelLinks).config({
        name: "delete-sales-channel-links-step",
    });
    (0, common_1.dismissRemoteLinkStep)(toDeleteShippingProfileLinks).config({
        name: "delete-shipping-profile-links-step",
    });
    const productIdEvents = (0, workflows_sdk_1.transform)({ updatedProducts }, ({ updatedProducts }) => {
        return updatedProducts?.map((p) => {
            return { id: p.id };
        });
    });
    (0, workflows_sdk_1.parallelize)((0, common_1.createRemoteLinkStep)(salesChannelLinks).config({
        name: "create-sales-channel-links-step",
    }), (0, common_1.createRemoteLinkStep)(shippingProfileLinks).config({
        name: "create-shipping-profile-links-step",
    }), (0, common_1.emitEventStep)({
        eventName: utils_1.ProductWorkflowEvents.UPDATED,
        data: productIdEvents,
    }));
    const productsUpdated = (0, workflows_sdk_1.createHook)("productsUpdated", {
        products: updatedProducts,
        additional_data: input.additional_data,
    });
    return new workflows_sdk_1.WorkflowResponse(updatedProducts, {
        hooks: [productsUpdated],
    });
});
