// @ariakit/core@0.4.14/cjs/__chunks/RL7DKHOF.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _BVPEDDW6cjs = require('./BVPEDDW6.cjs');


var _ASR6PAFNcjs = require('./ASR6PAFN.cjs');


var _MLCPLY2Pcjs = require('./MLCPLY2P.cjs');



var _7EQBAZ46cjs = require('./7EQBAZ46.cjs');

// src/hovercard/hovercard-store.ts
function createHovercardStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const popover = _BVPEDDW6cjs.createPopoverStore.call(void 0, _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {}, props), {
    placement: _MLCPLY2Pcjs.defaultValue.call(void 0, 
      props.placement,
      syncState == null ? void 0 : syncState.placement,
      "bottom"
    )
  }));
  const timeout = _MLCPLY2Pcjs.defaultValue.call(void 0, props.timeout, syncState == null ? void 0 : syncState.timeout, 500);
  const initialState = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {}, popover.getState()), {
    timeout,
    showTimeout: _MLCPLY2Pcjs.defaultValue.call(void 0, props.showTimeout, syncState == null ? void 0 : syncState.showTimeout),
    hideTimeout: _MLCPLY2Pcjs.defaultValue.call(void 0, props.hideTimeout, syncState == null ? void 0 : syncState.hideTimeout),
    autoFocusOnShow: _MLCPLY2Pcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.autoFocusOnShow, false)
  });
  const hovercard = _ASR6PAFNcjs.createStore.call(void 0, initialState, popover, props.store);
  return _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {}, popover), hovercard), {
    setAutoFocusOnShow: (value) => hovercard.setState("autoFocusOnShow", value)
  });
}



exports.createHovercardStore = createHovercardStore;
