// @ariakit/react-core@0.4.15/cjs/tag/tag-remove.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _DIH46ZNHcjs = require('../__chunks/DIH46ZNH.cjs');




var _KJ3PVFDWcjs = require('../__chunks/KJ3PVFDW.cjs');
require('../__chunks/CWQU4R75.cjs');
require('../__chunks/FVVTEJ4K.cjs');


var _3YJARQUNcjs = require('../__chunks/3YJARQUN.cjs');




var _6TI3NXX6cjs = require('../__chunks/6TI3NXX6.cjs');




var _L5A2ID7Scjs = require('../__chunks/L5A2ID7S.cjs');
require('../__chunks/FDRJDQ5Y.cjs');




var _7EQBAZ46cjs = require('../__chunks/7EQBAZ46.cjs');

// src/tag/tag-remove.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useTagRemove = _6TI3NXX6cjs.createHook.call(void 0, 
  function useTagRemove2(_a) {
    var _b = _a, {
      store,
      value: valueProp,
      removeOnClick = true
    } = _b, props = _7EQBAZ46cjs.__objRest.call(void 0, _b, [
      "store",
      "value",
      "removeOnClick"
    ]);
    const context = _KJ3PVFDWcjs.useTagContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "TagRemove must receive a `store` prop or be wrapped in a TagProvider component."
    );
    const id = _L5A2ID7Scjs.useId.call(void 0, props.id);
    const setRemoveId = _react.useContext.call(void 0, _KJ3PVFDWcjs.TagRemoveIdContext);
    const valueFromContext = _react.useContext.call(void 0, _KJ3PVFDWcjs.TagValueContext);
    const withinTag = valueFromContext !== null;
    const value = valueProp != null ? valueProp : valueFromContext;
    _react.useEffect.call(void 0, () => {
      setRemoveId == null ? void 0 : setRemoveId(id);
      return () => setRemoveId == null ? void 0 : setRemoveId();
    }, [id, setRemoveId]);
    const onClickProp = props.onClick;
    const removeOnClickProp = _L5A2ID7Scjs.useBooleanEvent.call(void 0, removeOnClick);
    const onClick = _L5A2ID7Scjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      if (!store) return;
      if (!value) return;
      if (!removeOnClickProp(event)) return;
      const { inputElement } = store.getState();
      store.removeValue(value);
      inputElement == null ? void 0 : inputElement.focus();
    });
    const children = /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, 
      "svg",
      {
        display: "block",
        viewBox: "0 0 24 24",
        fill: "none",
        stroke: "currentColor",
        strokeWidth: 1.5,
        strokeLinecap: "round",
        strokeLinejoin: "round",
        height: "1em",
        width: "1em",
        "aria-hidden": true,
        children: [
          /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "path", { d: "M7.28782 5.46243C7.80963 4.92308 8.07054 4.65341 8.37751 4.46038C8.64963 4.28926 8.94749 4.16299 9.25969 4.0864C9.61186 4 9.98709 4 10.7375 4H16.8C17.9201 4 18.4802 4 18.908 4.21799C19.2843 4.40973 19.5903 4.71569 19.782 5.09202C20 5.51984 20 6.0799 20 7.2V16.8C20 17.9201 20 18.4802 19.782 18.908C19.5903 19.2843 19.2843 19.5903 18.908 19.782C18.4802 20 17.9201 20 16.8 20H10.7512C9.99584 20 9.61818 20 9.264 19.9126C8.95004 19.8351 8.65067 19.7073 8.37748 19.5342C8.06931 19.339 7.80802 19.0663 7.28544 18.521L2.59879 13.63C2.03714 13.0439 1.75631 12.7508 1.65115 12.4158C1.55859 12.121 1.55935 11.8048 1.65332 11.5104C1.7601 11.1759 2.04233 10.8842 2.60679 10.3008L7.28782 5.46243Z" }),
          /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "path", { d: "M10.052 9L16 15M15.948 9L10 15" })
        ]
      }
    );
    const touchDevice = _DIH46ZNHcjs.useTouchDevice.call(void 0, ) && withinTag;
    props = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {
      id,
      children,
      role: touchDevice ? "button" : void 0,
      "aria-hidden": !touchDevice,
      "aria-label": touchDevice ? `Remove ${value}` : withinTag ? "Press Delete or Backspace to remove" : void 0
    }, props), {
      onClick,
      render: withinTag ? /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _3YJARQUNcjs.Role.span, { render: props.render }) : props.render
    });
    return props;
  }
);
var TagRemove = _6TI3NXX6cjs.forwardRef.call(void 0, function TagRemove2(props) {
  const htmlProps = useTagRemove(props);
  return _6TI3NXX6cjs.createElement.call(void 0, TagName, htmlProps);
});



exports.TagRemove = TagRemove; exports.useTagRemove = useTagRemove;
