// @medusajs/locking@2.6.1/dist/providers/in-memory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryLockingProvider = void 0;
const utils_1 = require("@medusajs/framework/utils");
class InMemoryLockingProvider {
    constructor() {
        this.locks = new Map();
    }
    getPromise() {
        let resolve;
        const pro = new Promise((ok) => {
            resolve = ok;
        });
        return {
            promise: pro,
            resolve,
        };
    }
    async execute(keys, job, args) {
        const timeout = Math.max(args?.timeout ?? 5, 1);
        const timeoutSeconds = Number.isNaN(timeout) ? 1 : timeout;
        const cancellationToken = { cancelled: false };
        const promises = [];
        if (timeoutSeconds > 0) {
            promises.push(this.getTimeout(timeoutSeconds, cancellationToken));
        }
        promises.push(this.acquire_(keys, {
            expire: timeoutSeconds,
            awaitQueue: true,
        }, cancellationToken));
        await Promise.race(promises);
        try {
            return await job();
        }
        finally {
            await this.release(keys);
        }
    }
    async acquire(keys, args) {
        return this.acquire_(keys, args);
    }
    async acquire_(keys, args, cancellationToken) {
        keys = Array.isArray(keys) ? keys : [keys];
        const { ownerId, expire } = args ?? {};
        for (const key of keys) {
            const lock = this.locks.get(key);
            const now = Date.now();
            if (!lock) {
                this.locks.set(key, {
                    ownerId: ownerId ?? null,
                    expiration: expire ? now + expire * 1000 : null,
                    currentPromise: this.getPromise(),
                });
                continue;
            }
            if (lock.expiration && lock.expiration <= now) {
                lock.currentPromise?.resolve?.();
                this.locks.set(key, {
                    ownerId: ownerId ?? null,
                    expiration: expire ? now + expire * 1000 : null,
                    currentPromise: this.getPromise(),
                });
                continue;
            }
            if (lock.ownerId !== null && lock.ownerId === ownerId) {
                if (expire) {
                    lock.expiration = now + expire * 1000;
                    this.locks.set(key, lock);
                }
                continue;
            }
            if (lock.currentPromise && args?.awaitQueue) {
                await lock.currentPromise.promise;
                if (cancellationToken?.cancelled) {
                    return;
                }
                return this.acquire(keys, args);
            }
            throw new Error(`Failed to acquire lock for key "${key}"`);
        }
    }
    async release(keys, args) {
        const { ownerId } = args ?? {};
        keys = Array.isArray(keys) ? keys : [keys];
        let success = true;
        for (const key of keys) {
            const lock = this.locks.get(key);
            if (!lock) {
                success = false;
                continue;
            }
            if ((0, utils_1.isDefined)(ownerId) && lock.ownerId !== ownerId) {
                success = false;
                continue;
            }
            lock.currentPromise?.resolve?.();
            this.locks.delete(key);
        }
        return success;
    }
    async releaseAll(args) {
        const { ownerId } = args ?? {};
        if (!(0, utils_1.isDefined)(ownerId)) {
            for (const [key, lock] of this.locks.entries()) {
                lock.currentPromise?.resolve?.();
                this.locks.delete(key);
            }
        }
        else {
            for (const [key, lock] of this.locks.entries()) {
                if (lock.ownerId === ownerId) {
                    lock.currentPromise?.resolve?.();
                    this.locks.delete(key);
                }
            }
        }
    }
    async getTimeout(seconds, cancellationToken) {
        return new Promise((_, reject) => {
            setTimeout(() => {
                cancellationToken.cancelled = true;
                reject(new Error("Timed-out acquiring lock."));
            }, seconds * 1000);
        });
    }
}
exports.InMemoryLockingProvider = InMemoryLockingProvider;
InMemoryLockingProvider.identifier = "in-memory";
