// @medusajs/medusa@2.6.1/dist/api/admin/customers/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminCustomerAddressesParams = exports.AdminCustomerAddressParams = exports.AdminUpdateCustomerAddress = exports.AdminCreateCustomerAddress = exports.CreateCustomerAddress = exports.AdminUpdateCustomer = exports.UpdateCustomer = exports.AdminCreateCustomer = exports.CreateCustomer = exports.AdminCustomersParams = exports.AdminCustomersParamsFields = exports.AdminCustomerGroupInCustomerParams = exports.AdminCustomerParams = void 0;
const zod_1 = require("zod");
const common_validators_1 = require("../../utils/common-validators");
const validators_1 = require("../../utils/validators");
exports.AdminCustomerParams = (0, validators_1.createSelectParams)();
exports.AdminCustomerGroupInCustomerParams = zod_1.z.object({
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    name: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    created_at: (0, validators_1.createOperatorMap)().optional(),
    updated_at: (0, validators_1.createOperatorMap)().optional(),
    deleted_at: (0, validators_1.createOperatorMap)().optional(),
});
exports.AdminCustomersParamsFields = zod_1.z.object({
    q: zod_1.z.string().optional(),
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    email: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    groups: zod_1.z
        .union([
        exports.AdminCustomerGroupInCustomerParams,
        zod_1.z.string(),
        zod_1.z.array(zod_1.z.string()),
    ])
        .optional(),
    company_name: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    first_name: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    last_name: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    has_account: (0, common_validators_1.booleanString)().optional(),
    created_by: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    created_at: (0, validators_1.createOperatorMap)().optional(),
    updated_at: (0, validators_1.createOperatorMap)().optional(),
    deleted_at: (0, validators_1.createOperatorMap)().optional(),
});
exports.AdminCustomersParams = (0, validators_1.createFindParams)({
    limit: 50,
    offset: 0,
})
    .merge(exports.AdminCustomersParamsFields)
    .merge((0, common_validators_1.applyAndAndOrOperators)(exports.AdminCustomersParamsFields));
exports.CreateCustomer = zod_1.z.object({
    email: zod_1.z.string().email().nullish(),
    company_name: zod_1.z.string().nullish(),
    first_name: zod_1.z.string().nullish(),
    last_name: zod_1.z.string().nullish(),
    phone: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
});
exports.AdminCreateCustomer = (0, validators_1.WithAdditionalData)(exports.CreateCustomer);
exports.UpdateCustomer = zod_1.z.object({
    email: zod_1.z.string().email().nullish(),
    company_name: zod_1.z.string().nullish(),
    first_name: zod_1.z.string().nullish(),
    last_name: zod_1.z.string().nullish(),
    phone: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
});
exports.AdminUpdateCustomer = (0, validators_1.WithAdditionalData)(exports.UpdateCustomer);
exports.CreateCustomerAddress = zod_1.z.object({
    address_name: zod_1.z.string().nullish(),
    is_default_shipping: zod_1.z.boolean().optional(),
    is_default_billing: zod_1.z.boolean().optional(),
    company: zod_1.z.string().nullish(),
    first_name: zod_1.z.string().nullish(),
    last_name: zod_1.z.string().nullish(),
    address_1: zod_1.z.string().nullish(),
    address_2: zod_1.z.string().nullish(),
    city: zod_1.z.string().nullish(),
    country_code: zod_1.z.string().nullish(),
    province: zod_1.z.string().nullish(),
    postal_code: zod_1.z.string().nullish(),
    phone: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
});
exports.AdminCreateCustomerAddress = (0, validators_1.WithAdditionalData)(exports.CreateCustomerAddress);
exports.AdminUpdateCustomerAddress = exports.AdminCreateCustomerAddress;
exports.AdminCustomerAddressParams = (0, validators_1.createSelectParams)();
exports.AdminCustomerAddressesParams = (0, validators_1.createFindParams)({
    offset: 0,
    limit: 50,
}).merge(zod_1.z.object({
    q: zod_1.z.string().optional(),
    company: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    city: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    country_code: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    province: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    postal_code: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
}));
