// @medusajs/medusa@2.6.1/dist/api/admin/returns/[id]/request/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DELETE = exports.POST = void 0;
const core_flows_1 = require("@medusajs/core-flows");
const utils_1 = require("@medusajs/framework/utils");
const POST = async (req, res) => {
    const { id } = req.params;
    const remoteQuery = req.scope.resolve(utils_1.ContainerRegistrationKeys.REMOTE_QUERY);
    const { result } = await (0, core_flows_1.confirmReturnRequestWorkflow)(req.scope).run({
        input: {
            return_id: id,
            confirmed_by: req.auth_context.actor_id,
        },
    });
    const queryObject = (0, utils_1.remoteQueryObjectFromString)({
        entryPoint: "return",
        variables: {
            id,
            filters: {
                ...req.filterableFields,
            },
        },
        fields: req.queryConfig.fields,
    });
    const [orderReturn] = await remoteQuery(queryObject);
    res.json({
        order_preview: result,
        return: orderReturn,
    });
};
exports.POST = POST;
const DELETE = async (req, res) => {
    const { id } = req.params;
    await (0, core_flows_1.cancelReturnRequestWorkflow)(req.scope).run({
        input: {
            return_id: id,
        },
    });
    res.status(200).json({
        id,
        object: "return",
        deleted: true,
    });
};
exports.DELETE = DELETE;
