// @medusajs/cart@2.6.1/dist/models/line-item.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@medusajs/framework/utils");
const cart_1 = __importDefault(require("./cart"));
const line_item_adjustment_1 = __importDefault(require("./line-item-adjustment"));
const line_item_tax_line_1 = __importDefault(require("./line-item-tax-line"));
const LineItem = utils_1.model
    .define({ name: "LineItem", tableName: "cart_line_item" }, {
    id: utils_1.model.id({ prefix: "cali" }).primaryKey(),
    title: utils_1.model.text(),
    subtitle: utils_1.model.text().nullable(),
    thumbnail: utils_1.model.text().nullable(),
    quantity: utils_1.model.number(),
    variant_id: utils_1.model.text().nullable(),
    product_id: utils_1.model.text().nullable(),
    product_title: utils_1.model.text().nullable(),
    product_description: utils_1.model.text().nullable(),
    product_subtitle: utils_1.model.text().nullable(),
    product_type: utils_1.model.text().nullable(),
    product_type_id: utils_1.model.text().nullable(),
    product_collection: utils_1.model.text().nullable(),
    product_handle: utils_1.model.text().nullable(),
    variant_sku: utils_1.model.text().nullable(),
    variant_barcode: utils_1.model.text().nullable(),
    variant_title: utils_1.model.text().nullable(),
    variant_option_values: utils_1.model.json().nullable(),
    requires_shipping: utils_1.model.boolean().default(true),
    is_discountable: utils_1.model.boolean().default(true),
    is_tax_inclusive: utils_1.model.boolean().default(false),
    is_custom_price: utils_1.model.boolean().default(false),
    compare_at_unit_price: utils_1.model.bigNumber().nullable(),
    unit_price: utils_1.model.bigNumber(),
    metadata: utils_1.model.json().nullable(),
    adjustments: utils_1.model.hasMany(() => line_item_adjustment_1.default, {
        mappedBy: "item",
    }),
    tax_lines: utils_1.model.hasMany(() => line_item_tax_line_1.default, {
        mappedBy: "item",
    }),
    cart: utils_1.model.belongsTo(() => cart_1.default, {
        mappedBy: "items",
    }),
})
    .indexes([
    {
        name: "IDX_line_item_cart_id",
        on: ["cart_id"],
        where: "deleted_at IS NULL",
    },
    {
        name: "IDX_line_item_variant_id",
        on: ["variant_id"],
        where: "deleted_at IS NULL AND variant_id IS NOT NULL",
    },
    {
        name: "IDX_line_item_product_id",
        on: ["product_id"],
        where: "deleted_at IS NULL AND product_id IS NOT NULL",
    },
    {
        name: "IDX_line_item_product_type_id",
        on: ["product_type_id"],
        where: "deleted_at IS NULL AND product_type_id IS NOT NULL",
    },
])
    .cascades({
    delete: ["adjustments", "tax_lines"],
});
exports.default = LineItem;
