// @medusajs/ui@4.0.7/dist/cjs/blocks/data-table/use-data-table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useDataTable = void 0;
const tslib_1 = require("tslib");
const react_table_1 = require("@tanstack/react-table");
const React = tslib_1.__importStar(require("react"));
const types_1 = require("./types");
const useDataTable = ({ rowCount = 0, filters, commands, rowSelection, sorting, filtering, pagination, search, onRowClick, autoResetPageIndex = true, isLoading = false, ...options }) => {
    var _a, _b, _c, _d, _e, _f;
    const { state: sortingState, onSortingChange } = sorting !== null && sorting !== void 0 ? sorting : {};
    const { state: filteringState, onFilteringChange } = filtering !== null && filtering !== void 0 ? filtering : {};
    const { state: paginationState, onPaginationChange } = pagination !== null && pagination !== void 0 ? pagination : {};
    const { state: rowSelectionState, onRowSelectionChange, enableRowSelection } = rowSelection !== null && rowSelection !== void 0 ? rowSelection : {};
    const autoResetPageIndexHandler = React.useCallback(() => {
        return autoResetPageIndex
            ? () => paginationState &&
                (onPaginationChange === null || onPaginationChange === void 0 ? void 0 : onPaginationChange({ ...paginationState, pageIndex: 0 }))
            : undefined;
    }, [autoResetPageIndex, paginationState, onPaginationChange]);
    const sortingStateHandler = React.useCallback(() => {
        return onSortingChange
            ? (updaterOrValue) => {
                var _a;
                (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
                onSortingChangeTransformer(onSortingChange, sortingState)(updaterOrValue);
            }
            : undefined;
    }, [onSortingChange, sortingState, autoResetPageIndexHandler]);
    const rowSelectionStateHandler = React.useCallback(() => {
        return onRowSelectionChange
            ? (updaterOrValue) => {
                var _a;
                (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
                onRowSelectionChangeTransformer(onRowSelectionChange, rowSelectionState)(updaterOrValue);
            }
            : undefined;
    }, [onRowSelectionChange, rowSelectionState, autoResetPageIndexHandler]);
    const filteringStateHandler = React.useCallback(() => {
        return onFilteringChange
            ? (updaterOrValue) => {
                var _a;
                (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
                onFilteringChangeTransformer(onFilteringChange, filteringState)(updaterOrValue);
            }
            : undefined;
    }, [onFilteringChange, filteringState, autoResetPageIndexHandler]);
    const paginationStateHandler = React.useCallback(() => {
        return onPaginationChange
            ? onPaginationChangeTransformer(onPaginationChange, paginationState)
            : undefined;
    }, [onPaginationChange, paginationState]);
    const instance = (0, react_table_1.useReactTable)({
        ...options,
        getCoreRowModel: (0, react_table_1.getCoreRowModel)(),
        state: {
            rowSelection: rowSelectionState !== null && rowSelectionState !== void 0 ? rowSelectionState : {},
            sorting: sortingState ? [sortingState] : undefined,
            columnFilters: Object.entries(filteringState !== null && filteringState !== void 0 ? filteringState : {}).map(([id, filter]) => ({
                id,
                value: filter,
            })),
            pagination: paginationState,
        },
        enableRowSelection,
        rowCount,
        onColumnFiltersChange: filteringStateHandler(),
        onRowSelectionChange: rowSelectionStateHandler(),
        onSortingChange: sortingStateHandler(),
        onPaginationChange: paginationStateHandler(),
        manualSorting: true,
        manualPagination: true,
        manualFiltering: true,
    });
    const getSorting = React.useCallback(() => {
        var _a, _b;
        return (_b = (_a = instance.getState().sorting) === null || _a === void 0 ? void 0 : _a[0]) !== null && _b !== void 0 ? _b : null;
    }, [instance]);
    const setSorting = React.useCallback((sortingOrUpdater) => {
        var _a, _b, _c;
        const currentSort = (_b = (_a = instance.getState().sorting) === null || _a === void 0 ? void 0 : _a[0]) !== null && _b !== void 0 ? _b : null;
        const newSorting = typeof sortingOrUpdater === "function"
            ? sortingOrUpdater(currentSort)
            : sortingOrUpdater;
        (_c = autoResetPageIndexHandler()) === null || _c === void 0 ? void 0 : _c();
        instance.setSorting([newSorting]);
    }, [instance, autoResetPageIndexHandler]);
    const getFilters = React.useCallback(() => {
        return filters !== null && filters !== void 0 ? filters : [];
    }, [filters]);
    const getFilterOptions = React.useCallback((id) => {
        const filter = getFilters().find((filter) => filter.id === id);
        if (!filter) {
            return null;
        }
        return filter.options;
    }, [getFilters]);
    const getFilterMeta = React.useCallback((id) => {
        return getFilters().find((filter) => filter.id === id) || null;
    }, [getFilters]);
    const getFiltering = React.useCallback(() => {
        var _a;
        const state = (_a = instance.getState().columnFilters) !== null && _a !== void 0 ? _a : [];
        return Object.fromEntries(state.map((filter) => [filter.id, filter.value]));
    }, [instance]);
    const addFilter = React.useCallback((filter) => {
        var _a;
        if (filter.value) {
            (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
        }
        onFilteringChange === null || onFilteringChange === void 0 ? void 0 : onFilteringChange({ ...getFiltering(), [filter.id]: filter.value });
    }, [onFilteringChange, getFiltering, autoResetPageIndexHandler]);
    const removeFilter = React.useCallback((id) => {
        var _a;
        const currentFilters = getFiltering();
        delete currentFilters[id];
        (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
        onFilteringChange === null || onFilteringChange === void 0 ? void 0 : onFilteringChange(currentFilters);
    }, [onFilteringChange, getFiltering, autoResetPageIndexHandler]);
    const clearFilters = React.useCallback(() => {
        var _a;
        (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
        onFilteringChange === null || onFilteringChange === void 0 ? void 0 : onFilteringChange({});
    }, [onFilteringChange, autoResetPageIndexHandler]);
    const updateFilter = React.useCallback((filter) => {
        addFilter(filter);
    }, [addFilter]);
    const { state: searchState, onSearchChange, debounce = 300 } = search !== null && search !== void 0 ? search : {};
    const [localSearch, setLocalSearch] = React.useState(searchState !== null && searchState !== void 0 ? searchState : "");
    const timeoutRef = React.useRef();
    React.useEffect(() => {
        setLocalSearch(searchState !== null && searchState !== void 0 ? searchState : "");
    }, [searchState]);
    const getSearch = React.useCallback(() => {
        return localSearch;
    }, [localSearch]);
    const debouncedSearchChange = React.useMemo(() => {
        if (!onSearchChange) {
            return undefined;
        }
        return (value) => {
            var _a;
            if (timeoutRef.current) {
                clearTimeout(timeoutRef.current);
            }
            if (debounce <= 0) {
                (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
                onSearchChange(value);
                return;
            }
            timeoutRef.current = setTimeout(() => {
                var _a;
                (_a = autoResetPageIndexHandler()) === null || _a === void 0 ? void 0 : _a();
                onSearchChange(value);
            }, debounce);
        };
    }, [onSearchChange, debounce, autoResetPageIndexHandler]);
    React.useEffect(() => {
        return () => {
            if (timeoutRef.current) {
                clearTimeout(timeoutRef.current);
            }
        };
    }, []);
    const onSearchChangeHandler = React.useCallback((search) => {
        setLocalSearch(search);
        debouncedSearchChange === null || debouncedSearchChange === void 0 ? void 0 : debouncedSearchChange(search);
    }, [debouncedSearchChange]);
    const getCommands = React.useCallback(() => {
        return commands !== null && commands !== void 0 ? commands : [];
    }, [commands]);
    const getRowSelection = React.useCallback(() => {
        return instance.getState().rowSelection;
    }, [instance]);
    const rows = instance.getRowModel().rows;
    const emptyState = React.useMemo(() => {
        const hasRows = rows.length > 0;
        const hasSearch = Boolean(searchState);
        const hasFilters = Object.keys(filteringState !== null && filteringState !== void 0 ? filteringState : {}).length > 0;
        if (hasRows) {
            return types_1.DataTableEmptyState.POPULATED;
        }
        if (hasSearch || hasFilters) {
            return types_1.DataTableEmptyState.FILTERED_EMPTY;
        }
        return types_1.DataTableEmptyState.EMPTY;
    }, [rows, searchState, filteringState]);
    const showSkeleton = React.useMemo(() => {
        return isLoading === true && rows.length === 0;
    }, [isLoading, rows]);
    const enablePagination = !!pagination;
    const enableFiltering = !!filtering;
    const enableSorting = !!sorting;
    const enableSearch = !!search;
    return {
        // Table
        getHeaderGroups: instance.getHeaderGroups,
        getRowModel: instance.getRowModel,
        getAllColumns: instance.getAllColumns,
        // Pagination
        enablePagination,
        getCanNextPage: instance.getCanNextPage,
        getCanPreviousPage: instance.getCanPreviousPage,
        nextPage: instance.nextPage,
        previousPage: instance.previousPage,
        getPageCount: instance.getPageCount,
        pageIndex: (_c = (_b = (_a = instance.getState()) === null || _a === void 0 ? void 0 : _a.pagination) === null || _b === void 0 ? void 0 : _b.pageIndex) !== null && _c !== void 0 ? _c : 0,
        pageSize: (_f = (_e = (_d = instance.getState()) === null || _d === void 0 ? void 0 : _d.pagination) === null || _e === void 0 ? void 0 : _e.pageSize) !== null && _f !== void 0 ? _f : 10,
        rowCount,
        // Search
        enableSearch,
        getSearch,
        onSearchChange: onSearchChangeHandler,
        // Sorting
        enableSorting,
        getSorting,
        setSorting,
        // Filtering
        enableFiltering,
        getFilters,
        getFilterOptions,
        getFilterMeta,
        getFiltering,
        addFilter,
        removeFilter,
        clearFilters,
        updateFilter,
        // Commands
        getCommands,
        getRowSelection,
        // Handlers
        onRowClick,
        // Empty State
        emptyState,
        // Loading
        isLoading,
        showSkeleton,
    };
};
exports.useDataTable = useDataTable;
function onSortingChangeTransformer(onSortingChange, state) {
    return (updaterOrValue) => {
        const value = typeof updaterOrValue === "function"
            ? updaterOrValue(state ? [state] : [])
            : updaterOrValue;
        const columnSort = value[0];
        onSortingChange(columnSort);
    };
}
function onRowSelectionChangeTransformer(onRowSelectionChange, state) {
    return (updaterOrValue) => {
        const value = typeof updaterOrValue === "function"
            ? updaterOrValue(state !== null && state !== void 0 ? state : {})
            : updaterOrValue;
        onRowSelectionChange(value);
    };
}
function onFilteringChangeTransformer(onFilteringChange, state) {
    return (updaterOrValue) => {
        const value = typeof updaterOrValue === "function"
            ? updaterOrValue(Object.entries(state !== null && state !== void 0 ? state : {}).map(([id, filter]) => ({
                id,
                value: filter,
            })))
            : updaterOrValue;
        const transformedValue = Object.fromEntries(value.map((filter) => [filter.id, filter]));
        onFilteringChange(transformedValue);
    };
}
function onPaginationChangeTransformer(onPaginationChange, state) {
    return (updaterOrValue) => {
        const value = typeof updaterOrValue === "function"
            ? updaterOrValue(state !== null && state !== void 0 ? state : { pageIndex: 0, pageSize: 10 })
            : updaterOrValue;
        onPaginationChange(value);
    };
}
