// @medusajs/orchestration@2.6.1/dist/transaction/errors.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _BaseStepErrror_stepResponse;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonSerializableCheckPointError = exports.TransactionTimeoutError = exports.TransactionStepTimeoutError = exports.SkipStepResponse = exports.PermanentStepFailureError = void 0;
class BaseStepErrror extends Error {
    constructor(name, message, stepResponse) {
        super(message);
        _BaseStepErrror_stepResponse.set(this, void 0);
        this.name = name;
        __classPrivateFieldSet(this, _BaseStepErrror_stepResponse, stepResponse, "f");
    }
    getStepResponse() {
        return __classPrivateFieldGet(this, _BaseStepErrror_stepResponse, "f");
    }
}
_BaseStepErrror_stepResponse = new WeakMap();
class PermanentStepFailureError extends BaseStepErrror {
    static isPermanentStepFailureError(error) {
        return (error instanceof PermanentStepFailureError ||
            error?.name === "PermanentStepFailure");
    }
    constructor(message, stepResponse) {
        super("PermanentStepFailure", message, stepResponse);
    }
}
exports.PermanentStepFailureError = PermanentStepFailureError;
class SkipStepResponse extends BaseStepErrror {
    static isSkipStepResponse(error) {
        return (error instanceof SkipStepResponse || error?.name === "SkipStepResponse");
    }
    constructor(message, stepResponse) {
        super("SkipStepResponse", message, stepResponse);
    }
}
exports.SkipStepResponse = SkipStepResponse;
class TransactionStepTimeoutError extends BaseStepErrror {
    static isTransactionStepTimeoutError(error) {
        return (error instanceof TransactionStepTimeoutError ||
            error?.name === "TransactionStepTimeoutError");
    }
    constructor(message, stepResponse) {
        super("TransactionStepTimeoutError", message, stepResponse);
    }
}
exports.TransactionStepTimeoutError = TransactionStepTimeoutError;
class TransactionTimeoutError extends BaseStepErrror {
    static isTransactionTimeoutError(error) {
        return (error instanceof TransactionTimeoutError ||
            error?.name === "TransactionTimeoutError");
    }
    constructor(message, stepResponse) {
        super("TransactionTimeoutError", message, stepResponse);
    }
}
exports.TransactionTimeoutError = TransactionTimeoutError;
class NonSerializableCheckPointError extends Error {
    static isNonSerializableCheckPointError(error) {
        return (error instanceof NonSerializableCheckPointError ||
            error?.name === "NonSerializableCheckPointError");
    }
    constructor(message) {
        super(message);
        this.name = "NonSerializableCheckPointError";
    }
}
exports.NonSerializableCheckPointError = NonSerializableCheckPointError;
