// @medusajs/utils@2.6.1/dist/orchestration/symbol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SymbolMedusaWorkflowResponse = exports.SymbolWorkflowStepTransformer = exports.SymbolWorkflowStepBind = exports.SymbolWorkflowStepResponse = exports.SymbolWorkflowWorkflowData = exports.SymbolWorkflowHook = exports.SymbolWorkflowStep = exports.SymbolInputReference = exports.SymbolMedusaWorkflowComposerCondition = exports.SymbolMedusaWorkflowComposerContext = void 0;
exports.SymbolMedusaWorkflowComposerContext = Symbol.for("MedusaWorkflowComposerContext").toString();
exports.SymbolMedusaWorkflowComposerCondition = Symbol.for("MedusaWorkflowComposerCondition").toString();
exports.SymbolInputReference = Symbol.for("WorkflowInputReference").toString();
exports.SymbolWorkflowStep = Symbol.for("WorkflowStep").toString();
exports.SymbolWorkflowHook = Symbol.for("WorkflowHook").toString();
exports.SymbolWorkflowWorkflowData = Symbol.for("WorkflowWorkflowData").toString();
exports.SymbolWorkflowStepResponse = Symbol.for("WorkflowStepResponse").toString();
exports.SymbolWorkflowStepBind = Symbol.for("WorkflowStepBind").toString();
exports.SymbolWorkflowStepTransformer = Symbol.for("WorkflowStepTransformer").toString();
exports.SymbolMedusaWorkflowResponse = Symbol.for("MedusaWorkflowResponse").toString();
