// @medusajs/core-flows@2.6.1/dist/order/workflows/return/confirm-return-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.confirmReturnRequestWorkflow = exports.confirmReturnRequestWorkflowId = exports.confirmReturnRequestValidationStep = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../../common");
const create_return_fulfillment_1 = require("../../../fulfillment/workflows/create-return-fulfillment");
const steps_1 = require("../../steps");
const confirm_order_changes_1 = require("../../steps/confirm-order-changes");
const create_return_items_from_actions_1 = require("../../steps/return/create-return-items-from-actions");
const order_validation_1 = require("../../utils/order-validation");
const create_or_update_order_payment_collection_1 = require("../create-or-update-order-payment-collection");
/**
 * This step validates that a return request can be confirmed.
 * If the order or return is canceled or the order change is not active, the step will throw an error.
 *
 * :::note
 *
 * You can retrieve an order, order change, and return details using [Query](https://docs.medusajs.com/learn/fundamentals/module-links/query),
 * or [useQueryGraphStep](https://docs.medusajs.com/resources/references/medusa-workflows/steps/useQueryGraphStep).
 *
 * :::
 *
 * @example
 * const data = confirmReturnRequestValidationStep({
 *   order: {
 *     id: "order_123",
 *     // other order details...
 *   },
 *   orderReturn: {
 *     id: "return_123",
 *     // other order return details...
 *   },
 *   orderChange: {
 *     id: "orch_123",
 *     // other order change details...
 *   }
 * })
 */
exports.confirmReturnRequestValidationStep = (0, workflows_sdk_1.createStep)("validate-confirm-return-request", async function ({ order, orderChange, orderReturn, }) {
    (0, order_validation_1.throwIfIsCancelled)(order, "Order");
    (0, order_validation_1.throwIfIsCancelled)(orderReturn, "Return");
    (0, order_validation_1.throwIfOrderChangeIsNotActive)({ orderChange });
});
/**
 * This step confirms that a requested return has at least one item
 */
const confirmIfReturnItemsArePresent = (0, workflows_sdk_1.createStep)("confirm-if-return-items-are-present", async function ({ returnItems }) {
    if (returnItems.length > 0) {
        return;
    }
    throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Order return request should have at least 1 item`);
});
function prepareFulfillmentData({ order, items, returnShippingOption, }) {
    const orderItemsMap = new Map(order.items.map((i) => [i.id, i]));
    const fulfillmentItems = items.map((i) => {
        const orderItem = orderItemsMap.get(i.item_id);
        return {
            line_item_id: i.item_id,
            quantity: i.quantity,
            return_quantity: i.quantity,
            title: orderItem.variant_title ?? orderItem.title,
            sku: orderItem.variant_sku || "",
            barcode: orderItem.variant_barcode || "",
        };
    });
    const locationId = returnShippingOption.service_zone.fulfillment_set.location?.id;
    // delivery address is the stock location address
    const address = returnShippingOption.service_zone.fulfillment_set.location?.address ?? {};
    delete address.id;
    return {
        input: {
            location_id: locationId,
            provider_id: returnShippingOption.provider_id,
            shipping_option_id: returnShippingOption.id,
            items: fulfillmentItems,
            delivery_address: address,
            order: order,
        },
    };
}
function extractReturnShippingOptionId({ orderPreview, orderReturn }) {
    let returnShippingMethod;
    for (const shippingMethod of orderPreview.shipping_methods ?? []) {
        const modifiedShippingMethod_ = shippingMethod;
        if (!modifiedShippingMethod_.actions) {
            continue;
        }
        returnShippingMethod = modifiedShippingMethod_.actions.find((action) => {
            return (action.action === utils_1.ChangeActionType.SHIPPING_ADD &&
                action.return_id === orderReturn.id);
        });
    }
    if (!returnShippingMethod) {
        return null;
    }
    return returnShippingMethod.shipping_option_id;
}
function getUpdateReturnData({ orderReturn }) {
    return (0, workflows_sdk_1.transform)({ orderReturn }, ({ orderReturn }) => {
        return [
            {
                id: orderReturn.id,
                status: utils_1.ReturnStatus.REQUESTED,
                requested_at: new Date(),
            },
        ];
    });
}
exports.confirmReturnRequestWorkflowId = "confirm-return-request";
/**
 * This workflow confirms a return request. It's used by the
 * [Confirm Return Request Admin API Route](https://docs.medusajs.com/api/admin#returns_postreturnsidrequest).
 *
 * You can use this workflow within your customizations or your own custom workflows, allowing you to confirm a return request
 * in your custom flow.
 *
 * @example
 * const { result } = await confirmReturnRequestWorkflow(container)
 * .run({
 *   input: {
 *     return_id: "return_123",
 *   }
 * })
 *
 * @summary
 *
 * Confirm a return request.
 */
exports.confirmReturnRequestWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.confirmReturnRequestWorkflowId, function (input) {
    const orderReturn = (0, common_1.useRemoteQueryStep)({
        entry_point: "return",
        fields: ["id", "status", "order_id", "location_id", "canceled_at"],
        variables: { id: input.return_id },
        list: false,
        throw_if_key_not_found: true,
    });
    const order = (0, common_1.useRemoteQueryStep)({
        entry_point: "orders",
        fields: [
            "id",
            "version",
            "canceled_at",
            "items.id",
            "items.title",
            "items.variant_title",
            "items.variant_sku",
            "items.variant_barcode",
        ],
        variables: { id: orderReturn.order_id },
        list: false,
        throw_if_key_not_found: true,
    }).config({ name: "order-query" });
    const orderChange = (0, common_1.useRemoteQueryStep)({
        entry_point: "order_change",
        fields: [
            "id",
            "status",
            "actions.id",
            "actions.action",
            "actions.details",
            "actions.reference",
            "actions.reference_id",
            "actions.internal_note",
        ],
        variables: {
            filters: {
                order_id: orderReturn.order_id,
                return_id: orderReturn.id,
                status: [utils_1.OrderChangeStatus.PENDING, utils_1.OrderChangeStatus.REQUESTED],
            },
        },
        list: false,
    }).config({ name: "order-change-query" });
    const returnItemActions = (0, workflows_sdk_1.transform)({ orderChange }, (data) => {
        return data.orderChange.actions.filter((act) => act.action === utils_1.ChangeActionType.RETURN_ITEM);
    });
    (0, exports.confirmReturnRequestValidationStep)({ order, orderReturn, orderChange });
    const orderPreview = (0, steps_1.previewOrderChangeStep)(order.id);
    const createdReturnItems = (0, create_return_items_from_actions_1.createReturnItemsFromActionsStep)({
        returnId: orderReturn.id,
        changes: returnItemActions,
    });
    confirmIfReturnItemsArePresent({ returnItems: createdReturnItems });
    const returnShippingOptionId = (0, workflows_sdk_1.transform)({ orderPreview, orderReturn }, extractReturnShippingOptionId);
    (0, workflows_sdk_1.when)({ returnShippingOptionId }, ({ returnShippingOptionId }) => {
        return !!returnShippingOptionId;
    }).then(() => {
        const returnShippingOption = (0, common_1.useRemoteQueryStep)({
            entry_point: "shipping_options",
            fields: [
                "id",
                "provider_id",
                "service_zone.fulfillment_set.location.id",
                "service_zone.fulfillment_set.location.address.*",
            ],
            variables: {
                id: returnShippingOptionId,
            },
            list: false,
        }).config({ name: "return-shipping-option" });
        const fulfillmentData = (0, workflows_sdk_1.transform)({ order, items: createdReturnItems, returnShippingOption }, prepareFulfillmentData);
        const returnFulfillment = create_return_fulfillment_1.createReturnFulfillmentWorkflow.runAsStep(fulfillmentData);
        const link = (0, workflows_sdk_1.transform)({ orderReturn, fulfillment: returnFulfillment }, (data) => {
            return [
                {
                    [utils_1.Modules.ORDER]: { return_id: data.orderReturn.id },
                    [utils_1.Modules.FULFILLMENT]: { fulfillment_id: data.fulfillment.id },
                },
            ];
        });
        (0, common_1.createRemoteLinkStep)(link);
    });
    const updateReturnData = getUpdateReturnData({ orderReturn });
    (0, workflows_sdk_1.parallelize)((0, steps_1.updateReturnsStep)(updateReturnData), (0, confirm_order_changes_1.confirmOrderChanges)({
        changes: [orderChange],
        orderId: order.id,
        confirmed_by: input.confirmed_by,
    }), (0, common_1.emitEventStep)({
        eventName: utils_1.OrderWorkflowEvents.RETURN_REQUESTED,
        data: {
            order_id: order.id,
            return_id: orderReturn.id,
        },
    }));
    create_or_update_order_payment_collection_1.createOrUpdateOrderPaymentCollectionWorkflow.runAsStep({
        input: {
            order_id: order.id,
        },
    });
    return new workflows_sdk_1.WorkflowResponse(orderPreview);
});
