// @medusajs/workflow-engine-redis@2.6.1/dist/services/workflow-orchestrator.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _WorkflowOrchestratorService_logger;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowOrchestratorService = void 0;
const orchestration_1 = require("@medusajs/framework/orchestration");
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const timers_1 = require("timers");
const ulid_1 = require("ulid");
const AnySubscriber = "any";
class WorkflowOrchestratorService {
    constructor({ dataLoaderOnly, redisDistributedTransactionStorage, redisPublisher, redisSubscriber, sharedContainer, }) {
        this.instanceId = (0, ulid_1.ulid)();
        this.subscribers = new Map();
        this.activeStepsCount = 0;
        _WorkflowOrchestratorService_logger.set(this, void 0);
        this.container_ = sharedContainer;
        this.redisPublisher = redisPublisher;
        this.redisSubscriber = redisSubscriber;
        __classPrivateFieldSet(this, _WorkflowOrchestratorService_logger, this.container_.resolve("logger", { allowUnregistered: true }) ?? console, "f");
        redisDistributedTransactionStorage.setWorkflowOrchestratorService(this);
        if (!dataLoaderOnly) {
            orchestration_1.DistributedTransaction.setStorage(redisDistributedTransactionStorage);
            orchestration_1.WorkflowScheduler.setStorage(redisDistributedTransactionStorage);
        }
        this.redisDistributedTransactionStorage_ =
            redisDistributedTransactionStorage;
        this.redisSubscriber.on("message", async (_, message) => {
            const { instanceId, data } = JSON.parse(message);
            await this.notify(data, false, instanceId);
        });
    }
    async onApplicationShutdown() {
        await this.redisDistributedTransactionStorage_.onApplicationShutdown();
    }
    async onApplicationPrepareShutdown() {
        // eslint-disable-next-line max-len
        await this.redisDistributedTransactionStorage_.onApplicationPrepareShutdown();
        while (this.activeStepsCount > 0) {
            await new Promise((resolve) => (0, timers_1.setTimeout)(resolve, 1000));
        }
    }
    async onApplicationStart() {
        await this.redisDistributedTransactionStorage_.onApplicationStart();
    }
    async triggerParentStep(transaction, result) {
        const metadata = transaction.flow.metadata;
        const { parentStepIdempotencyKey } = metadata ?? {};
        if (parentStepIdempotencyKey) {
            const hasFailed = [
                utils_1.TransactionState.REVERTED,
                utils_1.TransactionState.FAILED,
            ].includes(transaction.flow.state);
            if (hasFailed) {
                await this.setStepFailure({
                    idempotencyKey: parentStepIdempotencyKey,
                    stepResponse: result,
                    options: {
                        logOnError: true,
                    },
                });
            }
            else {
                await this.setStepSuccess({
                    idempotencyKey: parentStepIdempotencyKey,
                    stepResponse: result,
                    options: {
                        logOnError: true,
                    },
                });
            }
        }
    }
    async run(workflowIdOrWorkflow, options, sharedContext = {}) {
        const { input, transactionId, resultFrom, logOnError, events: eventHandlers, container, } = options ?? {};
        let { throwOnError, context } = options ?? {};
        throwOnError ??= true;
        context ??= {};
        context.transactionId ??= transactionId ?? (0, ulid_1.ulid)();
        const workflowId = (0, utils_1.isString)(workflowIdOrWorkflow)
            ? workflowIdOrWorkflow
            : workflowIdOrWorkflow.getName();
        if (!workflowId) {
            throw new Error("Workflow ID is required");
        }
        const events = this.buildWorkflowEvents({
            customEventHandlers: eventHandlers,
            workflowId,
            transactionId: context.transactionId,
        });
        const exportedWorkflow = workflows_sdk_1.MedusaWorkflow.getWorkflow(workflowId);
        if (!exportedWorkflow) {
            throw new Error(`Workflow with id "${workflowId}" not found.`);
        }
        const originalOnFinishHandler = events.onFinish;
        delete events.onFinish;
        const ret = await exportedWorkflow.run({
            input,
            throwOnError: false,
            logOnError,
            resultFrom,
            context,
            events,
            container: container ?? this.container_,
        });
        const hasFinished = ret.transaction.hasFinished();
        const metadata = ret.transaction.getFlow().metadata;
        const { parentStepIdempotencyKey } = metadata ?? {};
        const hasFailed = [
            utils_1.TransactionState.REVERTED,
            utils_1.TransactionState.FAILED,
        ].includes(ret.transaction.getFlow().state);
        const acknowledgement = {
            transactionId: context.transactionId,
            workflowId: workflowId,
            parentStepIdempotencyKey,
            hasFinished,
            hasFailed,
        };
        if (hasFinished) {
            const { result, errors } = ret;
            await originalOnFinishHandler({
                transaction: ret.transaction,
                result,
                errors,
            });
            await this.triggerParentStep(ret.transaction, result);
        }
        if (throwOnError && ret.thrownError) {
            throw ret.thrownError;
        }
        return { acknowledgement, ...ret };
    }
    async getRunningTransaction(workflowId, transactionId, options, sharedContext = {}) {
        let { context } = options ?? {};
        if (!workflowId) {
            throw new Error("Workflow ID is required");
        }
        if (!transactionId) {
            throw new Error("TransactionId ID is required");
        }
        context ??= {};
        context.transactionId ??= transactionId;
        const exportedWorkflow = workflows_sdk_1.MedusaWorkflow.getWorkflow(workflowId);
        if (!exportedWorkflow) {
            throw new Error(`Workflow with id "${workflowId}" not found.`);
        }
        const transaction = await exportedWorkflow.getRunningTransaction(transactionId, context);
        return transaction;
    }
    async setStepSuccess({ idempotencyKey, stepResponse, options, }, sharedContext = {}) {
        const { context, logOnError, resultFrom, container, events: eventHandlers, } = options ?? {};
        let { throwOnError } = options ?? {};
        throwOnError ??= true;
        const [idempotencyKey_, { workflowId, transactionId }] = this.buildIdempotencyKeyAndParts(idempotencyKey);
        const exportedWorkflow = workflows_sdk_1.MedusaWorkflow.getWorkflow(workflowId);
        if (!exportedWorkflow) {
            throw new Error(`Workflow with id "${workflowId}" not found.`);
        }
        const events = this.buildWorkflowEvents({
            customEventHandlers: eventHandlers,
            transactionId,
            workflowId,
        });
        const originalOnFinishHandler = events.onFinish;
        delete events.onFinish;
        const ret = await exportedWorkflow.registerStepSuccess({
            idempotencyKey: idempotencyKey_,
            context,
            resultFrom,
            throwOnError: false,
            logOnError,
            events,
            response: stepResponse,
            container: container ?? this.container_,
        });
        if (ret.transaction.hasFinished()) {
            const { result, errors } = ret;
            await originalOnFinishHandler({
                transaction: ret.transaction,
                result,
                errors,
            });
            await this.triggerParentStep(ret.transaction, result);
        }
        if (throwOnError && ret.thrownError) {
            throw ret.thrownError;
        }
        return ret;
    }
    async setStepFailure({ idempotencyKey, stepResponse, options, }, sharedContext = {}) {
        const { context, logOnError, resultFrom, container, events: eventHandlers, } = options ?? {};
        let { throwOnError } = options ?? {};
        throwOnError ??= true;
        const [idempotencyKey_, { workflowId, transactionId }] = this.buildIdempotencyKeyAndParts(idempotencyKey);
        const exportedWorkflow = workflows_sdk_1.MedusaWorkflow.getWorkflow(workflowId);
        if (!exportedWorkflow) {
            throw new Error(`Workflow with id "${workflowId}" not found.`);
        }
        const events = this.buildWorkflowEvents({
            customEventHandlers: eventHandlers,
            transactionId,
            workflowId,
        });
        const originalOnFinishHandler = events.onFinish;
        delete events.onFinish;
        const ret = await exportedWorkflow.registerStepFailure({
            idempotencyKey: idempotencyKey_,
            context,
            resultFrom,
            throwOnError: false,
            logOnError,
            events,
            response: stepResponse,
            container: container ?? this.container_,
        });
        if (ret.transaction.hasFinished()) {
            const { result, errors } = ret;
            await originalOnFinishHandler({
                transaction: ret.transaction,
                result,
                errors,
            });
            await this.triggerParentStep(ret.transaction, result);
        }
        if (throwOnError && ret.thrownError) {
            throw ret.thrownError;
        }
        return ret;
    }
    subscribe({ workflowId, transactionId, subscriber, subscriberId }, sharedContext = {}) {
        subscriber._id = subscriberId;
        const subscribers = this.subscribers.get(workflowId) ?? new Map();
        // Subscribe instance to redis
        if (!this.subscribers.has(workflowId)) {
            void this.redisSubscriber.subscribe(this.getChannelName(workflowId));
        }
        const handlerIndex = (handlers) => {
            return handlers.indexOf((s) => s === subscriber || s._id === subscriberId);
        };
        if (transactionId) {
            const transactionSubscribers = subscribers.get(transactionId) ?? [];
            const subscriberIndex = handlerIndex(transactionSubscribers);
            if (subscriberIndex !== -1) {
                transactionSubscribers.slice(subscriberIndex, 1);
            }
            transactionSubscribers.push(subscriber);
            subscribers.set(transactionId, transactionSubscribers);
            this.subscribers.set(workflowId, subscribers);
            return;
        }
        const workflowSubscribers = subscribers.get(AnySubscriber) ?? [];
        const subscriberIndex = handlerIndex(workflowSubscribers);
        if (subscriberIndex !== -1) {
            workflowSubscribers.slice(subscriberIndex, 1);
        }
        workflowSubscribers.push(subscriber);
        subscribers.set(AnySubscriber, workflowSubscribers);
        this.subscribers.set(workflowId, subscribers);
    }
    unsubscribe({ workflowId, transactionId, subscriberOrId }, sharedContext = {}) {
        const subscribers = this.subscribers.get(workflowId) ?? new Map();
        const filterSubscribers = (handlers) => {
            return handlers.filter((handler) => {
                return handler._id
                    ? handler._id !== subscriberOrId
                    : handler !== subscriberOrId;
            });
        };
        // Unsubscribe instance
        if (!this.subscribers.has(workflowId)) {
            void this.redisSubscriber.unsubscribe(this.getChannelName(workflowId));
        }
        if (transactionId) {
            const transactionSubscribers = subscribers.get(transactionId) ?? [];
            const newTransactionSubscribers = filterSubscribers(transactionSubscribers);
            subscribers.set(transactionId, newTransactionSubscribers);
            this.subscribers.set(workflowId, subscribers);
            return;
        }
        const workflowSubscribers = subscribers.get(AnySubscriber) ?? [];
        const newWorkflowSubscribers = filterSubscribers(workflowSubscribers);
        subscribers.set(AnySubscriber, newWorkflowSubscribers);
        this.subscribers.set(workflowId, subscribers);
    }
    async notify(options, publish = true, instanceId = this.instanceId) {
        if (!publish && instanceId === this.instanceId) {
            return;
        }
        if (publish) {
            const channel = this.getChannelName(options.workflowId);
            const message = JSON.stringify({
                instanceId: this.instanceId,
                data: options,
            });
            await this.redisPublisher.publish(channel, message);
        }
        const { eventType, workflowId, transactionId, errors, result, step, response, } = options;
        const subscribers = this.subscribers.get(workflowId) ?? new Map();
        const notifySubscribers = (handlers) => {
            handlers.forEach((handler) => {
                const args = {
                    eventType,
                    workflowId,
                    transactionId,
                    step,
                    response,
                    result,
                    errors,
                };
                const isPromise = "then" in handler;
                if (isPromise) {
                    ;
                    handler(args).catch((e) => {
                        __classPrivateFieldGet(this, _WorkflowOrchestratorService_logger, "f").error(e);
                    });
                }
                else {
                    try {
                        handler(args);
                    }
                    catch (e) {
                        __classPrivateFieldGet(this, _WorkflowOrchestratorService_logger, "f").error(e);
                    }
                }
            });
        };
        if (transactionId) {
            const transactionSubscribers = subscribers.get(transactionId) ?? [];
            notifySubscribers(transactionSubscribers);
            // removes transaction id subscribers on finish
            if (eventType === "onFinish") {
                subscribers.delete(transactionId);
            }
        }
        const workflowSubscribers = subscribers.get(AnySubscriber) ?? [];
        notifySubscribers(workflowSubscribers);
    }
    getChannelName(workflowId) {
        return `orchestrator:${workflowId}`;
    }
    buildWorkflowEvents({ customEventHandlers, workflowId, transactionId, }) {
        const notify = async ({ eventType, step, result, response, errors, }) => {
            await this.notify({
                workflowId,
                transactionId,
                eventType,
                response,
                step,
                result,
                errors,
            });
        };
        return {
            onTimeout: async ({ transaction }) => {
                customEventHandlers?.onTimeout?.({ transaction });
                await notify({ eventType: "onTimeout" });
            },
            onBegin: async ({ transaction }) => {
                customEventHandlers?.onBegin?.({ transaction });
                await notify({ eventType: "onBegin" });
            },
            onResume: async ({ transaction }) => {
                customEventHandlers?.onResume?.({ transaction });
                await notify({ eventType: "onResume" });
            },
            onCompensateBegin: async ({ transaction }) => {
                customEventHandlers?.onCompensateBegin?.({ transaction });
                await notify({ eventType: "onCompensateBegin" });
            },
            onFinish: async ({ transaction, result, errors }) => {
                customEventHandlers?.onFinish?.({ transaction, result, errors });
                await notify({ eventType: "onFinish" });
            },
            onStepBegin: async ({ step, transaction }) => {
                customEventHandlers?.onStepBegin?.({ step, transaction });
                this.activeStepsCount++;
                await notify({ eventType: "onStepBegin", step });
            },
            onStepSuccess: async ({ step, transaction }) => {
                const stepName = step.definition.action;
                const response = await (0, workflows_sdk_1.resolveValue)(transaction.getContext().invoke[stepName], transaction);
                customEventHandlers?.onStepSuccess?.({ step, transaction, response });
                await notify({ eventType: "onStepSuccess", step, response });
                this.activeStepsCount--;
            },
            onStepFailure: async ({ step, transaction }) => {
                const stepName = step.definition.action;
                const errors = transaction
                    .getErrors(orchestration_1.TransactionHandlerType.INVOKE)
                    .filter((err) => err.action === stepName);
                customEventHandlers?.onStepFailure?.({ step, transaction, errors });
                await notify({ eventType: "onStepFailure", step, errors });
                this.activeStepsCount--;
            },
            onStepAwaiting: async ({ step, transaction }) => {
                customEventHandlers?.onStepAwaiting?.({ step, transaction });
                await notify({ eventType: "onStepAwaiting", step });
                this.activeStepsCount--;
            },
            onCompensateStepSuccess: async ({ step, transaction }) => {
                const stepName = step.definition.action;
                const response = transaction.getContext().compensate[stepName];
                customEventHandlers?.onCompensateStepSuccess?.({
                    step,
                    transaction,
                    response,
                });
                await notify({ eventType: "onCompensateStepSuccess", step, response });
            },
            onCompensateStepFailure: async ({ step, transaction }) => {
                const stepName = step.definition.action;
                const errors = transaction
                    .getErrors(orchestration_1.TransactionHandlerType.COMPENSATE)
                    .filter((err) => err.action === stepName);
                customEventHandlers?.onStepFailure?.({ step, transaction, errors });
                await notify({ eventType: "onCompensateStepFailure", step, errors });
            },
        };
    }
    buildIdempotencyKeyAndParts(idempotencyKey) {
        const parts = {
            workflowId: "",
            transactionId: "",
            stepId: "",
            action: "invoke",
        };
        let idempotencyKey_ = idempotencyKey;
        const setParts = (workflowId, transactionId, stepId, action) => {
            parts.workflowId = workflowId;
            parts.transactionId = transactionId;
            parts.stepId = stepId;
            parts.action = action;
        };
        if (!(0, utils_1.isString)(idempotencyKey)) {
            const { workflowId, transactionId, stepId, action } = idempotencyKey;
            idempotencyKey_ = [workflowId, transactionId, stepId, action].join(":");
            setParts(workflowId, transactionId, stepId, action);
        }
        else {
            const [workflowId, transactionId, stepId, action] = idempotencyKey_.split(":");
            setParts(workflowId, transactionId, stepId, action);
        }
        return [idempotencyKey_, parts];
    }
}
exports.WorkflowOrchestratorService = WorkflowOrchestratorService;
_WorkflowOrchestratorService_logger = new WeakMap();
__decorate([
    (0, utils_1.InjectSharedContext)(),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], WorkflowOrchestratorService.prototype, "run", null);
__decorate([
    (0, utils_1.InjectSharedContext)(),
    __param(3, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, String, Object, Object]),
    __metadata("design:returntype", Promise)
], WorkflowOrchestratorService.prototype, "getRunningTransaction", null);
__decorate([
    (0, utils_1.InjectSharedContext)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], WorkflowOrchestratorService.prototype, "setStepSuccess", null);
__decorate([
    (0, utils_1.InjectSharedContext)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], WorkflowOrchestratorService.prototype, "setStepFailure", null);
__decorate([
    (0, utils_1.InjectSharedContext)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", void 0)
], WorkflowOrchestratorService.prototype, "subscribe", null);
__decorate([
    (0, utils_1.InjectSharedContext)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", void 0)
], WorkflowOrchestratorService.prototype, "unsubscribe", null);
