// @medusajs/pricing@2.6.1/dist/services/pricing-module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@medusajs/framework/types");
const utils_1 = require("@medusajs/framework/utils");
const _models_1 = require("../models");
const _utils_1 = require("../utils");
const joiner_config_1 = require("../joiner-config");
const generateMethodForModels = {
    PriceSet: _models_1.PriceSet,
    PriceList: _models_1.PriceList,
    PriceListRule: _models_1.PriceListRule,
    PriceRule: _models_1.PriceRule,
    Price: _models_1.Price,
    PricePreference: _models_1.PricePreference,
};
class PricingModuleService extends utils_1.ModulesSdkUtils.MedusaService(generateMethodForModels) {
    constructor({ baseRepository, pricingRepository, priceSetService, priceRuleService, priceService, pricePreferenceService, priceListService, priceListRuleService, }, moduleDeclaration) {
        // @ts-ignore
        super(...arguments);
        this.moduleDeclaration = moduleDeclaration;
        this.baseRepository_ = baseRepository;
        this.pricingRepository_ = pricingRepository;
        this.priceSetService_ = priceSetService;
        this.priceRuleService_ = priceRuleService;
        this.priceService_ = priceService;
        this.pricePreferenceService_ = pricePreferenceService;
        this.priceListService_ = priceListService;
        this.priceListRuleService_ = priceListRuleService;
    }
    __joinerConfig() {
        return joiner_config_1.joinerConfig;
    }
    setupCalculatedPriceConfig_(filters, config) {
        const fieldIdx = config.relations?.indexOf("calculated_price");
        const shouldCalculatePrice = fieldIdx > -1;
        const pricingContext = filters.context ?? {};
        delete filters.context;
        if (!shouldCalculatePrice) {
            return;
        }
        // cleanup virtual field "calculated_price"
        config.relations?.splice(fieldIdx, 1);
        return pricingContext;
    }
    // @ts-expect-error
    async retrievePriceSet(id, config, sharedContext) {
        const priceSet = await this.priceSetService_.retrieve(id, this.normalizePriceSetConfig(config), sharedContext);
        return await this.baseRepository_.serialize(priceSet);
    }
    // @ts-expect-error
    async listPriceSets(filters = {}, config = {}, sharedContext = {}) {
        const normalizedConfig = this.normalizePriceSetConfig(config);
        const pricingContext = this.setupCalculatedPriceConfig_(filters, normalizedConfig);
        const priceSets = await super.listPriceSets(filters, normalizedConfig, sharedContext);
        if (!pricingContext || !priceSets.length) {
            return priceSets;
        }
        const calculatedPrices = await this.calculatePrices({ id: priceSets.map((p) => p.id) }, { context: pricingContext }, sharedContext);
        const calculatedPricesMap = new Map();
        for (const calculatedPrice of calculatedPrices) {
            calculatedPricesMap.set(calculatedPrice.id, calculatedPrice);
        }
        for (const priceSet of priceSets) {
            const calculatedPrice = calculatedPricesMap.get(priceSet.id);
            priceSet.calculated_price = calculatedPrice ?? null;
        }
        return priceSets;
    }
    // @ts-expect-error
    async listAndCountPriceSets(filters = {}, config = {}, sharedContext = {}) {
        const normalizedConfig = this.normalizePriceSetConfig(config);
        const pricingContext = this.setupCalculatedPriceConfig_(filters, normalizedConfig);
        const [priceSets, count] = await super.listAndCountPriceSets(filters, normalizedConfig, sharedContext);
        if (!pricingContext || !priceSets.length) {
            return [priceSets, count];
        }
        const calculatedPrices = await this.calculatePrices({ id: priceSets.map((p) => p.id) }, { context: pricingContext }, sharedContext);
        const calculatedPricesMap = new Map();
        for (const calculatedPrice of calculatedPrices) {
            calculatedPricesMap.set(calculatedPrice.id, calculatedPrice);
        }
        for (const priceSet of priceSets) {
            const calculatedPrice = calculatedPricesMap.get(priceSet.id);
            priceSet.calculated_price = calculatedPrice ?? null;
        }
        return [priceSets, count];
    }
    async calculatePrices(pricingFilters, pricingContext = { context: {} }, sharedContext = {}) {
        const results = await this.pricingRepository_.calculatePrices(pricingFilters, pricingContext, sharedContext);
        const pricesSetPricesMap = (0, utils_1.groupBy)(results, "price_set_id");
        const priceIds = [];
        pricesSetPricesMap.forEach((prices, key) => {
            const priceListPrice = prices.find((p) => p.price_list_id);
            const defaultPrice = prices?.find((p) => !p.price_list_id);
            if (!prices.length || (!priceListPrice && !defaultPrice)) {
                pricesSetPricesMap.delete(key);
                return;
            }
            let calculatedPrice = defaultPrice;
            let originalPrice = defaultPrice;
            /**
             * When deciding which price to use we follow the following logic:
             * - If the price list is of type OVERRIDE, we always use the price list price.
             * - If the price list is of type SALE, we use the lowest price between the price list price and the default price
             */
            if (priceListPrice) {
                switch (priceListPrice.price_list_type) {
                    case utils_1.PriceListType.OVERRIDE:
                        calculatedPrice = priceListPrice;
                        originalPrice = priceListPrice;
                        break;
                    case utils_1.PriceListType.SALE: {
                        let lowestPrice = priceListPrice;
                        if (defaultPrice?.amount && priceListPrice.amount) {
                            lowestPrice = utils_1.MathBN.lte(priceListPrice.amount, defaultPrice.amount)
                                ? priceListPrice
                                : defaultPrice;
                        }
                        calculatedPrice = lowestPrice;
                        break;
                    }
                }
            }
            pricesSetPricesMap.set(key, { calculatedPrice, originalPrice });
            priceIds.push(...(0, utils_1.deduplicate)([calculatedPrice?.id, originalPrice?.id].filter(Boolean)));
        });
        // We use the price rules to get the right preferences for the price
        const priceRulesForPrices = await this.priceRuleService_.list({ price_id: priceIds }, {});
        const priceRulesPriceMap = (0, utils_1.groupBy)(priceRulesForPrices, "price_id");
        // Note: For now the preferences are intentionally kept very simple and explicit - they use either the region or currency,
        // so we hard-code those as the possible filters here. This can be made more flexible if needed later on.
        const pricingPreferences = await this.pricePreferenceService_.list({
            $or: Object.entries(pricingContext)
                .filter(([key, val]) => {
                return key === "region_id" || key === "currency_code";
            })
                .map(([key, val]) => ({
                attribute: key,
                value: val,
            })),
        }, {}, sharedContext);
        const calculatedPrices = pricingFilters.id
            .map((priceSetId) => {
            const prices = pricesSetPricesMap.get(priceSetId);
            if (!prices) {
                return null;
            }
            const { calculatedPrice, originalPrice, } = prices;
            return {
                id: priceSetId,
                is_calculated_price_price_list: !!calculatedPrice?.price_list_id,
                is_calculated_price_tax_inclusive: isTaxInclusive(priceRulesPriceMap.get(calculatedPrice.id), pricingPreferences, calculatedPrice.currency_code, pricingContext.context?.region_id),
                calculated_amount: (0, utils_1.isPresent)(calculatedPrice?.amount)
                    ? parseFloat(calculatedPrice?.amount)
                    : null,
                raw_calculated_amount: calculatedPrice?.raw_amount || null,
                is_original_price_price_list: !!originalPrice?.price_list_id,
                is_original_price_tax_inclusive: originalPrice?.id
                    ? isTaxInclusive(priceRulesPriceMap.get(originalPrice.id), pricingPreferences, originalPrice.currency_code || calculatedPrice.currency_code, pricingContext.context?.region_id)
                    : false,
                original_amount: (0, utils_1.isPresent)(originalPrice?.amount)
                    ? parseFloat(originalPrice?.amount)
                    : null,
                raw_original_amount: originalPrice?.raw_amount || null,
                currency_code: calculatedPrice?.currency_code || null,
                calculated_price: {
                    id: calculatedPrice?.id || null,
                    price_list_id: calculatedPrice?.price_list_id || null,
                    price_list_type: calculatedPrice?.price_list_type || null,
                    min_quantity: parseInt(calculatedPrice?.min_quantity || "") || null,
                    max_quantity: parseInt(calculatedPrice?.max_quantity || "") || null,
                },
                original_price: {
                    id: originalPrice?.id || null,
                    price_list_id: originalPrice?.price_list_id || null,
                    price_list_type: originalPrice?.price_list_type || null,
                    min_quantity: parseInt(originalPrice?.min_quantity || "") || null,
                    max_quantity: parseInt(originalPrice?.max_quantity || "") || null,
                },
            };
        })
            .filter(Boolean);
        return JSON.parse(JSON.stringify(calculatedPrices));
    }
    // @ts-expect-error
    async createPriceSets(data, sharedContext = {}) {
        const input = Array.isArray(data) ? data : [data];
        const priceSets = await this.createPriceSets_(input, sharedContext);
        // TODO: Remove the need to refetch the data here
        const dbPriceSets = await this.listPriceSets({ id: priceSets.map((p) => p.id) }, this.normalizePriceSetConfig({
            relations: ["prices", "prices.price_rules"],
        }), sharedContext);
        // Ensure the output to be in the same order as the input
        const results = priceSets.map((priceSet) => {
            return dbPriceSets.find((p) => p.id === priceSet.id);
        });
        return await this.baseRepository_.serialize(Array.isArray(data) ? results : results[0]);
    }
    async upsertPriceSets(data, sharedContext = {}) {
        const input = Array.isArray(data) ? data : [data];
        const forUpdate = input.filter((priceSet) => !!priceSet.id);
        const forCreate = input.filter((priceSet) => !priceSet.id);
        const operations = [];
        if (forCreate.length) {
            operations.push(this.createPriceSets_(forCreate, sharedContext));
        }
        if (forUpdate.length) {
            operations.push(this.updatePriceSets_(forUpdate, sharedContext));
        }
        const result = (await (0, utils_1.promiseAll)(operations)).flat();
        return await this.baseRepository_.serialize(Array.isArray(data) ? result : result[0]);
    }
    // @ts-expect-error
    async updatePriceSets(idOrSelector, data, sharedContext = {}) {
        let normalizedInput = [];
        if ((0, utils_1.isString)(idOrSelector)) {
            // Check if the ID exists, it will throw if not.
            await this.priceSetService_.retrieve(idOrSelector, {}, sharedContext);
            normalizedInput = [{ id: idOrSelector, ...data }];
        }
        else {
            const priceSets = await this.priceSetService_.list(idOrSelector, {}, sharedContext);
            normalizedInput = priceSets.map((priceSet) => ({
                id: priceSet.id,
                ...data,
            }));
        }
        const updateResult = await this.updatePriceSets_(normalizedInput, sharedContext);
        const priceSets = await this.baseRepository_.serialize(updateResult);
        return (0, utils_1.isString)(idOrSelector) ? priceSets[0] : priceSets;
    }
    async updatePriceSets_(data, sharedContext = {}) {
        // TODO: Since money IDs are rarely passed, this will delete all previous data and insert new entries.
        // We can make the `insert` inside upsertWithReplace do an `upsert` instead to avoid this
        const normalizedData = await this.normalizeUpdateData(data);
        const priceListPrices = await this.priceService_.list({
            price_set_id: normalizedData.map(({ id }) => id),
            price_list_id: { $ne: null },
        });
        const prices = normalizedData.flatMap((priceSet) => priceSet.prices || []);
        const { entities: upsertedPrices, performedActions } = await this.priceService_.upsertWithReplace(prices, { relations: ["price_rules"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        const priceSetsToUpsert = normalizedData.map((priceSet) => {
            const { prices, ...rest } = priceSet;
            return {
                ...rest,
                prices: [
                    ...upsertedPrices
                        .filter((p) => p.price_set_id === priceSet.id)
                        .map((price) => {
                        // @ts-ignore
                        delete price.price_rules;
                        return price;
                    }),
                    ...priceListPrices
                        .filter((p) => p.price_set_id === priceSet.id)
                        .map((price) => ({
                        id: price.id,
                        amount: price.amount,
                        price_set_id: price.price_set_id,
                        price_list_id: price.price_list_id,
                    })),
                ],
            };
        });
        const { entities: priceSets, performedActions: priceSetPerformedActions } = await this.priceSetService_.upsertWithReplace(priceSetsToUpsert, { relations: ["prices"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions: priceSetPerformedActions,
            sharedContext,
        });
        return priceSets.map((ps) => {
            if (ps.prices) {
                ps.prices = ps.prices.filter((p) => !p.price_list_id);
            }
            return ps;
        });
    }
    async normalizeUpdateData(data) {
        return data.map((priceSet) => {
            return {
                ...priceSet,
                prices: this.normalizePrices(priceSet.prices?.map((p) => ({ ...p, price_set_id: priceSet.id })), []),
            };
        });
    }
    normalizePrices(data, existingPrices, priceListId) {
        const pricesToUpsert = new Map();
        const existingPricesMap = new Map();
        existingPrices?.forEach((price) => {
            existingPricesMap.set(hashPrice(price), price);
        });
        data?.forEach((price) => {
            const cleanRules = price.rules ? (0, utils_1.removeNullish)(price.rules) : {};
            const ruleOperators = Object.values(utils_1.PricingRuleOperator);
            const rules = Object.entries(cleanRules)
                .map(([attribute, value]) => {
                if (Array.isArray(value)) {
                    return value.map((customRule) => {
                        if (!ruleOperators.includes(customRule.operator)) {
                            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `operator should be one of ${ruleOperators.join(", ")}`);
                        }
                        if (typeof customRule.value !== "number") {
                            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `value should be a number`);
                        }
                        return {
                            attribute,
                            operator: customRule.operator,
                            value: customRule.value,
                        };
                    });
                }
                return {
                    attribute,
                    value,
                };
            })
                .flat(1);
            const hasRulesInput = (0, utils_1.isPresent)(price.rules);
            const entry = {
                ...price,
                price_list_id: priceListId,
                price_rules: hasRulesInput ? rules : undefined,
                rules_count: hasRulesInput ? rules.length : undefined,
            };
            delete entry.rules;
            const entryHash = hashPrice(entry);
            // We want to keep the existing rules as they might already have ids, but any other data should come from the updated input
            const existing = existingPricesMap.get(entryHash);
            pricesToUpsert.set(entryHash, {
                ...entry,
                id: existing?.id ?? entry.id,
                price_rules: existing?.price_rules ?? entry.price_rules,
            });
            return entry;
        });
        return Array.from(pricesToUpsert.values());
    }
    async addPrices(data, sharedContext = {}) {
        const input = Array.isArray(data) ? data : [data];
        await this.addPrices_(input, sharedContext);
        const dbPrices = await this.listPriceSets({ id: input.map((d) => d.priceSetId) }, { relations: ["prices"] }, sharedContext);
        const orderedPriceSets = input.map((inputItem) => {
            return dbPrices.find((p) => p.id === inputItem.priceSetId);
        });
        return Array.isArray(data) ? orderedPriceSets : orderedPriceSets[0];
    }
    // @ts-ignore
    async createPriceLists(data, sharedContext = {}) {
        const priceLists = await this.createPriceLists_(data, sharedContext);
        return await this.baseRepository_.serialize(priceLists);
    }
    // @ts-ignore
    async updatePriceLists(data, sharedContext = {}) {
        const priceLists = await this.updatePriceLists_(data, sharedContext);
        return await this.baseRepository_.serialize(priceLists);
    }
    async updatePriceListPrices(data, sharedContext = {}) {
        const prices = await this.updatePriceListPrices_(data, sharedContext);
        return await this.baseRepository_.serialize(prices);
    }
    async removePrices(ids, sharedContext = {}) {
        await this.removePrices_(ids, sharedContext);
    }
    async addPriceListPrices(data, sharedContext = {}) {
        const prices = await this.addPriceListPrices_(data, sharedContext);
        return await this.baseRepository_.serialize(prices);
    }
    async setPriceListRules(data, sharedContext = {}) {
        const [priceList] = await this.setPriceListRules_([data], sharedContext);
        return await this.baseRepository_.serialize(priceList);
    }
    async removePriceListRules(data, sharedContext = {}) {
        const [priceList] = await this.removePriceListRules_([data], sharedContext);
        return await this.baseRepository_.serialize(priceList);
    }
    async createPricePreferences(data, sharedContext = {}) {
        const normalized = Array.isArray(data) ? data : [data];
        const preferences = await this.createPricePreferences_(normalized, sharedContext);
        const serialized = await this.baseRepository_.serialize(preferences);
        return Array.isArray(data) ? serialized : serialized[0];
    }
    async upsertPricePreferences(data, sharedContext = {}) {
        const input = Array.isArray(data) ? data : [data];
        const forUpdate = input.filter((pricePreference) => !!pricePreference.id);
        const forCreate = input.filter((pricePreference) => !pricePreference.id);
        const operations = [];
        if (forCreate.length) {
            operations.push(this.createPricePreferences_(forCreate, sharedContext));
        }
        if (forUpdate.length) {
            operations.push(this.updatePricePreferences_(forUpdate, sharedContext));
        }
        const result = (await (0, utils_1.promiseAll)(operations)).flat();
        return await this.baseRepository_.serialize(Array.isArray(data) ? result : result[0]);
    }
    // @ts-expect-error
    async updatePricePreferences(idOrSelector, data, sharedContext = {}) {
        let normalizedInput = [];
        if ((0, utils_1.isString)(idOrSelector)) {
            // Check if the ID exists, it will throw if not.
            await this.pricePreferenceService_.retrieve(idOrSelector, {}, sharedContext);
            normalizedInput = [{ id: idOrSelector, ...data }];
        }
        else {
            const pricePreferences = await this.pricePreferenceService_.list(idOrSelector, {}, sharedContext);
            normalizedInput = pricePreferences.map((pricePreference) => ({
                id: pricePreference.id,
                ...data,
            }));
        }
        const updateResult = await this.updatePricePreferences_(normalizedInput, sharedContext);
        const pricePreferences = await this.baseRepository_.serialize(updateResult);
        return (0, utils_1.isString)(idOrSelector) ? pricePreferences[0] : pricePreferences;
    }
    async createPricePreferences_(data, sharedContext = {}) {
        const preferences = await this.pricePreferenceService_.create(data.map((d) => ({
            ...d,
            is_tax_inclusive: d.is_tax_inclusive ?? false,
        })), sharedContext);
        return preferences;
    }
    async updatePricePreferences_(data, sharedContext = {}) {
        const preferences = await this.pricePreferenceService_.update(data, sharedContext);
        return preferences;
    }
    async createPriceSets_(data, sharedContext = {}) {
        const input = Array.isArray(data) ? data : [data];
        const toCreate = input.map((inputData) => {
            const entry = {
                ...inputData,
                prices: this.normalizePrices(inputData.prices, []),
            };
            return entry;
        });
        // Bulk create price sets
        const createdPriceSets = await this.priceSetService_.create(toCreate, sharedContext);
        const eventsData = createdPriceSets.reduce((eventsData, priceSet) => {
            eventsData.priceSets.push({
                id: priceSet.id,
            });
            priceSet.prices.map((price) => {
                eventsData.prices.push({
                    id: price.id,
                });
                price.price_rules.map((priceRule) => {
                    eventsData.priceRules.push({
                        id: priceRule.id,
                    });
                });
            });
            return eventsData;
        }, {
            priceSets: [],
            priceRules: [],
            prices: [],
        });
        _utils_1.eventBuilders.createdPriceSet({
            data: eventsData.priceSets,
            sharedContext,
        });
        _utils_1.eventBuilders.createdPrice({
            data: eventsData.prices,
            sharedContext,
        });
        _utils_1.eventBuilders.createdPriceRule({
            data: eventsData.priceRules,
            sharedContext,
        });
        return createdPriceSets;
    }
    async addPrices_(input, sharedContext = {}) {
        const priceSets = await this.listPriceSets({ id: input.map((d) => d.priceSetId) }, { relations: ["prices", "prices.price_rules"] }, sharedContext);
        const existingPrices = priceSets
            .map((p) => p.prices)
            .flat();
        const pricesToUpsert = input
            .map((addPrice) => this.normalizePrices(addPrice.prices?.map((p) => ({
            ...p,
            price_set_id: addPrice.priceSetId,
        })), existingPrices))
            .filter(Boolean)
            .flat();
        const priceSetMap = new Map(priceSets.map((p) => [p.id, p]));
        pricesToUpsert.forEach((price) => {
            const priceSet = priceSetMap.get(price.price_set_id);
            if (!priceSet) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Price set with id: ${price.price_set_id} not found`);
            }
        });
        const { entities, performedActions } = await this.priceService_.upsertWithReplace(pricesToUpsert, { relations: ["price_rules"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        return entities;
    }
    async createPriceLists_(data, sharedContext = {}) {
        const normalized = this.normalizePriceListDate(data);
        const priceListsToCreate = normalized.map((priceListData) => {
            const entry = {
                ...priceListData,
                rules: undefined,
            };
            if (priceListData.prices) {
                entry.prices = this.normalizePrices(priceListData.prices, []);
            }
            if (priceListData.rules) {
                const cleanRules = priceListData.rules
                    ? (0, utils_1.removeNullish)(priceListData.rules)
                    : {};
                const rules = Object.entries(cleanRules);
                const numberOfRules = rules.length;
                const rulesDataMap = new Map();
                rules.map(([attribute, value]) => {
                    const rule = {
                        attribute,
                        value,
                    };
                    rulesDataMap.set(JSON.stringify(rule), rule);
                });
                entry.price_list_rules = Array.from(rulesDataMap.values());
                entry.rules_count = numberOfRules;
            }
            return entry;
        });
        const priceLists = await this.priceListService_.create(priceListsToCreate, sharedContext);
        /**
         * Preparing data for emitting events
         */
        const eventsData = priceLists.reduce((eventsData, priceList) => {
            eventsData.priceList.push({
                id: priceList.id,
            });
            priceList.price_list_rules.map((listRule) => {
                eventsData.priceListRules.push({
                    id: listRule.id,
                });
            });
            priceList.prices.map((price) => {
                eventsData.prices.push({
                    id: price.id,
                });
                price.price_rules.map((priceRule) => {
                    eventsData.priceRules.push({
                        id: priceRule.id,
                    });
                });
            });
            return eventsData;
        }, {
            priceList: [],
            priceListRules: [],
            priceRules: [],
            prices: [],
        });
        /**
         * Emitting events for all created entities
         */
        _utils_1.eventBuilders.createdPriceList({
            data: eventsData.priceList,
            sharedContext,
        });
        _utils_1.eventBuilders.createdPriceListRule({
            data: eventsData.priceListRules,
            sharedContext,
        });
        _utils_1.eventBuilders.createdPrice({
            data: eventsData.prices,
            sharedContext,
        });
        _utils_1.eventBuilders.createdPriceRule({
            data: eventsData.priceRules,
            sharedContext,
        });
        return priceLists;
    }
    async updatePriceLists_(data, sharedContext = {}) {
        const existingPriceLists = await this.priceListService_.list({ id: data.map((d) => d.id) }, {}, sharedContext);
        if (existingPriceLists.length !== data.length) {
            const diff = (0, utils_1.arrayDifference)(data.map((d) => d.id), existingPriceLists.map((p) => p.id));
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Price lists with ids: '${diff.join(", ")}' not found`);
        }
        const normalizedData = this.normalizePriceListDate(data).map((priceList) => {
            const entry = {
                ...priceList,
                rules: undefined,
                price_list_rules: undefined,
            };
            if (typeof priceList.rules === "object") {
                const cleanRules = priceList.rules
                    ? (0, utils_1.removeNullish)(priceList.rules)
                    : {};
                const rules = Object.entries(cleanRules);
                const numberOfRules = rules.length;
                const rulesDataMap = new Map();
                rules.map(([attribute, value]) => {
                    const rule = {
                        attribute,
                        value,
                    };
                    rulesDataMap.set(JSON.stringify(rule), rule);
                });
                entry.price_list_rules = Array.from(rulesDataMap.values());
                entry.rules_count = numberOfRules;
            }
            return entry;
        });
        const { entities, performedActions } = await this.priceListService_.upsertWithReplace(normalizedData, {
            relations: ["price_list_rules"],
        });
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        return entities;
    }
    async updatePriceListPrices_(data, sharedContext = {}) {
        const priceLists = await this.listPriceLists({ id: data.map((p) => p.price_list_id) }, { relations: ["prices", "prices.price_rules"] }, sharedContext);
        const existingPrices = priceLists
            .map((p) => p.prices ?? [])
            .flat();
        const pricesToUpsert = data
            .map((addPrice) => this.normalizePrices(addPrice.prices, existingPrices, addPrice.price_list_id))
            .filter(Boolean)
            .flat();
        const priceListMap = new Map(priceLists.map((p) => [p.id, p]));
        for (const { price_list_id: priceListId } of data) {
            const priceList = priceListMap.get(priceListId);
            if (!priceList) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Price list with id: ${priceListId} not found`);
            }
        }
        const { entities, performedActions } = await this.priceService_.upsertWithReplace(pricesToUpsert, { relations: ["price_rules"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        return entities;
    }
    async removePrices_(ids, sharedContext = {}) {
        await this.priceService_.delete(ids, sharedContext);
    }
    async addPriceListPrices_(data, sharedContext = {}) {
        const priceLists = await this.listPriceLists({ id: data.map((p) => p.price_list_id) }, { relations: ["prices", "prices.price_rules"] }, sharedContext);
        const existingPrices = priceLists
            .map((p) => p.prices ?? [])
            .flat();
        const pricesToUpsert = data
            .map((addPrice) => this.normalizePrices(addPrice.prices, existingPrices, addPrice.price_list_id))
            .filter(Boolean)
            .flat();
        const priceListMap = new Map(priceLists.map((p) => [p.id, p]));
        pricesToUpsert.forEach((price) => {
            const priceList = priceListMap.get(price.price_list_id);
            if (!priceList) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Price list with id: ${price.price_list_id} not found`);
            }
        });
        const { entities, performedActions } = await this.priceService_.upsertWithReplace(pricesToUpsert, { relations: ["price_rules"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        return entities;
    }
    async setPriceListRules_(data, sharedContext = {}) {
        // TODO: re think this method
        const priceLists = await this.priceListService_.list({ id: data.map((d) => d.price_list_id) }, {
            relations: ["price_list_rules"],
        }, sharedContext);
        const rulesMap = new Map();
        data.forEach((rule) => {
            if (!rulesMap.has(rule.price_list_id)) {
                rulesMap.set(rule.price_list_id, []);
            }
            Object.entries(rule.rules).forEach(([key, value]) => {
                rulesMap.get(rule.price_list_id).push([key, value]);
            });
        });
        const priceListsUpsert = priceLists
            .map((priceList) => {
            const priceListRules = priceList.price_list_rules;
            const allRules = new Map(priceListRules.toArray().map((r) => [r.attribute, r.value]));
            const rules = rulesMap.get(priceList.id);
            if (!rules?.length) {
                return;
            }
            rules.forEach(([key, value]) => {
                allRules.set(key, value);
            });
            return {
                ...priceList,
                rules_count: allRules.size,
                price_list_rules: Array.from(allRules).map(([attribute, value]) => ({
                    attribute,
                    value,
                })),
            };
        })
            .filter(Boolean);
        const { entities, performedActions } = await this.priceListService_.upsertWithReplace(priceListsUpsert, { relations: ["price_list_rules"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        return entities;
    }
    async removePriceListRules_(data, sharedContext = {}) {
        // TODO: re think this method
        const priceLists = await this.priceListService_.list({ id: data.map((d) => d.price_list_id) }, {
            relations: ["price_list_rules"],
        }, sharedContext);
        const rulesMap = new Map();
        data.forEach((rule) => {
            if (!rulesMap.has(rule.price_list_id)) {
                rulesMap.set(rule.price_list_id, []);
            }
            rule.rules.forEach((key) => {
                rulesMap.get(rule.price_list_id).push([key, undefined]);
            });
        });
        const priceListsUpsert = priceLists
            .map((priceList) => {
            const priceListRules = priceList.price_list_rules;
            const allRules = new Map(priceListRules.toArray().map((r) => [r.attribute, r.value]));
            const rules = rulesMap.get(priceList.id);
            if (!rules?.length) {
                return;
            }
            rules.forEach(([key, value]) => {
                allRules.set(key, value);
            });
            return {
                ...priceList,
                rules_count: allRules.size,
                price_list_rules: Array.from(allRules)
                    .map(([attribute, value]) => ({
                    attribute,
                    value,
                }))
                    .filter((r) => !!r.value),
            };
        })
            .filter(Boolean);
        const { entities, performedActions } = await this.priceListService_.upsertWithReplace(priceListsUpsert, { relations: ["price_list_rules"] }, sharedContext);
        composeAllEvents({
            eventBuilders: _utils_1.eventBuilders,
            performedActions,
            sharedContext,
        });
        return entities;
    }
    normalizePriceListDate(data) {
        return data.map((priceListData) => {
            (0, _utils_1.validatePriceListDates)(priceListData);
            if (!!priceListData.starts_at) {
                priceListData.starts_at = (0, utils_1.GetIsoStringFromDate)(priceListData.starts_at);
            }
            if (!!priceListData.ends_at) {
                priceListData.ends_at = (0, utils_1.GetIsoStringFromDate)(priceListData.ends_at);
            }
            return priceListData;
        });
    }
    normalizePriceSetConfig(config) {
        return {
            options: {
                populateWhere: { prices: { price_list_id: null } },
            },
            ...config,
        };
    }
}
exports.default = PricingModuleService;
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "retrievePriceSet", null);
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "listPriceSets", null);
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "listAndCountPriceSets", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "calculatePrices", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)()
    // @ts-expect-error
    ,
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "createPriceSets", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "upsertPriceSets", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)()
    // @ts-expect-error
    ,
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePriceSets", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePriceSets_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "addPrices", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)()
    // @ts-ignore
    ,
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "createPriceLists", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    (0, utils_1.EmitEvents)()
    // @ts-ignore
    ,
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePriceLists", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePriceListPrices", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "removePrices", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "addPriceListPrices", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "setPriceListRules", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "removePriceListRules", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "createPricePreferences", null);
__decorate([
    (0, utils_1.InjectManager)(),
    (0, utils_1.EmitEvents)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "upsertPricePreferences", null);
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePricePreferences", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "createPricePreferences_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePricePreferences_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "createPriceSets_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "addPrices_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "createPriceLists_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePriceLists_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "updatePriceListPrices_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "removePrices_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "addPriceListPrices_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "setPriceListRules_", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PricingModuleService.prototype, "removePriceListRules_", null);
const composeAllEvents = ({ eventBuilders, performedActions, sharedContext, }) => {
    for (const action of Object.keys(performedActions)) {
        for (const entity of Object.keys(performedActions[action])) {
            const eventName = action + (0, utils_1.upperCaseFirst)(entity);
            if (!eventBuilders[eventName]) {
                continue;
            }
            eventBuilders[eventName]({
                data: performedActions[action][entity] ?? [],
                sharedContext,
            });
        }
    }
};
const isTaxInclusive = (priceRules, preferences, currencyCode, regionId) => {
    const regionRule = priceRules?.find((rule) => rule.attribute === "region_id" && rule.value === regionId);
    const regionPreference = preferences.find((p) => p.attribute === "region_id" && p.value === regionId);
    const currencyPreference = preferences.find((p) => p.attribute === "currency_code" && p.value === currencyCode);
    if (regionRule && regionPreference) {
        return regionPreference.is_tax_inclusive;
    }
    if (currencyPreference) {
        return currencyPreference.is_tax_inclusive;
    }
    return false;
};
const hashPrice = (price) => {
    const data = Object.entries({
        currency_code: price.currency_code,
        price_set_id: "price_set_id" in price ? price.price_set_id ?? null : null,
        price_list_id: "price_list_id" in price ? price.price_list_id ?? null : null,
        min_quantity: price.min_quantity ? price.min_quantity.toString() : null,
        max_quantity: price.max_quantity ? price.max_quantity.toString() : null,
        ...("price_rules" in price
            ? price.price_rules?.reduce((agg, pr) => {
                agg[pr.attribute] = pr.value;
                return agg;
            }, {})
            : {}),
    }).sort(([a], [b]) => a.localeCompare(b));
    return (0, utils_1.simpleHash)(JSON.stringify(data));
};
