// @medusajs/link-modules@2.6.1/dist/definitions/order-payment-collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderPaymentCollection = void 0;
const utils_1 = require("@medusajs/framework/utils");
exports.OrderPaymentCollection = {
    serviceName: utils_1.LINKS.OrderPaymentCollection,
    isLink: true,
    databaseConfig: {
        tableName: "order_payment_collection",
        idPrefix: "ordpay",
    },
    alias: [
        {
            name: ["order_payment_collection", "order_payment_collections"],
            entity: "LinkOrderPaymentCollection",
        },
    ],
    primaryKeys: ["id", "order_id", "payment_collection_id"],
    relationships: [
        {
            serviceName: utils_1.Modules.ORDER,
            entity: "Order",
            primaryKey: "id",
            foreignKey: "order_id",
            alias: "order",
            args: {
                methodSuffix: "Orders",
            },
        },
        {
            serviceName: utils_1.Modules.PAYMENT,
            entity: "PaymentCollection",
            primaryKey: "id",
            foreignKey: "payment_collection_id",
            alias: "payment_collection",
            args: {
                methodSuffix: "PaymentCollections",
            },
            deleteCascade: true,
        },
    ],
    extends: [
        {
            serviceName: utils_1.Modules.ORDER,
            entity: "Order",
            fieldAlias: {
                payment_collections: {
                    path: "payment_collections_link.payment_collection",
                    isList: true,
                },
            },
            relationship: {
                serviceName: utils_1.LINKS.OrderPaymentCollection,
                primaryKey: "order_id",
                foreignKey: "id",
                alias: "payment_collections_link",
            },
        },
        {
            serviceName: utils_1.Modules.PAYMENT,
            entity: "PaymentCollection",
            fieldAlias: {
                order: "order_link.order",
            },
            relationship: {
                serviceName: utils_1.LINKS.OrderPaymentCollection,
                primaryKey: "payment_collection_id",
                foreignKey: "id",
                alias: "order_link",
            },
        },
    ],
};
