// @medusajs/link-modules@2.6.1/dist/definitions/order-fulfillment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderFulfillment = void 0;
const utils_1 = require("@medusajs/framework/utils");
exports.OrderFulfillment = {
    serviceName: utils_1.LINKS.OrderFulfillment,
    isLink: true,
    databaseConfig: {
        tableName: "order_fulfillment",
        idPrefix: "ordful",
    },
    alias: [
        {
            name: ["order_fulfillment", "order_fulfillments"],
            entity: "LinkOrderFulfillment",
        },
    ],
    primaryKeys: ["id", "order_id", "fulfillment_id"],
    relationships: [
        {
            serviceName: utils_1.Modules.ORDER,
            entity: "Order",
            primaryKey: "id",
            foreignKey: "order_id",
            alias: "order",
            args: {
                methodSuffix: "Orders",
            },
        },
        {
            serviceName: utils_1.Modules.FULFILLMENT,
            entity: "Fulfillment",
            primaryKey: "id",
            foreignKey: "fulfillment_id",
            alias: "fulfillments",
            args: {
                methodSuffix: "Fulfillments",
            },
        },
    ],
    extends: [
        {
            serviceName: utils_1.Modules.ORDER,
            entity: "Order",
            fieldAlias: {
                fulfillments: {
                    path: "fulfillment_link.fulfillments",
                    isList: true,
                },
            },
            relationship: {
                serviceName: utils_1.LINKS.OrderFulfillment,
                primaryKey: "order_id",
                foreignKey: "id",
                alias: "fulfillment_link",
                isList: true,
            },
        },
        {
            serviceName: utils_1.Modules.FULFILLMENT,
            entity: "Fulfillment",
            fieldAlias: {
                order: "order_link.order",
            },
            relationship: {
                serviceName: utils_1.LINKS.OrderFulfillment,
                primaryKey: "fulfillment_id",
                foreignKey: "id",
                alias: "order_link",
            },
        },
    ],
};
