// @medusajs/link-modules@2.6.1/dist/definitions/product-sales-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductSalesChannel = void 0;
const utils_1 = require("@medusajs/framework/utils");
exports.ProductSalesChannel = {
    serviceName: utils_1.LINKS.ProductSalesChannel,
    isLink: true,
    databaseConfig: {
        tableName: "product_sales_channel",
        idPrefix: "prodsc",
    },
    alias: [
        {
            name: "product_sales_channel",
        },
        {
            name: "product_sales_channels",
        },
    ],
    primaryKeys: ["id", "product_id", "sales_channel_id"],
    relationships: [
        {
            serviceName: utils_1.Modules.PRODUCT,
            entity: "Product",
            primaryKey: "id",
            foreignKey: "product_id",
            alias: "product",
            args: {
                methodSuffix: "Products",
            },
        },
        {
            serviceName: utils_1.Modules.SALES_CHANNEL,
            entity: "SalesChannel",
            primaryKey: "id",
            foreignKey: "sales_channel_id",
            alias: "sales_channel",
            args: {
                methodSuffix: "SalesChannels",
            },
        },
    ],
    extends: [
        {
            serviceName: utils_1.Modules.PRODUCT,
            entity: "Product",
            fieldAlias: {
                sales_channels: {
                    path: "sales_channels_link.sales_channel",
                    isList: true,
                },
            },
            relationship: {
                serviceName: utils_1.LINKS.ProductSalesChannel,
                primaryKey: "product_id",
                foreignKey: "id",
                alias: "sales_channels_link",
                isList: true,
            },
        },
        {
            serviceName: utils_1.Modules.SALES_CHANNEL,
            entity: "SalesChannel",
            relationship: {
                serviceName: utils_1.LINKS.ProductSalesChannel,
                primaryKey: "sales_channel_id",
                foreignKey: "id",
                alias: "products_link",
                isList: true,
            },
        },
    ],
};
