// @medusajs/link-modules@2.6.1/dist/definitions/product-shipping-profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductShippingProfile = void 0;
const utils_1 = require("@medusajs/framework/utils");
exports.ProductShippingProfile = {
    serviceName: utils_1.LINKS.ProductShippingProfile,
    isLink: true,
    databaseConfig: {
        tableName: "product_shipping_profile",
        idPrefix: "prodsp",
    },
    alias: [
        {
            name: "product_shipping_profile",
        },
        {
            name: "product_shipping_profiles",
        },
    ],
    primaryKeys: ["id", "product_id", "shipping_profile_id"],
    relationships: [
        {
            serviceName: utils_1.Modules.PRODUCT,
            entity: "Product",
            primaryKey: "id",
            foreignKey: "product_id",
            alias: "product",
            args: {
                methodSuffix: "Products",
            },
        },
        {
            serviceName: utils_1.Modules.FULFILLMENT,
            entity: "ShippingProfile",
            primaryKey: "id",
            foreignKey: "shipping_profile_id",
            alias: "shipping_profile",
            args: {
                methodSuffix: "ShippingProfiles",
            },
        },
    ],
    extends: [
        {
            serviceName: utils_1.Modules.PRODUCT,
            entity: "Product",
            fieldAlias: {
                shipping_profile: {
                    path: "shipping_profiles_link.shipping_profile",
                    isList: false,
                },
            },
            relationship: {
                serviceName: utils_1.LINKS.ProductShippingProfile,
                primaryKey: "product_id",
                foreignKey: "id",
                alias: "shipping_profiles_link",
                isList: false,
            },
        },
        {
            serviceName: utils_1.Modules.FULFILLMENT,
            entity: "ShippingProfile",
            relationship: {
                serviceName: utils_1.LINKS.ProductShippingProfile,
                primaryKey: "shipping_profile_id",
                foreignKey: "id",
                alias: "products_link",
                isList: true,
            },
        },
    ],
};
