// @medusajs/link-modules@2.6.1/dist/definitions/product-variant-inventory-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductVariantInventoryItem = void 0;
const utils_1 = require("@medusajs/framework/utils");
exports.ProductVariantInventoryItem = {
    serviceName: utils_1.LINKS.ProductVariantInventoryItem,
    isLink: true,
    databaseConfig: {
        tableName: "product_variant_inventory_item",
        idPrefix: "pvitem",
        extraFields: {
            required_quantity: {
                type: "integer",
                defaultValue: "1",
            },
        },
    },
    alias: [
        {
            name: [
                "product_variant_inventory_item",
                "product_variant_inventory_items",
            ],
            entity: "LinkProductVariantInventoryItem",
        },
    ],
    primaryKeys: ["id", "variant_id", "inventory_item_id"],
    relationships: [
        {
            serviceName: utils_1.Modules.PRODUCT,
            entity: "ProductVariant",
            primaryKey: "id",
            foreignKey: "variant_id",
            alias: "variant",
            args: {
                methodSuffix: "ProductVariants",
            },
        },
        {
            serviceName: utils_1.Modules.INVENTORY,
            entity: "InventoryItem",
            primaryKey: "id",
            foreignKey: "inventory_item_id",
            alias: "inventory",
            args: {
                methodSuffix: "InventoryItems",
            },
        },
    ],
    extends: [
        {
            serviceName: utils_1.Modules.PRODUCT,
            entity: "ProductVariant",
            fieldAlias: {
                inventory: "inventory_items.inventory",
            },
            relationship: {
                serviceName: utils_1.LINKS.ProductVariantInventoryItem,
                primaryKey: "variant_id",
                foreignKey: "id",
                alias: "inventory_items",
                isList: true,
            },
        },
        {
            serviceName: utils_1.Modules.INVENTORY,
            entity: "InventoryItem",
            fieldAlias: {
                variants: {
                    path: "variant_link.variant",
                    isList: true,
                },
            },
            relationship: {
                serviceName: utils_1.LINKS.ProductVariantInventoryItem,
                primaryKey: "inventory_item_id",
                foreignKey: "id",
                alias: "variant_link",
                isList: true,
            },
        },
    ],
};
