// @rokmohar/medusa-plugin-meilisearch@1.0.6/.medusa/server/src/modules/meilisearch/services/meilisearch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeiliSearchService = void 0;
const utils_1 = require("@medusajs/utils");
const meilisearch_1 = require("meilisearch");
const types_1 = require("../types");
const transformer_1 = require("../utils/transformer");
class MeiliSearchService extends utils_1.SearchUtils.AbstractSearchService {
    constructor(container, options) {
        super(container, options);
        this.isDefault = false;
        this.config_ = options;
        if (process.env.NODE_ENV !== 'development') {
            if (!options.config?.apiKey) {
                throw Error('Meilisearch API key is missing in plugin config. See https://github.com/rokmohar/medusa-plugin-meilisearch');
            }
        }
        if (!options.config?.host) {
            throw Error('Meilisearch host is missing in plugin config. See https://github.com/rokmohar/medusa-plugin-meilisearch');
        }
        this.client_ = new meilisearch_1.MeiliSearch(options.config);
    }
    async createIndex(indexName, options = { primaryKey: 'id' }) {
        return await this.client_.createIndex(indexName, options);
    }
    getIndex(indexName) {
        return this.client_.index(indexName);
    }
    async addDocuments(indexName, documents, type) {
        const transformedDocuments = this.getTransformedDocuments(type, documents);
        return await this.client_.index(indexName).addDocuments(transformedDocuments, { primaryKey: 'id' });
    }
    async replaceDocuments(indexName, documents, type) {
        const transformedDocuments = this.getTransformedDocuments(type, documents);
        return await this.client_.index(indexName).addDocuments(transformedDocuments, { primaryKey: 'id' });
    }
    async deleteDocument(indexName, documentId) {
        return await this.client_.index(indexName).deleteDocument(documentId);
    }
    async deleteDocuments(indexName, documentIds) {
        return await this.client_.index(indexName).deleteDocuments(documentIds);
    }
    async deleteAllDocuments(indexName) {
        return await this.client_.index(indexName).deleteAllDocuments();
    }
    async search(indexName, query, options) {
        const { paginationOptions, filter, additionalOptions } = options;
        return await this.client_.index(indexName).search(query, { filter, ...paginationOptions, ...additionalOptions });
    }
    async updateSettings(indexName, settings) {
        const indexSettings = settings.indexSettings ?? {};
        await this.upsertIndex(indexName, settings);
        return await this.client_.index(indexName).updateSettings(indexSettings);
    }
    async upsertIndex(indexName, settings) {
        try {
            await this.client_.getIndex(indexName);
        }
        catch (error) {
            if (error.code === types_1.meilisearchErrorCodes.INDEX_NOT_FOUND) {
                await this.createIndex(indexName, {
                    primaryKey: settings?.primaryKey ?? 'id',
                });
            }
        }
    }
    getTransformedDocuments(type, documents) {
        if (!documents?.length) {
            return [];
        }
        switch (type) {
            case utils_1.SearchUtils.indexTypes.PRODUCTS:
                const productsTransformer = this.config_.settings?.[utils_1.SearchUtils.indexTypes.PRODUCTS]?.transformer ?? transformer_1.transformProduct;
                return documents.map(productsTransformer);
            default:
                return documents;
        }
    }
}
exports.MeiliSearchService = MeiliSearchService;
MeiliSearchService.identifier = 'index-meilisearch';
