// @medusajs/order@2.6.1/dist/models/line-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderLineItem = void 0;
const utils_1 = require("@medusajs/framework/utils");
const line_item_adjustment_1 = require("./line-item-adjustment");
const line_item_tax_line_1 = require("./line-item-tax-line");
const _OrderLineItem = utils_1.model
    .define("OrderLineItem", {
    id: utils_1.model.id({ prefix: "ordli" }).primaryKey(),
    title: utils_1.model.text(),
    subtitle: utils_1.model.text().nullable(),
    thumbnail: utils_1.model.text().nullable(),
    variant_id: utils_1.model.text().nullable(),
    product_id: utils_1.model.text().nullable(),
    product_title: utils_1.model.text().nullable(),
    product_description: utils_1.model.text().nullable(),
    product_subtitle: utils_1.model.text().nullable(),
    product_type: utils_1.model.text().nullable(),
    product_type_id: utils_1.model.text().nullable(),
    product_collection: utils_1.model.text().nullable(),
    product_handle: utils_1.model.text().nullable(),
    variant_sku: utils_1.model.text().nullable(),
    variant_barcode: utils_1.model.text().nullable(),
    variant_title: utils_1.model.text().nullable(),
    variant_option_values: utils_1.model.json().nullable(),
    requires_shipping: utils_1.model.boolean().default(true),
    is_discountable: utils_1.model.boolean().default(true),
    is_tax_inclusive: utils_1.model.boolean().default(false),
    compare_at_unit_price: utils_1.model.bigNumber().nullable(),
    unit_price: utils_1.model.bigNumber().nullable(),
    is_custom_price: utils_1.model.boolean().default(false),
    metadata: utils_1.model.json().nullable(),
    tax_lines: utils_1.model.hasMany(() => line_item_tax_line_1.OrderLineItemTaxLine, {
        mappedBy: "item",
    }),
    adjustments: utils_1.model.hasMany(() => line_item_adjustment_1.OrderLineItemAdjustment, {
        mappedBy: "item",
    }),
})
    .cascades({
    delete: ["tax_lines", "adjustments"],
})
    .indexes([
    {
        name: "IDX_order_line_item_deleted_at",
        on: ["deleted_at"],
        unique: false,
        where: "deleted_at IS NOT NULL",
    },
    {
        name: "IDX_order_line_item_product_id",
        on: ["product_id"],
        unique: false,
        where: "deleted_at IS NOT NULL",
    },
    {
        name: "IDX_line_item_product_type_id",
        on: ["product_type_id"],
        unique: false,
        where: "deleted_at IS NOT NULL AND product_type_id IS NOT NULL",
    },
    {
        name: "IDX_order_line_item_variant_id",
        on: ["variant_id"],
        unique: false,
        where: "deleted_at IS NOT NULL",
    },
]);
exports.OrderLineItem = _OrderLineItem;
