// @medusajs/modules-sdk@2.6.1/dist/remote-query/remote-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteQuery = void 0;
const orchestration_1 = require("@medusajs/orchestration");
const utils_1 = require("@medusajs/utils");
const medusa_module_1 = require("../medusa-module");
const BASE_PREFIX = "";
const MAX_BATCH_SIZE = 4000;
const MAX_CONCURRENT_REQUESTS = 10;
class RemoteQuery {
    constructor({ modulesLoaded, customRemoteFetchData, servicesConfig = [], entitiesMap, }) {
        this.modulesMap = new Map();
        this.entitiesMap = new Map();
        const servicesConfig_ = [...servicesConfig];
        this.entitiesMap = entitiesMap;
        if (!modulesLoaded?.length) {
            modulesLoaded = medusa_module_1.MedusaModule.getLoadedModules().map((mod) => Object.values(mod)[0]);
        }
        for (const mod of modulesLoaded || []) {
            if (!mod.__definition.isQueryable) {
                continue;
            }
            const serviceName = mod.__definition.key;
            if (this.modulesMap.has(serviceName)) {
                throw new Error(`Duplicated instance of module ${serviceName} is not allowed.`);
            }
            this.modulesMap.set(serviceName, mod);
            servicesConfig_.push(mod.__joinerConfig);
        }
        this.customRemoteFetchData = customRemoteFetchData;
        this.remoteJoiner = new orchestration_1.RemoteJoiner(servicesConfig_, this.remoteFetchData.bind(this), {
            autoCreateServiceNameAlias: false,
            entitiesMap,
        });
    }
    getEntitiesMap() {
        return this.entitiesMap;
    }
    setFetchDataCallback(remoteFetchData) {
        this.remoteJoiner.setFetchDataCallback(remoteFetchData);
    }
    static getAllFieldsAndRelations(expand, prefix = BASE_PREFIX, args = {}) {
        expand = JSON.parse(JSON.stringify(expand));
        let fields = new Set();
        let relations = [];
        let shouldSelectAll = false;
        for (const field of expand.fields ?? []) {
            if (field === "*") {
                shouldSelectAll = true;
                break;
            }
            fields.add(prefix ? `${prefix}.${field}` : field);
        }
        const filters = expand.args?.find((arg) => arg.name === "filters")?.value ?? {};
        if ((0, utils_1.isPresent)(filters)) {
            args[prefix] = filters;
        }
        else if ((0, utils_1.isPresent)(expand.args)) {
            args[prefix] = expand.args;
        }
        for (const property in expand.expands ?? {}) {
            const newPrefix = prefix ? `${prefix}.${property}` : property;
            relations.push(newPrefix);
            fields.delete(newPrefix);
            const result = RemoteQuery.getAllFieldsAndRelations(expand.expands[property], newPrefix, args);
            result.select?.forEach(fields.add, fields);
            relations = relations.concat(result.relations);
        }
        const allFields = Array.from(fields);
        const select = allFields.length && !shouldSelectAll
            ? allFields
            : shouldSelectAll
                ? undefined
                : [];
        return {
            select,
            relations,
            args,
        };
    }
    hasPagination(options) {
        if (!options) {
            return false;
        }
        const attrs = ["skip", "cursor"];
        return Object.keys(options).some((key) => attrs.includes(key));
    }
    buildPagination(options, count) {
        return {
            skip: options.skip,
            take: options.take,
            // cursor: options.cursor, not yet supported
            // TODO: next cursor
            count,
        };
    }
    async fetchRemoteDataBatched(args) {
        const { serviceName, keyField, service, methodName, filters, options, ids, } = args;
        const getBatch = function* (idArray, batchSize) {
            for (let i = 0; i < idArray.length; i += batchSize) {
                yield idArray.slice(i, i + batchSize);
            }
        };
        const idsToFetch = getBatch(ids, MAX_BATCH_SIZE);
        const results = [];
        let running = 0;
        const fetchPromises = [];
        const processBatch = async (batch) => {
            running++;
            const batchFilters = { ...filters, [keyField]: batch };
            let result;
            try {
                if (RemoteQuery.traceFetchRemoteData) {
                    result = await RemoteQuery.traceFetchRemoteData(async () => service[methodName](batchFilters, options), serviceName, methodName, options);
                }
                else {
                    result = await service[methodName](batchFilters, options);
                }
                results.push(result);
            }
            finally {
                running--;
                processAllBatches();
            }
        };
        let batchesDone = () => { };
        const awaitBatches = new Promise((ok) => {
            batchesDone = ok;
        });
        const processAllBatches = async () => {
            let isDone = false;
            while (running < MAX_CONCURRENT_REQUESTS) {
                const nextBatch = idsToFetch.next();
                if (nextBatch.done) {
                    isDone = true;
                    break;
                }
                const batch = nextBatch.value;
                fetchPromises.push(processBatch(batch));
            }
            if (isDone) {
                await Promise.all(fetchPromises);
                batchesDone();
            }
        };
        processAllBatches();
        await awaitBatches;
        const flattenedResults = results.reduce((acc, result) => {
            if (Array.isArray(result) &&
                result.length === 2 &&
                Array.isArray(result[0])) {
                return acc.concat(result[0]);
            }
            return acc.concat(result);
        }, []);
        return flattenedResults;
    }
    async remoteFetchData(expand, keyField, ids, relationship) {
        if (this.customRemoteFetchData) {
            const resp = await this.customRemoteFetchData(expand, keyField, ids);
            if (resp !== undefined) {
                return resp;
            }
        }
        const serviceConfig = expand.serviceConfig;
        const service = this.modulesMap.get(serviceConfig.serviceName);
        let filters = {};
        const options = {
            ...RemoteQuery.getAllFieldsAndRelations(expand),
        };
        const availableOptions = [
            "skip",
            "take",
            "limit",
            "offset",
            "cursor",
            "sort",
            "order",
            "withDeleted",
        ];
        const availableOptionsAlias = new Map([
            ["limit", "take"],
            ["offset", "skip"],
        ]);
        for (const arg of expand.args || []) {
            if (arg.name === "filters" && arg.value) {
                filters = { ...filters, ...arg.value };
            }
            else if (arg.name === "context" && arg.value) {
                filters["context"] = arg.value;
            }
            else if (availableOptions.includes(arg.name)) {
                const argName = availableOptionsAlias.has(arg.name)
                    ? availableOptionsAlias.get(arg.name)
                    : arg.name;
                options[argName] = arg.value;
            }
        }
        if (ids) {
            filters[keyField] = ids;
        }
        delete options.args?.[BASE_PREFIX];
        if (Object.keys(options.args ?? {}).length) {
            filters = {
                ...filters,
                ...options?.args,
            };
            options.args = {};
        }
        const hasPagination = this.hasPagination(options);
        let methodName = hasPagination ? "listAndCount" : "list";
        if (relationship?.args?.methodSuffix) {
            methodName += (0, utils_1.toPascalCase)(relationship.args.methodSuffix);
        }
        else if (serviceConfig?.args?.methodSuffix) {
            methodName += (0, utils_1.toPascalCase)(serviceConfig.args.methodSuffix);
        }
        if (typeof service[methodName] !== "function") {
            throw new Error(`Method "${methodName}" does not exist on "${serviceConfig.serviceName}"`);
        }
        if (ids?.length && !hasPagination) {
            options.take = null;
        }
        if (ids && ids.length >= MAX_BATCH_SIZE && !hasPagination) {
            const data = await this.fetchRemoteDataBatched({
                serviceName: serviceConfig.serviceName,
                keyField,
                service,
                methodName,
                filters,
                options,
                ids,
            });
            return { data };
        }
        let result;
        if (RemoteQuery.traceFetchRemoteData) {
            result = await RemoteQuery.traceFetchRemoteData(async () => service[methodName](filters, options), serviceConfig.serviceName, methodName, options);
        }
        else {
            result = await service[methodName](filters, options);
        }
        if (hasPagination) {
            const [data, count] = result;
            return {
                data: {
                    rows: data,
                    metadata: this.buildPagination(options, count),
                },
                path: "rows",
            };
        }
        return {
            data: result,
        };
    }
    async query(query, variables, options) {
        let finalQuery = query;
        if ((0, utils_1.isString)(query)) {
            finalQuery = orchestration_1.RemoteJoiner.parseQuery(query, variables);
        }
        else if (!(0, utils_1.isString)(finalQuery?.service) && !(0, utils_1.isString)(finalQuery?.alias)) {
            finalQuery = (0, orchestration_1.toRemoteJoinerQuery)(query, variables);
        }
        return await this.remoteJoiner.query(finalQuery, options);
    }
}
exports.RemoteQuery = RemoteQuery;
