// @medusajs/modules-sdk@2.6.1/dist/remote-query/to-remote-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toRemoteQuery = toRemoteQuery;
const utils_1 = require("@medusajs/utils");
const parse_filters_1 = require("./parse-filters");
const FIELDS = "__fields";
const ARGUMENTS = "__args";
/**
 * convert a specific API configuration to a remote query object
 *
 * @param config
 * @param entitiesMap
 *
 * @example
 * const remoteQueryObject = toRemoteQuery({
 *   entity: "product",
 *   fields,
 *   filters: { variants: QueryFilter({ sku: "abc" }) },
 *   context: {
 *     variants: { calculated_price: QueryContext({ region_id: "reg_123" }) }
 *   }
 * });
 *
 * console.log(remoteQueryObject);
 */
function toRemoteQuery(config, entitiesMap) {
    const { entity, fields = [], filters = {}, context = {}, withDeleted, } = config;
    const joinerQuery = {
        [entity]: {
            __fields: [],
        },
    };
    function processNestedObjects(target, source, topLevel = true) {
        for (const key in source) {
            const src = topLevel ? source : source[key];
            if (!(0, utils_1.isObject)(src)) {
                target[key] = src;
                continue;
            }
            if (utils_1.QueryContext.isQueryContext(src)) {
                const normalizedFilters = { ...src };
                delete normalizedFilters.__type;
                const prop = "context";
                if (topLevel) {
                    target[ARGUMENTS] ??= {};
                    target[ARGUMENTS][prop] = normalizedFilters;
                    if (withDeleted) {
                        target[ARGUMENTS]["withDeleted"] = true;
                    }
                }
                else {
                    target[key] ??= {};
                    target[key][ARGUMENTS] ??= {};
                    target[key][ARGUMENTS][prop] = normalizedFilters;
                    if (withDeleted) {
                        target[key][ARGUMENTS]["withDeleted"] = true;
                    }
                }
            }
            else {
                if (!topLevel) {
                    target[key] ??= {};
                }
                const nextTarget = topLevel ? target : target[key];
                processNestedObjects(nextTarget, src, false);
            }
        }
    }
    // Process filters and context recursively
    processNestedObjects(joinerQuery[entity], context);
    for (const field of fields) {
        const fieldAsString = field;
        if (!fieldAsString.includes(".")) {
            joinerQuery[entity][FIELDS].push(field);
            continue;
        }
        const fieldSegments = fieldAsString.split(".");
        const fieldProperty = fieldSegments.pop();
        let combinedPath = "";
        const deepConfigRef = fieldSegments.reduce((acc, curr) => {
            combinedPath = combinedPath ? combinedPath + "." + curr : curr;
            acc[curr] ??= {};
            return acc[curr];
        }, joinerQuery[entity]);
        deepConfigRef[FIELDS] ??= [];
        deepConfigRef[FIELDS].push(fieldProperty);
    }
    if (config.pagination) {
        joinerQuery[entity][ARGUMENTS] ??= {};
        joinerQuery[entity][ARGUMENTS] = {
            ...joinerQuery[entity][ARGUMENTS],
            ...config.pagination,
        };
    }
    if (withDeleted) {
        joinerQuery[entity][ARGUMENTS] ??= {};
        joinerQuery[entity][ARGUMENTS]["withDeleted"] = true;
    }
    (0, parse_filters_1.parseAndAssignFilters)({
        entryPoint: entity,
        filters: filters,
        remoteQueryObject: joinerQuery,
    }, entitiesMap);
    return joinerQuery;
}
