// @medusajs/admin-vite-plugin@2.6.1/dist/index.mjs
// src/plugin.ts
import path2 from "path";

// src/custom-fields/generate-custom-field-displays.ts
import {
  isValidCustomFieldDisplayPath,
  isValidCustomFieldDisplayZone
} from "@medusajs/admin-shared";
import fs from "fs/promises";

// src/babel.ts
import { parse } from "@babel/parser";
import _traverse from "@babel/traverse";
import {
  isArrayExpression,
  isCallExpression,
  isFunctionDeclaration,
  isIdentifier,
  isJSXElement,
  isJSXFragment,
  isMemberExpression,
  isObjectExpression,
  isObjectProperty,
  isStringLiteral,
  isTemplateLiteral,
  isVariableDeclaration,
  isVariableDeclarator
} from "@babel/types";
var traverse;
if (typeof _traverse === "function") {
  traverse = _traverse;
} else {
  traverse = _traverse.default;
}

// src/logger.ts
import colors from "picocolors";
function getTimestamp() {
  const now = /* @__PURE__ */ new Date();
  return now.toLocaleTimeString("en-US", { hour12: true });
}
function getPrefix(type) {
  const timestamp = colors.dim(getTimestamp());
  const typeColor = type === "warn" ? colors.yellow : type === "info" ? colors.green : colors.red;
  const prefix = typeColor("[@medusajs/admin-vite-plugin]");
  return `${timestamp} ${prefix}`;
}
function getFile(options) {
  if (!options.file) {
    return "";
  }
  const value = Array.isArray(options.file) ? options.file.map((f) => f).join(", ") : options.file;
  return colors.dim(`${value}`);
}
function formatError(error) {
  if (error instanceof Error) {
    return colors.red(`${error.name}: ${error.message}
${error.stack}`);
  } else if (typeof error === "object") {
    return colors.red(JSON.stringify(error, null, 2));
  } else {
    return colors.red(String(error));
  }
}
var logger = {
  warn(msg, options = {}) {
    console.warn(`${getPrefix("warn")} ${msg} ${getFile(options)}`);
  },
  info(msg, options = {}) {
    console.info(`${getPrefix("info")} ${msg} ${getFile(options)}`);
  },
  error(msg, options = {}) {
    console.error(`${getPrefix("error")} ${msg} ${getFile(options)}`);
    if (options.error) {
      console.error(formatError(options.error));
    }
  }
};

// src/utils.ts
import { fdir } from "fdir";
import MagicString from "magic-string";
import crypto from "crypto";
import path from "path";
function normalizePath(file) {
  return path.normalize(file).replace(/\\/g, "/");
}
function getParserOptions(file) {
  const options = {
    sourceType: "module",
    plugins: ["jsx"]
  };
  if (file.endsWith(".tsx")) {
    options.plugins?.push("typescript");
  }
  return options;
}
function generateModule(code) {
  const magicString = new MagicString(code);
  return {
    code: magicString.toString(),
    map: magicString.generateMap({ hires: true })
  };
}
var VALID_FILE_EXTENSIONS = [".tsx", ".jsx", ".js"];
async function crawl(dir, file, depth) {
  const dirDepth = dir.split(path.sep).length;
  const crawler = new fdir().withBasePath().exclude((dirName) => dirName.startsWith("_")).filter((path3) => {
    return VALID_FILE_EXTENSIONS.some((ext) => path3.endsWith(ext));
  });
  if (file) {
    crawler.filter((path3) => {
      return VALID_FILE_EXTENSIONS.some((ext) => path3.endsWith(file + ext));
    });
  }
  if (depth) {
    crawler.filter((file2) => {
      const pathDepth = file2.split(path.sep).length - 1;
      if (depth.max && pathDepth > dirDepth + depth.max) {
        return false;
      }
      if (pathDepth < dirDepth + depth.min) {
        return false;
      }
      return true;
    });
  }
  return crawler.crawl(dir).withPromise();
}
function getConfigObjectProperties(path3) {
  if (isVariableDeclarator(path3.node)) {
    const configDeclaration = isIdentifier(path3.node.id, { name: "config" }) ? path3.node : null;
    if (configDeclaration && isCallExpression(configDeclaration.init) && configDeclaration.init.arguments.length > 0 && isObjectExpression(configDeclaration.init.arguments[0])) {
      return configDeclaration.init.arguments[0].properties;
    }
    return null;
  }
  const declaration = path3.node.declaration;
  if (isVariableDeclaration(declaration)) {
    const configDeclaration = declaration.declarations.find(
      (d) => isVariableDeclarator(d) && isIdentifier(d.id, { name: "config" })
    );
    if (configDeclaration && isCallExpression(configDeclaration.init) && configDeclaration.init.arguments.length > 0 && isObjectExpression(configDeclaration.init.arguments[0])) {
      return configDeclaration.init.arguments[0].properties;
    }
  }
  return null;
}
async function hasDefaultExport(ast) {
  let hasDefaultExport2 = false;
  traverse(ast, {
    ExportDefaultDeclaration() {
      hasDefaultExport2 = true;
    },
    AssignmentExpression(path3) {
      if (path3.node.left.type === "MemberExpression" && path3.node.left.object.type === "Identifier" && path3.node.left.object.name === "exports" && path3.node.left.property.type === "Identifier" && path3.node.left.property.name === "default") {
        hasDefaultExport2 = true;
      }
    },
    ExportNamedDeclaration(path3) {
      const specifiers = path3.node.specifiers;
      if (specifiers?.some(
        (s) => s.type === "ExportSpecifier" && s.exported.type === "Identifier" && s.exported.name === "default"
      )) {
        hasDefaultExport2 = true;
      }
    }
  });
  return hasDefaultExport2;
}
function generateHash(content) {
  return crypto.createHash("md5").update(content).digest("hex");
}
function isFileInAdminSubdirectory(file, subdirectory) {
  const normalizedPath = normalizePath(file);
  return normalizedPath.includes(`/src/admin/${subdirectory}/`);
}

// src/custom-fields/helpers.ts
import {
  isValidCustomFieldModel
} from "@medusajs/admin-shared";
function getModel(path3, file) {
  const configArgument = getConfigArgument(path3);
  if (!configArgument) {
    return null;
  }
  const modelProperty = configArgument.properties.find(
    (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "model" })
  );
  if (!modelProperty) {
    return null;
  }
  if (isTemplateLiteral(modelProperty.value)) {
    logger.warn(
      `'model' property cannot be a template literal (e.g. \`product\`).`,
      { file }
    );
    return null;
  }
  if (!isStringLiteral(modelProperty.value)) {
    logger.warn(
      `'model' is invalid. The 'model' property must be a string literal, e.g. 'product' or 'customer'.`,
      { file }
    );
    return null;
  }
  const model = modelProperty.value.value.trim();
  if (!isValidCustomFieldModel(model)) {
    logger.warn(
      `'model' is invalid, received: ${model}. The 'model' property must be set to a valid model, e.g. 'product' or 'customer'.`,
      { file }
    );
    return null;
  }
  return model;
}
function getConfigArgument(path3) {
  if (!isCallExpression(path3.node.declaration)) {
    return null;
  }
  if (!isIdentifier(path3.node.declaration.callee, {
    name: "unstable_defineCustomFieldsConfig"
  })) {
    return null;
  }
  const configArgument = path3.node.declaration.arguments[0];
  if (!isObjectExpression(configArgument)) {
    return null;
  }
  return configArgument;
}
function validateLink(path3, file) {
  const configArgument = getConfigArgument(path3);
  if (!configArgument) {
    return false;
  }
  const linkProperty = configArgument.properties.find(
    (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "link" })
  );
  if (!linkProperty) {
    logger.warn(`'link' property is missing.`, { file });
    return false;
  }
  return true;
}

// src/custom-fields/generate-custom-field-displays.ts
async function generateCustomFieldDisplays(sources) {
  const files = await getFilesFromSources(sources);
  const results = await getCustomFieldDisplayResults(files);
  const imports = results.map((result) => result.import).flat();
  const code = generateDisplayCode(results);
  return {
    imports,
    code
  };
}
async function getFilesFromSources(sources) {
  const files = (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/custom-fields`)
    )
  )).flat();
  return files;
}
function generateDisplayCode(results) {
  const groupedByModel = /* @__PURE__ */ new Map();
  results.forEach((result) => {
    const model = result.model;
    if (!groupedByModel.has(model)) {
      groupedByModel.set(model, []);
    }
    groupedByModel.get(model).push(result);
  });
  const segments = [];
  groupedByModel.forEach((results2, model) => {
    const displays = results2.map((result) => formatDisplays(result.displays)).filter((display) => display !== "").join(",\n");
    segments.push(`
      ${model}: [
        ${displays}
      ],
    `);
  });
  return `
    displays: {
      ${segments.join("\n")}
    }
  `;
}
function formatDisplays(displays) {
  if (!displays || displays.length === 0) {
    return "";
  }
  return displays.map(
    (display) => `
        {
          zone: "${display.zone}",
          Component: ${display.Component},
        }
      `
  ).join(",\n");
}
async function getCustomFieldDisplayResults(files) {
  return (await Promise.all(
    files.map(async (file, index) => parseDisplayFile(file, index))
  )).filter(Boolean);
}
async function parseDisplayFile(file, index) {
  const content = await fs.readFile(file, "utf8");
  let ast;
  try {
    ast = parse(content, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file`, { file, error: e });
    return null;
  }
  const import_ = generateImport(file, index);
  let displays = null;
  let model = null;
  let hasLink = false;
  try {
    traverse(ast, {
      ExportDefaultDeclaration(path3) {
        const _model = getModel(path3, file);
        if (!_model) {
          return;
        }
        model = _model;
        displays = getDisplays(path3, model, index, file);
        hasLink = validateLink(path3, file);
      }
    });
  } catch (err) {
    logger.error(`An error occurred while traversing the file.`, {
      file,
      error: err
    });
    return null;
  }
  if (!model) {
    logger.warn(`'model' property is missing.`, { file });
    return null;
  }
  if (!hasLink) {
    logger.warn(`'link' property is missing.`, { file });
    return null;
  }
  return {
    import: import_,
    model,
    displays
  };
}
function getDisplays(path3, model, index, file) {
  const configArgument = getConfigArgument(path3);
  if (!configArgument) {
    return null;
  }
  const displayProperty = configArgument.properties.find(
    (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "displays" })
  );
  if (!displayProperty) {
    return null;
  }
  if (!isArrayExpression(displayProperty.value)) {
    logger.warn(
      `'displays' is not an array. The 'displays' property must be an array of objects.`,
      { file }
    );
    return null;
  }
  const displays = [];
  displayProperty.value.elements.forEach((element, j) => {
    if (!isObjectExpression(element)) {
      return;
    }
    const zoneProperty = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "zone" })
    );
    if (!zoneProperty) {
      logger.warn(
        `'zone' property is missing at the ${j} index of the 'displays' property.`,
        { file }
      );
      return;
    }
    if (!isStringLiteral(zoneProperty.value)) {
      logger.warn(
        `'zone' property at index ${j} in the 'displays' property is not a string literal. 'zone' must be a string literal, e.g. 'general' or 'attributes'.`,
        { file }
      );
      return;
    }
    const zone = zoneProperty.value.value;
    const fullPath = getDisplayEntryPath(model, zone);
    if (!isValidCustomFieldDisplayZone(zone) || !isValidCustomFieldDisplayPath(fullPath)) {
      logger.warn(
        `'zone' is invalid at index ${j} in the 'displays' property. Received: ${zone}.`,
        { file }
      );
      return;
    }
    const componentProperty = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "component" })
    );
    if (!componentProperty) {
      logger.warn(
        `'component' property is missing at index ${j} in the 'displays' property.`,
        { file }
      );
      return;
    }
    displays.push({
      zone,
      Component: getDisplayComponent(index, j)
    });
  });
  return displays.length > 0 ? displays : null;
}
function getDisplayEntryPath(model, zone) {
  return `${model}.${zone}.$display`;
}
function getDisplayComponent(fileIndex, displayEntryIndex) {
  const import_ = generateCustomFieldConfigName(fileIndex);
  return `${import_}.displays[${displayEntryIndex}].component`;
}
function generateCustomFieldConfigName(index) {
  return `CustomFieldConfig${index}`;
}
function generateImport(file, index) {
  const path3 = normalizePath(file);
  return `import ${generateCustomFieldConfigName(index)} from "${path3}"`;
}

// src/custom-fields/generate-custom-field-forms.ts
import {
  isValidCustomFieldFormConfigPath,
  isValidCustomFieldFormFieldPath,
  isValidCustomFieldFormTab,
  isValidCustomFieldFormZone
} from "@medusajs/admin-shared";
import fs2 from "fs/promises";
import { outdent } from "outdent";
async function generateCustomFieldForms(sources) {
  const files = await getFilesFromSources2(sources);
  const results = await getCustomFieldResults(files);
  const imports = results.map((result) => result.import).flat();
  const code = generateCode(results);
  return {
    imports,
    code
  };
}
async function getFilesFromSources2(sources) {
  const files = (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/custom-fields`)
    )
  )).flat();
  return files;
}
function generateCode(results) {
  const groupedByModel = /* @__PURE__ */ new Map();
  results.forEach((result) => {
    const model = result.model;
    if (!groupedByModel.has(model)) {
      groupedByModel.set(model, []);
    }
    groupedByModel.get(model).push(result);
  });
  const segments = [];
  groupedByModel.forEach((results2, model) => {
    const configs = results2.map((result) => formatConfig(result.configs)).filter((config) => config !== "").join(",\n");
    const forms = results2.map((result) => formatForms(result.forms)).filter((form) => form !== "").join(",\n");
    segments.push(outdent`
      ${model}: {
        configs: [
          ${configs}
        ],
        forms: [
          ${forms}
        ],
      }
    `);
  });
  return outdent`
    customFields: {
      ${segments.join("\n")}
    }
  `;
}
function formatConfig(configs) {
  if (!configs || configs.length === 0) {
    return "";
  }
  return outdent`
    ${configs.map(
    (config) => outdent`
        {
          zone: "${config.zone}",
          fields: {
            ${config.fields.map(
      (field) => `${field.name}: {
              defaultValue: ${field.defaultValue},
              validation: ${field.validation},
            }`
    ).join(",\n")}
          },
        }
      `
  ).join(",\n")}
  `;
}
function formatForms(forms) {
  if (!forms || forms.length === 0) {
    return "";
  }
  return forms.map(
    (form) => outdent`
        {
          zone: "${form.zone}",
          tab: ${form.tab === void 0 ? void 0 : `"${form.tab}"`},
          fields: {
            ${form.fields.map(
      (field) => `${field.name}: {
              validation: ${field.validation},
              Component: ${field.Component},
              label: ${field.label},
              description: ${field.description},
              placeholder: ${field.placeholder},
            }`
    ).join(",\n")}
          },
        }
      `
  ).join(",\n");
}
async function getCustomFieldResults(files) {
  return (await Promise.all(files.map(async (file, index) => parseFile(file, index)))).filter(Boolean);
}
async function parseFile(file, index) {
  const content = await fs2.readFile(file, "utf8");
  let ast;
  try {
    ast = parse(content, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file`, { file, error: e });
    return null;
  }
  const import_ = generateImport2(file, index);
  let configs = [];
  let forms = [];
  let model = null;
  let hasLink = false;
  try {
    traverse(ast, {
      ExportDefaultDeclaration(path3) {
        const _model = getModel(path3, file);
        if (!_model) {
          return;
        }
        model = _model;
        hasLink = validateLink(path3, file);
        configs = getConfigs(path3, model, index, file);
        forms = getForms(path3, model, index, file);
      }
    });
  } catch (err) {
    logger.error(`An error occurred while traversing the file.`, {
      file,
      error: err
    });
    return null;
  }
  if (!model) {
    logger.warn(`'model' property is missing.`, { file });
    return null;
  }
  if (!hasLink) {
    logger.warn(`'link' property is missing.`, { file });
    return null;
  }
  return {
    import: import_,
    model,
    configs,
    forms
  };
}
function generateCustomFieldConfigName2(index) {
  return `CustomFieldConfig${index}`;
}
function generateImport2(file, index) {
  const path3 = normalizePath(file);
  return `import ${generateCustomFieldConfigName2(index)} from "${path3}"`;
}
function getForms(path3, model, index, file) {
  const formArray = getFormsArgument(path3, file);
  if (!formArray) {
    return null;
  }
  const forms = [];
  formArray.elements.forEach((element, j) => {
    if (!isObjectExpression(element)) {
      return;
    }
    const zoneProperty = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "zone" })
    );
    if (!zoneProperty) {
      logger.warn(
        `'zone' property is missing from the ${j} index of the 'forms' property. The 'zone' property is required to load a custom field form.`,
        { file }
      );
      return;
    }
    if (!isStringLiteral(zoneProperty.value)) {
      logger.warn(
        `'zone' property at the ${j} index of the 'forms' property is not a string literal. The 'zone' property must be a string literal, e.g. 'general' or 'attributes'.`,
        { file }
      );
      return;
    }
    const tabProperty = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "tab" })
    );
    let tab;
    if (tabProperty) {
      if (!isStringLiteral(tabProperty.value)) {
        logger.warn(
          `'tab' property at the ${j} index of the 'forms' property is not a string literal. The 'tab' property must be a string literal, e.g. 'general' or 'attributes'.`,
          { file }
        );
        return;
      }
      tab = tabProperty.value.value;
    }
    if (tab && !isValidCustomFieldFormTab(tab)) {
      logger.warn(
        `'tab' property at the ${j} index of the 'forms' property is not a valid custom field form tab for the '${model}' model. Received: ${tab}.`,
        { file }
      );
      return;
    }
    const zone = zoneProperty.value.value;
    const fullPath = getFormEntryFieldPath(model, zone, tab);
    if (!isValidCustomFieldFormZone(zone) || !isValidCustomFieldFormFieldPath(fullPath)) {
      logger.warn(
        `'zone' and 'tab' properties at the ${j} index of the 'forms' property are not a valid for the '${model}' model. Received: { zone: ${zone}, tab: ${tab} }.`,
        { file }
      );
      return;
    }
    const fieldsObject = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "fields" })
    );
    if (!fieldsObject) {
      logger.warn(
        `The 'fields' property is missing at the ${j} index of the 'forms' property. The 'fields' property is required to load a custom field form.`,
        { file }
      );
      return;
    }
    const fields = [];
    if (!isObjectExpression(fieldsObject.value)) {
      logger.warn(
        `The 'fields' property at the ${j} index of the 'forms' property is malformed. The 'fields' property must be an object.`,
        { file }
      );
      return;
    }
    fieldsObject.value.properties.forEach((field) => {
      if (!isObjectProperty(field) || !isIdentifier(field.key)) {
        return;
      }
      const name = field.key.name;
      if (!isObjectExpression(field.value) && !(isCallExpression(field.value) && isMemberExpression(field.value.callee) && isIdentifier(field.value.callee.object) && isIdentifier(field.value.callee.property) && field.value.callee.object.name === "form" && field.value.callee.property.name === "define" && field.value.arguments.length === 1 && isObjectExpression(field.value.arguments[0]))) {
        logger.warn(
          `'${name}' property in the 'fields' property at the ${j} index of the 'forms' property in ${file} is malformed. The property must be an object or a call to form.define().`,
          { file }
        );
        return;
      }
      const fieldObject = isObjectExpression(field.value) ? field.value : field.value.arguments[0];
      const labelProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "label" })
      );
      const descriptionProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "description" })
      );
      const componentProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "component" })
      );
      const validationProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "validation" })
      );
      const placeholderProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "placeholder" })
      );
      const label = getFormFieldSectionValue(
        !!labelProperty,
        index,
        j,
        name,
        "label"
      );
      const description = getFormFieldSectionValue(
        !!descriptionProperty,
        index,
        j,
        name,
        "description"
      );
      const placeholder = getFormFieldSectionValue(
        !!placeholderProperty,
        index,
        j,
        name,
        "placeholder"
      );
      const component = getFormFieldSectionValue(
        !!componentProperty,
        index,
        j,
        name,
        "component"
      );
      const validation = getFormFieldSectionValue(
        !!validationProperty,
        index,
        j,
        name,
        "validation"
      );
      fields.push({
        name,
        label,
        description,
        Component: component,
        validation,
        placeholder
      });
    });
    forms.push({
      zone,
      tab,
      fields
    });
  });
  return forms.length > 0 ? forms : null;
}
function getFormFieldSectionValue(exists, fileIndex, formIndex, fieldKey, value) {
  if (!exists) {
    return "undefined";
  }
  const import_ = generateCustomFieldConfigName2(fileIndex);
  return `${import_}.forms[${formIndex}].fields.${fieldKey}.${value}`;
}
function getFormEntryFieldPath(model, zone, tab) {
  return `${model}.${zone}.${tab ? `${tab}.` : ""}$field`;
}
function getConfigs(path3, model, index, file) {
  const formArray = getFormsArgument(path3, file);
  if (!formArray) {
    logger.warn(`'forms' property is missing.`, { file });
    return null;
  }
  const configs = [];
  formArray.elements.forEach((element, j) => {
    if (!isObjectExpression(element)) {
      logger.warn(
        `'forms' property at the ${j} index is malformed. The 'forms' property must be an object.`,
        { file }
      );
      return;
    }
    const zoneProperty = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "zone" })
    );
    if (!zoneProperty) {
      logger.warn(
        `'zone' property is missing from the ${j} index of the 'forms' property.`,
        { file }
      );
      return;
    }
    if (isTemplateLiteral(zoneProperty.value)) {
      logger.warn(
        `'zone' property at the ${j} index of the 'forms' property cannot be a template literal (e.g. \`general\`).`,
        { file }
      );
      return;
    }
    if (!isStringLiteral(zoneProperty.value)) {
      logger.warn(
        `'zone' property at the ${j} index of the 'forms' property is not a string literal (e.g. 'general' or 'attributes').`,
        { file }
      );
      return;
    }
    const zone = zoneProperty.value.value;
    const fullPath = getFormEntryConfigPath(model, zone);
    if (!isValidCustomFieldFormZone(zone) || !isValidCustomFieldFormConfigPath(fullPath)) {
      logger.warn(
        `'zone' property at the ${j} index of the 'forms' property is not a valid custom field form zone for the '${model}' model. Received: ${zone}.`
      );
      return;
    }
    const fieldsObject = element.properties.find(
      (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "fields" })
    );
    if (!fieldsObject) {
      logger.warn(
        `'fields' property is missing from the ${j} entry in the 'forms' property in ${file}.`,
        { file }
      );
      return;
    }
    const fields = [];
    if (!isObjectExpression(fieldsObject.value)) {
      logger.warn(
        `'fields' property at the ${j} index of the 'forms' property is malformed. The 'fields' property must be an object.`,
        { file }
      );
      return;
    }
    fieldsObject.value.properties.forEach((field) => {
      if (!isObjectProperty(field) || !isIdentifier(field.key)) {
        return;
      }
      const name = field.key.name;
      if (!isObjectExpression(field.value) && !(isCallExpression(field.value) && isMemberExpression(field.value.callee) && isIdentifier(field.value.callee.object) && isIdentifier(field.value.callee.property) && field.value.callee.object.name === "form" && field.value.callee.property.name === "define" && field.value.arguments.length === 1 && isObjectExpression(field.value.arguments[0]))) {
        logger.warn(
          `'${name}' property in the 'fields' property at the ${j} index of the 'forms' property in ${file} is malformed. The property must be an object or a call to form.define().`,
          { file }
        );
        return;
      }
      const fieldObject = isObjectExpression(field.value) ? field.value : field.value.arguments[0];
      const defaultValueProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "defaultValue" })
      );
      if (!defaultValueProperty) {
        logger.warn(
          `'defaultValue' property is missing at the ${j} index of the 'forms' property in ${file}.`,
          { file }
        );
        return;
      }
      const validationProperty = fieldObject.properties.find(
        (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "validation" })
      );
      if (!validationProperty) {
        logger.warn(
          `'validation' property is missing at the ${j} index of the 'forms' property in ${file}.`,
          { file }
        );
        return;
      }
      const defaultValue = getFormFieldValue(index, j, name, "defaultValue");
      const validation = getFormFieldValue(index, j, name, "validation");
      fields.push({
        name,
        defaultValue,
        validation
      });
    });
    configs.push({
      zone,
      fields
    });
  });
  return configs.length > 0 ? configs : null;
}
function getFormFieldValue(fileIndex, formIndex, fieldKey, value) {
  const import_ = generateCustomFieldConfigName2(fileIndex);
  return `${import_}.forms[${formIndex}].fields.${fieldKey}.${value}`;
}
function getFormEntryConfigPath(model, zone) {
  return `${model}.${zone}.$config`;
}
function getFormsArgument(path3, file) {
  const configArgument = getConfigArgument(path3);
  if (!configArgument) {
    return null;
  }
  const formProperty = configArgument.properties.find(
    (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "forms" })
  );
  if (!formProperty) {
    return null;
  }
  if (!isArrayExpression(formProperty.value)) {
    logger.warn(
      `The 'forms' property is malformed. The 'forms' property must be an array of objects.`,
      { file }
    );
    return null;
  }
  return formProperty.value;
}

// src/custom-fields/generate-custom-field-hashes.ts
import fs3 from "fs/promises";
async function generateCustomFieldHashes(sources) {
  const files = await getFilesFromSources3(sources);
  const contents = await Promise.all(files.map(getCustomFieldContents));
  const linkContents = contents.map((c) => c.link).filter(Boolean);
  const formContents = contents.map((c) => c.form).filter(Boolean);
  const displayContents = contents.map((c) => c.display).filter(Boolean);
  const totalLinkContent = linkContents.join("");
  const totalFormContent = formContents.join("");
  const totalDisplayContent = displayContents.join("");
  return {
    linkHash: generateHash(totalLinkContent),
    formHash: generateHash(totalFormContent),
    displayHash: generateHash(totalDisplayContent)
  };
}
async function getFilesFromSources3(sources) {
  return (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/custom-fields`)
    )
  )).flat();
}
async function getCustomFieldContents(file) {
  const code = await fs3.readFile(file, "utf-8");
  let ast = null;
  try {
    ast = parse(code, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file.`, {
      file,
      error: e
    });
    return { link: null, form: null, display: null };
  }
  let linkContent = null;
  let formContent = null;
  let displayContent = null;
  try {
    traverse(ast, {
      ExportDefaultDeclaration(path3) {
        const configArgument = getConfigArgument(path3);
        if (!configArgument) {
          return;
        }
        configArgument.properties.forEach((prop) => {
          if (!isObjectProperty(prop) || !prop.key || !isIdentifier(prop.key)) {
            return;
          }
          switch (prop.key.name) {
            case "link":
              linkContent = code.slice(prop.start, prop.end);
              break;
            case "forms":
              formContent = code.slice(prop.start, prop.end);
              break;
            case "display":
              displayContent = code.slice(prop.start, prop.end);
              break;
          }
        });
      }
    });
  } catch (e) {
    logger.error(
      `An error occurred while processing ${file}. See the below error for more details:
${e}`,
      { file, error: e }
    );
    return { link: null, form: null, display: null };
  }
  return { link: linkContent, form: formContent, display: displayContent };
}

// src/custom-fields/generate-custom-field-links.ts
import fs4 from "fs/promises";
async function generateCustomFieldLinks(sources) {
  const files = await getFilesFromSources4(sources);
  const results = await getCustomFieldLinkResults(files);
  const imports = results.map((result) => result.import);
  const code = generateCode2(results);
  return {
    imports,
    code
  };
}
async function getFilesFromSources4(sources) {
  const files = (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/custom-fields`)
    )
  )).flat();
  return files;
}
function generateCode2(results) {
  const groupedByModel = /* @__PURE__ */ new Map();
  results.forEach((result) => {
    const model = result.model;
    if (!groupedByModel.has(model)) {
      groupedByModel.set(model, []);
    }
    groupedByModel.get(model).push(result);
  });
  const segments = [];
  groupedByModel.forEach((results2, model) => {
    const links = results2.map((result) => result.link).join(",\n");
    segments.push(`
      ${model}: [
        ${links}
      ],
    `);
  });
  return `
    links: {
      ${segments.join("\n")}
    }
  `;
}
async function getCustomFieldLinkResults(files) {
  return (await Promise.all(files.map(async (file, index) => parseFile2(file, index)))).filter(Boolean);
}
async function parseFile2(file, index) {
  const content = await fs4.readFile(file, "utf8");
  let ast;
  try {
    ast = parse(content, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file`, { file, error: e });
    return null;
  }
  const import_ = generateImport3(file, index);
  let link = null;
  let model = null;
  try {
    traverse(ast, {
      ExportDefaultDeclaration(path3) {
        const _model = getModel(path3, file);
        if (!_model) {
          return;
        }
        model = _model;
        link = getLink(path3, index, file);
      }
    });
  } catch (err) {
    logger.error(`An error occurred while traversing the file.`, {
      file,
      error: err
    });
    return null;
  }
  if (!link || !model) {
    return null;
  }
  return {
    import: import_,
    model,
    link
  };
}
function generateCustomFieldConfigName3(index) {
  return `CustomFieldConfig${index}`;
}
function generateImport3(file, index) {
  const path3 = normalizePath(file);
  return `import ${generateCustomFieldConfigName3(index)} from "${path3}"`;
}
function getLink(path3, index, file) {
  const configArgument = getConfigArgument(path3);
  if (!configArgument) {
    return null;
  }
  const linkProperty = configArgument.properties.find(
    (p) => isObjectProperty(p) && isIdentifier(p.key, { name: "link" })
  );
  if (!linkProperty) {
    logger.warn(`'link' is missing.`, { file });
    return null;
  }
  const import_ = generateCustomFieldConfigName3(index);
  return `${import_}.link`;
}

// src/routes/generate-menu-items.ts
import {
  NESTED_ROUTE_POSITIONS
} from "@medusajs/admin-shared";
import fs5 from "fs/promises";
import { outdent as outdent2 } from "outdent";

// src/routes/helpers.ts
function getRoute(file) {
  const importPath = normalizePath(file);
  return importPath.replace(/.*\/admin\/(routes)/, "").replace(/\[([^\]]+)\]/g, ":$1").replace(
    new RegExp(
      `/page\\.(${VALID_FILE_EXTENSIONS.map((ext) => ext.slice(1)).join(
        "|"
      )})$`
    ),
    ""
  );
}

// src/routes/generate-menu-items.ts
async function generateMenuItems(sources) {
  const files = await getFilesFromSources5(sources);
  const results = await getMenuItemResults(files);
  const imports = results.map((result) => result.import);
  const code = generateCode3(results);
  return { imports, code };
}
function generateCode3(results) {
  return outdent2`
        menuItems: [
            ${results.map((result) => formatMenuItem(result.menuItem)).join(",\n")}
        ]
    }
  `;
}
function formatMenuItem(route) {
  const { label, icon, path: path3, nested } = route;
  return `{
    label: ${label},
    icon: ${icon || "undefined"},
    path: "${path3}",
    nested: ${nested ? `"${nested}"` : "undefined"}
  }`;
}
async function getFilesFromSources5(sources) {
  const files = (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/routes`, "page", { min: 1 })
    )
  )).flat();
  return files;
}
async function getMenuItemResults(files) {
  const results = await Promise.all(files.map(parseFile3));
  return results.filter((item) => item !== null);
}
async function parseFile3(file, index) {
  const config = await getRouteConfig(file);
  if (!config) {
    return null;
  }
  if (!config.label) {
    logger.warn(`Config is missing a label.`, {
      file
    });
  }
  const import_ = generateImport4(file, index);
  const menuItem = generateMenuItem(config, file, index);
  return {
    import: import_,
    menuItem
  };
}
function generateImport4(file, index) {
  const path3 = normalizePath(file);
  return `import { config as ${generateRouteConfigName(index)} } from "${path3}"`;
}
function generateMenuItem(config, file, index) {
  const configName = generateRouteConfigName(index);
  return {
    label: `${configName}.label`,
    icon: config.icon ? `${configName}.icon` : void 0,
    path: getRoute(file),
    nested: config.nested
  };
}
async function getRouteConfig(file) {
  const code = await fs5.readFile(file, "utf-8");
  let ast = null;
  try {
    ast = parse(code, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file.`, {
      file,
      error: e
    });
    return null;
  }
  let config = null;
  let configFound = false;
  try {
    traverse(ast, {
      /**
       * For bundled files, the config will not be a named export,
       * but instead a variable declaration.
       */
      VariableDeclarator(path3) {
        if (configFound) {
          return;
        }
        const properties = getConfigObjectProperties(path3);
        if (!properties) {
          return;
        }
        config = processConfigProperties(properties, file);
        if (config) {
          configFound = true;
        }
      },
      /**
       * For unbundled files, the `config` will always be a named export.
       */
      ExportNamedDeclaration(path3) {
        if (configFound) {
          return;
        }
        const properties = getConfigObjectProperties(path3);
        if (!properties) {
          return;
        }
        config = processConfigProperties(properties, file);
        if (config) {
          configFound = true;
        }
      }
    });
  } catch (e) {
    logger.error(`An error occurred while traversing the file.`, {
      file,
      error: e
    });
  }
  return config;
}
function processConfigProperties(properties, file) {
  const hasProperty = (name) => properties.some(
    (prop) => isObjectProperty(prop) && isIdentifier(prop.key, { name })
  );
  const hasLabel = hasProperty("label");
  if (!hasLabel) {
    return null;
  }
  const nested = properties.find(
    (prop) => isObjectProperty(prop) && isIdentifier(prop.key, { name: "nested" })
  );
  let nestedValue = void 0;
  if (isStringLiteral(nested?.value)) {
    nestedValue = nested.value.value;
  }
  if (nestedValue && !NESTED_ROUTE_POSITIONS.includes(nestedValue)) {
    logger.error(
      `Invalid nested route position: "${nestedValue}". Allowed values are: ${NESTED_ROUTE_POSITIONS.join(
        ", "
      )}`,
      { file }
    );
    return null;
  }
  return {
    label: hasLabel,
    icon: hasProperty("icon"),
    nested: nestedValue
  };
}
function generateRouteConfigName(index) {
  return `RouteConfig${index}`;
}

// src/routes/generate-route-hashes.ts
import fs6 from "fs/promises";
async function generateRouteHashes(sources) {
  const files = await getFilesFromSources6(sources);
  const contents = await Promise.all(files.map(getRouteContents));
  const defaultExportContents = contents.map((c) => c.defaultExport).filter(Boolean);
  const configContents = contents.map((c) => c.config).filter(Boolean);
  const totalDefaultExportContent = defaultExportContents.join("");
  const totalConfigContent = configContents.join("");
  return {
    defaultExportHash: generateHash(totalDefaultExportContent),
    configHash: generateHash(totalConfigContent)
  };
}
async function getFilesFromSources6(sources) {
  return (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/routes`, "page", { min: 1 })
    )
  )).flat();
}
async function getRouteContents(file) {
  const code = await fs6.readFile(file, "utf-8");
  let ast = null;
  try {
    ast = parse(code, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file.`, {
      file,
      error: e
    });
    return { defaultExport: null, config: null };
  }
  let defaultExportContent = null;
  let configContent = null;
  try {
    traverse(ast, {
      ExportDefaultDeclaration(path3) {
        defaultExportContent = code.slice(path3.node.start, path3.node.end);
      },
      ExportNamedDeclaration(path3) {
        const properties = getConfigObjectProperties(path3);
        if (properties) {
          configContent = code.slice(path3.node.start, path3.node.end);
        }
      }
    });
  } catch (e) {
    logger.error(
      `An error occurred while processing ${file}. See the below error for more details:
${e}`,
      { file, error: e }
    );
    return { defaultExport: null, config: null };
  }
  return { defaultExport: defaultExportContent, config: configContent };
}

// src/routes/generate-routes.ts
import fs7 from "fs/promises";
import { outdent as outdent3 } from "outdent";
async function generateRoutes(sources) {
  const files = await getFilesFromSources7(sources);
  const results = await getRouteResults(files);
  const imports = results.map((result) => result.imports).flat();
  const code = generateCode4(results);
  return {
    imports,
    code
  };
}
function generateCode4(results) {
  return outdent3`
        routes: [
            ${results.map((result) => formatRoute(result.route)).join(",\n")}
        ]
    }
  `;
}
function formatRoute(route) {
  let base = `{
    Component: ${route.Component},
    path: "${route.path}"`;
  if (route.handle) {
    base += `,
    handle: ${route.handle}`;
  }
  if (route.loader) {
    base += `,
    loader: ${route.loader}`;
  }
  if (route.children?.length) {
    return `${base},
    children: [
      ${route.children.map((child) => formatRoute(child)).join(",\n      ")}
    ]
  }`;
  }
  return `${base}
  }`;
}
async function getFilesFromSources7(sources) {
  const files = (await Promise.all(
    Array.from(sources).map(
      async (source) => crawl(`${source}/routes`, "page", { min: 1 })
    )
  )).flat();
  return files;
}
async function getRouteResults(files) {
  const results = (await Promise.all(files.map(parseFile4))).filter(
    (result) => result !== null
  );
  const routeMap = /* @__PURE__ */ new Map();
  results.forEach((result) => {
    const routePath = result.route.path;
    const isParallel = routePath.includes("/@");
    if (isParallel) {
      const parentPath = routePath.split("/@")[0];
      const parent = routeMap.get(parentPath);
      if (parent) {
        parent.route.children = parent.route.children || [];
        const finalRoute = {
          ...result.route,
          path: result.route.path.replace("@", "")
        };
        parent.route.children.push(finalRoute);
        parent.imports.push(...result.imports);
      }
    } else {
      routeMap.set(routePath, result);
    }
  });
  return Array.from(routeMap.values());
}
async function parseFile4(file, index) {
  const code = await fs7.readFile(file, "utf-8");
  let ast = null;
  try {
    ast = parse(code, getParserOptions(file));
  } catch (e) {
    logger.error("An error occurred while parsing the file.", {
      file,
      error: e
    });
    return null;
  }
  if (!await isValidRouteFile(ast, file)) {
    return null;
  }
  const { hasHandle, hasLoader } = await hasNamedExports(ast, file);
  const routePath = getRoute(file);
  const imports = generateImports(file, index, hasHandle, hasLoader);
  const route = generateRoute(routePath, index, hasHandle, hasLoader);
  return {
    imports,
    route
  };
}
async function isValidRouteFile(ast, file) {
  try {
    return await hasDefaultExport(ast);
  } catch (e) {
    logger.error(
      `An error occurred while checking for a default export in ${file}. The file will be ignored. See the below error for more details:
${e}`
    );
    return false;
  }
}
function generateImports(file, index, hasHandle, hasLoader) {
  const imports = [];
  const route = generateRouteComponentName(index);
  const importPath = normalizePath(file);
  if (!hasHandle && !hasLoader) {
    imports.push(`import ${route} from "${importPath}"`);
  } else {
    const namedImports = [
      hasHandle && `handle as ${generateHandleName(index)}`,
      hasLoader && `loader as ${generateLoaderName(index)}`
    ].filter(Boolean).join(", ");
    imports.push(`import ${route}, { ${namedImports} } from "${importPath}"`);
  }
  return imports;
}
function generateRoute(route, index, hasHandle, hasLoader) {
  return {
    Component: generateRouteComponentName(index),
    path: route,
    handle: hasHandle ? generateHandleName(index) : void 0,
    loader: hasLoader ? generateLoaderName(index) : void 0
  };
}
function generateRouteComponentName(index) {
  return `RouteComponent${index}`;
}
function generateHandleName(index) {
  return `handle${index}`;
}
function generateLoaderName(index) {
  return `loader${index}`;
}
async function hasNamedExports(ast, file) {
  let hasHandle = false;
  let hasLoader = false;
  try {
    traverse(ast, {
      ExportNamedDeclaration(path3) {
        const declaration = path3.node.declaration;
        if (declaration?.type === "VariableDeclaration") {
          declaration.declarations.forEach((decl) => {
            if (decl.id.type === "Identifier" && decl.id.name === "handle") {
              hasHandle = true;
            }
            if (decl.id.type === "Identifier" && decl.id.name === "loader") {
              hasLoader = true;
            }
          });
        }
        if (declaration?.type === "FunctionDeclaration" && declaration.id?.name === "loader") {
          hasLoader = true;
        }
      }
    });
  } catch (e) {
    logger.error("An error occurred while checking for named exports.", {
      file,
      error: e
    });
  }
  return { hasHandle, hasLoader };
}

// src/virtual-modules/generate-virtual-display-module.ts
import { outdent as outdent4 } from "outdent";
async function generateVirtualDisplayModule(sources) {
  const displays = await generateCustomFieldDisplays(sources);
  const code = outdent4`
    ${displays.imports.join("\n")}

    export default {
      ${displays.code}
    }
  `;
  return generateModule(code);
}

// src/virtual-modules/generate-virtual-form-module.ts
import outdent6 from "outdent";

// src/widgets/generate-widget-hash.ts
import fs8 from "fs/promises";

// src/widgets/helpers.ts
async function getWidgetFilesFromSources(sources) {
  return (await Promise.all(
    Array.from(sources).map(async (source) => crawl(`${source}/widgets`))
  )).flat();
}

// src/widgets/generate-widget-hash.ts
async function generateWidgetHash(sources) {
  const files = await getWidgetFilesFromSources(sources);
  const contents = await Promise.all(files.map(getWidgetContents));
  const totalContent = contents.flatMap(({ config, defaultExport }) => [config, defaultExport]).filter(Boolean).join("");
  return generateHash(totalContent);
}
async function getWidgetContents(file) {
  const code = await fs8.readFile(file, "utf-8");
  let ast;
  try {
    ast = parse(code, getParserOptions(file));
  } catch (e) {
    logger.error(
      `An error occurred while parsing the file. Due to the error we cannot validate whether the widget has changed. If your changes aren't correctly reflected try restarting the dev server.`,
      {
        file,
        error: e
      }
    );
    return { config: null, defaultExport: null };
  }
  let configContent = null;
  let defaultExportContent = null;
  traverse(ast, {
    ExportNamedDeclaration(path3) {
      const properties = getConfigObjectProperties(path3);
      if (properties) {
        configContent = code.slice(path3.node.start, path3.node.end);
      }
    },
    ExportDefaultDeclaration(path3) {
      defaultExportContent = code.slice(path3.node.start, path3.node.end);
    }
  });
  return { config: configContent, defaultExport: defaultExportContent };
}

// src/widgets/generate-widgets.ts
import { isValidInjectionZone } from "@medusajs/admin-shared";
import fs9 from "fs/promises";
import outdent5 from "outdent";
async function generateWidgets(sources) {
  const files = await getWidgetFilesFromSources(sources);
  const results = await getWidgetResults(files);
  const imports = results.map((r) => r.import);
  const code = generateCode5(results);
  return {
    imports,
    code
  };
}
async function getWidgetResults(files) {
  return (await Promise.all(files.map(parseFile5))).filter(
    (r) => r !== null
  );
}
function generateCode5(results) {
  return outdent5`
    widgets: [
      ${results.map((r) => formatWidget(r.widget)).join(",\n")}
    ]
  `;
}
function formatWidget(widget) {
  return outdent5`
    {
        Component: ${widget.Component},
        zone: [${widget.zone.map((z) => `"${z}"`).join(", ")}]
    }
  `;
}
async function parseFile5(file, index) {
  const code = await fs9.readFile(file, "utf-8");
  let ast;
  try {
    ast = parse(code, getParserOptions(file));
  } catch (e) {
    logger.error(`An error occurred while parsing the file.`, {
      file,
      error: e
    });
    return null;
  }
  let fileHasDefaultExport = false;
  try {
    fileHasDefaultExport = await hasDefaultExport(ast);
  } catch (e) {
    logger.error(`An error occurred while checking for a default export.`, {
      file,
      error: e
    });
    return null;
  }
  if (!fileHasDefaultExport) {
    return null;
  }
  let zone;
  try {
    zone = await getWidgetZone(ast, file);
  } catch (e) {
    logger.error(`An error occurred while traversing the file.`, {
      file,
      error: e
    });
    return null;
  }
  if (!zone) {
    logger.warn(`'zone' property is missing from the widget config.`, { file });
    return null;
  }
  const import_ = generateImport5(file, index);
  const widget = generateWidget(zone, index);
  return {
    widget,
    import: import_
  };
}
function generateWidgetComponentName(index) {
  return `WidgetComponent${index}`;
}
function generateWidgetConfigName(index) {
  return `WidgetConfig${index}`;
}
function generateImport5(file, index) {
  const path3 = normalizePath(file);
  return `import ${generateWidgetComponentName(
    index
  )}, { config as ${generateWidgetConfigName(index)} } from "${path3}"`;
}
function generateWidget(zone, index) {
  return {
    Component: generateWidgetComponentName(index),
    zone
  };
}
async function getWidgetZone(ast, file) {
  const zones = [];
  let zoneFound = false;
  traverse(ast, {
    /**
     * In case we are processing a bundled file, the `config` will most likely
     * not be a named export. Instead we look for a `VariableDeclaration` named
     * `config` and extract the `zone` property from it.
     */
    VariableDeclarator(path3) {
      if (zoneFound) {
        return;
      }
      if (path3.node.id.type === "Identifier" && path3.node.id.name === "config" && path3.node.init?.type === "CallExpression") {
        const arg = path3.node.init.arguments[0];
        if (arg?.type === "ObjectExpression") {
          const zoneProperty = arg.properties.find(
            (p) => p.type === "ObjectProperty" && p.key.name === "zone"
          );
          if (zoneProperty?.type === "ObjectProperty") {
            extractZoneValues(zoneProperty.value, zones, file);
            zoneFound = true;
          }
        }
      }
    },
    /**
     * For unbundled files, the `config` will always be a named export.
     */
    ExportNamedDeclaration(path3) {
      if (zoneFound) {
        return;
      }
      const declaration = path3.node.declaration;
      if (declaration?.type === "VariableDeclaration" && declaration.declarations[0]?.type === "VariableDeclarator" && declaration.declarations[0].id.type === "Identifier" && declaration.declarations[0].id.name === "config" && declaration.declarations[0].init?.type === "CallExpression") {
        const arg = declaration.declarations[0].init.arguments[0];
        if (arg?.type === "ObjectExpression") {
          const zoneProperty = arg.properties.find(
            (p) => p.type === "ObjectProperty" && p.key.name === "zone"
          );
          if (zoneProperty?.type === "ObjectProperty") {
            extractZoneValues(zoneProperty.value, zones, file);
            zoneFound = true;
          }
        }
      }
    }
  });
  if (!zoneFound) {
    logger.warn(`'zone' property is missing from the widget config.`, { file });
    return null;
  }
  const validatedZones = zones.filter(isValidInjectionZone);
  if (validatedZones.length === 0) {
    logger.warn(`'zone' property is not a valid injection zone.`, {
      file
    });
    return null;
  }
  return validatedZones;
}
function extractZoneValues(value, zones, file) {
  if (isTemplateLiteral(value)) {
    logger.warn(
      `'zone' property cannot be a template literal (e.g. \`product.details.after\`).`,
      { file }
    );
    return;
  }
  if (isStringLiteral(value)) {
    zones.push(value.value);
  } else if (isArrayExpression(value)) {
    const values = value.elements.filter((e) => isStringLiteral(e)).map((e) => e.value);
    zones.push(...values);
  } else {
    logger.warn(`'zone' property is not a string or array.`, { file });
    return;
  }
}

// src/virtual-modules/generate-virtual-form-module.ts
async function generateVirtualFormModule(sources) {
  const menuItems = await generateMenuItems(sources);
  const widgets = await generateWidgets(sources);
  const customFields = await generateCustomFieldForms(sources);
  const imports = [
    ...menuItems.imports,
    ...widgets.imports,
    ...customFields.imports
  ];
  const code = outdent6`
          ${imports.join("\n")}

          export default {
            ${menuItems.code},
            ${widgets.code},
            ${customFields.code},
          }
        `;
  return generateModule(code);
}

// src/virtual-modules/generate-virtual-link-module.ts
import { outdent as outdent7 } from "outdent";
async function generateVirtualLinkModule(sources) {
  const links = await generateCustomFieldLinks(sources);
  const code = outdent7`
    ${links.imports.join("\n")}

    export default {
      ${links.code}
    }
  `;
  return generateModule(code);
}

// src/virtual-modules/generate-virtual-menu-item-module.ts
import outdent8 from "outdent";
async function generateVirtualMenuItemModule(sources) {
  const menuItems = await generateMenuItems(sources);
  const code = outdent8`
          ${menuItems.imports.join("\n")}

          export default {
            ${menuItems.code},
          }
        `;
  return generateModule(code);
}

// src/virtual-modules/generate-virtual-route-module.ts
import { outdent as outdent9 } from "outdent";
async function generateVirtualRouteModule(sources) {
  const routes = await generateRoutes(sources);
  const imports = [...routes.imports];
  const code = outdent9`
    ${imports.join("\n")}

    export default {
      ${routes.code}
    }
  `;
  return generateModule(code);
}

// src/virtual-modules/generate-virtual-widget-module.ts
import outdent10 from "outdent";
async function generateVirtualWidgetModule(sources) {
  const widgets = await generateWidgets(sources);
  const imports = [...widgets.imports];
  const code = outdent10`
    ${imports.join("\n")}

    export default {
        ${widgets.code},
    }
  `;
  return generateModule(code);
}

// src/vmod.ts
import {
  DISPLAY_VIRTUAL_MODULE,
  FORM_VIRTUAL_MODULE,
  LINK_VIRTUAL_MODULE,
  MENU_ITEM_VIRTUAL_MODULE,
  ROUTE_VIRTUAL_MODULE,
  WIDGET_VIRTUAL_MODULE
} from "@medusajs/admin-shared";
var RESOLVED_LINK_VIRTUAL_MODULE = `\0${LINK_VIRTUAL_MODULE}`;
var RESOLVED_FORM_VIRTUAL_MODULE = `\0${FORM_VIRTUAL_MODULE}`;
var RESOLVED_DISPLAY_VIRTUAL_MODULE = `\0${DISPLAY_VIRTUAL_MODULE}`;
var RESOLVED_ROUTE_VIRTUAL_MODULE = `\0${ROUTE_VIRTUAL_MODULE}`;
var RESOLVED_MENU_ITEM_VIRTUAL_MODULE = `\0${MENU_ITEM_VIRTUAL_MODULE}`;
var RESOLVED_WIDGET_VIRTUAL_MODULE = `\0${WIDGET_VIRTUAL_MODULE}`;
var VIRTUAL_MODULES = [
  LINK_VIRTUAL_MODULE,
  FORM_VIRTUAL_MODULE,
  DISPLAY_VIRTUAL_MODULE,
  ROUTE_VIRTUAL_MODULE,
  MENU_ITEM_VIRTUAL_MODULE,
  WIDGET_VIRTUAL_MODULE
];
var RESOLVED_VIRTUAL_MODULES = [
  RESOLVED_LINK_VIRTUAL_MODULE,
  RESOLVED_FORM_VIRTUAL_MODULE,
  RESOLVED_DISPLAY_VIRTUAL_MODULE,
  RESOLVED_ROUTE_VIRTUAL_MODULE,
  RESOLVED_MENU_ITEM_VIRTUAL_MODULE,
  RESOLVED_WIDGET_VIRTUAL_MODULE
];
function resolveVirtualId(id) {
  return `\0${id}`;
}
function isVirtualModuleId(id) {
  return VIRTUAL_MODULES.includes(id);
}
function isResolvedVirtualModuleId(id) {
  return RESOLVED_VIRTUAL_MODULES.includes(
    id
  );
}
var resolvedVirtualModuleIds = {
  link: RESOLVED_LINK_VIRTUAL_MODULE,
  form: RESOLVED_FORM_VIRTUAL_MODULE,
  display: RESOLVED_DISPLAY_VIRTUAL_MODULE,
  route: RESOLVED_ROUTE_VIRTUAL_MODULE,
  menuItem: RESOLVED_MENU_ITEM_VIRTUAL_MODULE,
  widget: RESOLVED_WIDGET_VIRTUAL_MODULE
};
var virtualModuleIds = {
  link: LINK_VIRTUAL_MODULE,
  form: FORM_VIRTUAL_MODULE,
  display: DISPLAY_VIRTUAL_MODULE,
  route: ROUTE_VIRTUAL_MODULE,
  menuItem: MENU_ITEM_VIRTUAL_MODULE,
  widget: WIDGET_VIRTUAL_MODULE
};
var vmod = {
  resolved: resolvedVirtualModuleIds,
  virtual: virtualModuleIds
};

// src/plugin.ts
var medusaVitePlugin = (options) => {
  const hashMap = /* @__PURE__ */ new Map();
  const _sources = new Set(options?.sources ?? []);
  let watcher;
  function isFileInSources(file) {
    for (const source of _sources) {
      if (file.startsWith(path2.resolve(source))) {
        return true;
      }
    }
    return false;
  }
  async function loadVirtualModule(config) {
    const hash = await config.hashGenerator(_sources);
    hashMap.set(config.hashKey, hash);
    return config.moduleGenerator(_sources);
  }
  async function handleFileChange(server, config) {
    const hashes = await config.hashGenerator(_sources);
    for (const module of config.modules) {
      const newHash = hashes[module.hashKey];
      if (newHash !== hashMap.get(module.virtualModule)) {
        const moduleToReload = server.moduleGraph.getModuleById(
          module.resolvedModule
        );
        if (moduleToReload) {
          await server.reloadModule(moduleToReload);
        }
        hashMap.set(module.virtualModule, newHash);
      }
    }
  }
  return {
    name: "@medusajs/admin-vite-plugin",
    enforce: "pre",
    configureServer(server) {
      watcher = server.watcher;
      watcher?.add(Array.from(_sources));
      watcher?.on("all", async (_event, file) => {
        if (!isFileInSources(file)) {
          return;
        }
        for (const config of watcherConfigs) {
          if (isFileInAdminSubdirectory(file, config.subdirectory)) {
            await handleFileChange(server, config);
          }
        }
      });
    },
    resolveId(id) {
      if (!isVirtualModuleId(id)) {
        return null;
      }
      return resolveVirtualId(id);
    },
    async load(id) {
      if (!isResolvedVirtualModuleId(id)) {
        return null;
      }
      const config = loadConfigs[id];
      if (!config) {
        return null;
      }
      return loadVirtualModule(config);
    },
    async closeBundle() {
      if (watcher) {
        await watcher.close();
      }
    }
  };
};
var loadConfigs = {
  [vmod.resolved.widget]: {
    hashGenerator: async (sources) => generateWidgetHash(sources),
    moduleGenerator: async (sources) => generateVirtualWidgetModule(sources),
    hashKey: vmod.virtual.widget
  },
  [vmod.resolved.link]: {
    hashGenerator: async (sources) => (await generateCustomFieldHashes(sources)).linkHash,
    moduleGenerator: async (sources) => generateVirtualLinkModule(sources),
    hashKey: vmod.virtual.link
  },
  [vmod.resolved.form]: {
    hashGenerator: async (sources) => (await generateCustomFieldHashes(sources)).formHash,
    moduleGenerator: async (sources) => generateVirtualFormModule(sources),
    hashKey: vmod.virtual.form
  },
  [vmod.resolved.display]: {
    hashGenerator: async (sources) => (await generateCustomFieldHashes(sources)).displayHash,
    moduleGenerator: async (sources) => generateVirtualDisplayModule(sources),
    hashKey: vmod.virtual.display
  },
  [vmod.resolved.route]: {
    hashGenerator: async (sources) => (await generateRouteHashes(sources)).defaultExportHash,
    moduleGenerator: async (sources) => generateVirtualRouteModule(sources),
    hashKey: vmod.virtual.route
  },
  [vmod.resolved.menuItem]: {
    hashGenerator: async (sources) => (await generateRouteHashes(sources)).configHash,
    moduleGenerator: async (sources) => generateVirtualMenuItemModule(sources),
    hashKey: vmod.virtual.menuItem
  }
};
var watcherConfigs = [
  {
    subdirectory: "routes",
    hashGenerator: async (sources) => generateRouteHashes(sources),
    modules: [
      {
        virtualModule: vmod.virtual.route,
        resolvedModule: vmod.resolved.route,
        hashKey: "defaultExportHash"
      },
      {
        virtualModule: vmod.virtual.menuItem,
        resolvedModule: vmod.resolved.menuItem,
        hashKey: "configHash"
      }
    ]
  },
  {
    subdirectory: "widgets",
    hashGenerator: async (sources) => ({
      widgetConfigHash: await generateWidgetHash(sources)
    }),
    modules: [
      {
        virtualModule: vmod.virtual.widget,
        resolvedModule: vmod.resolved.widget,
        hashKey: "widgetConfigHash"
      }
    ]
  },
  {
    subdirectory: "custom-fields",
    hashGenerator: async (sources) => generateCustomFieldHashes(sources),
    modules: [
      {
        virtualModule: vmod.virtual.link,
        resolvedModule: vmod.resolved.link,
        hashKey: "linkHash"
      },
      {
        virtualModule: vmod.virtual.form,
        resolvedModule: vmod.resolved.form,
        hashKey: "formHash"
      },
      {
        virtualModule: vmod.virtual.display,
        resolvedModule: vmod.resolved.display,
        hashKey: "displayHash"
      }
    ]
  }
];

// src/index.ts
var src_default = medusaVitePlugin;
export {
  src_default as default
};
