// @ariakit/react-core@0.4.15/cjs/__chunks/CWQU4R75.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _FVVTEJ4Kcjs = require('./FVVTEJ4K.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/composite/composite-context.tsx
var _react = require('react');
var ctx = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_FVVTEJ4Kcjs.CollectionContextProvider],
  [_FVVTEJ4Kcjs.CollectionScopedContextProvider]
);
var useCompositeContext = ctx.useContext;
var useCompositeScopedContext = ctx.useScopedContext;
var useCompositeProviderContext = ctx.useProviderContext;
var CompositeContextProvider = ctx.ContextProvider;
var CompositeScopedContextProvider = ctx.ScopedContextProvider;
var CompositeItemContext = _react.createContext.call(void 0, 
  void 0
);
var CompositeRowContext = _react.createContext.call(void 0, 
  void 0
);









exports.useCompositeContext = useCompositeContext; exports.useCompositeScopedContext = useCompositeScopedContext; exports.useCompositeProviderContext = useCompositeProviderContext; exports.CompositeContextProvider = CompositeContextProvider; exports.CompositeScopedContextProvider = CompositeScopedContextProvider; exports.CompositeItemContext = CompositeItemContext; exports.CompositeRowContext = CompositeRowContext;
