// @ariakit/react-core@0.4.15/cjs/__chunks/EPZ3LJJN.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _4KO75GA3cjs = require('./4KO75GA3.cjs');


var _VF7H2WBXcjs = require('./VF7H2WBX.cjs');




var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');


var _L5A2ID7Scjs = require('./L5A2ID7S.cjs');




var _7EQBAZ46cjs = require('./7EQBAZ46.cjs');

// src/dialog/dialog-dismiss.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useDialogDismiss = _6TI3NXX6cjs.createHook.call(void 0, 
  function useDialogDismiss2(_a) {
    var _b = _a, { store } = _b, props = _7EQBAZ46cjs.__objRest.call(void 0, _b, ["store"]);
    const context = _VF7H2WBXcjs.useDialogScopedContext.call(void 0, );
    store = store || context;
    const onClickProp = props.onClick;
    const onClick = _L5A2ID7Scjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      store == null ? void 0 : store.hide();
    });
    const children = _react.useMemo.call(void 0, 
      () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, 
        "svg",
        {
          "aria-label": "Dismiss popup",
          display: "block",
          fill: "none",
          stroke: "currentColor",
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeWidth: 1.5,
          viewBox: "0 0 16 16",
          height: "1em",
          width: "1em",
          children: [
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "4", x2: "12", y2: "12" }),
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "12", x2: "12", y2: "4" })
          ]
        }
      ),
      []
    );
    props = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {
      "data-dialog-dismiss": "",
      children
    }, props), {
      onClick
    });
    props = _4KO75GA3cjs.useButton.call(void 0, props);
    return props;
  }
);
var DialogDismiss = _6TI3NXX6cjs.forwardRef.call(void 0, function DialogDismiss2(props) {
  const htmlProps = useDialogDismiss(props);
  return _6TI3NXX6cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss;
