// @ariakit/react-core@0.4.15/cjs/__chunks/HLFPESUH.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _CWQU4R75cjs = require('./CWQU4R75.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/menubar/menubar-context.tsx
var _react = require('react');
var menubar = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_CWQU4R75cjs.CompositeContextProvider],
  [_CWQU4R75cjs.CompositeScopedContextProvider]
);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarProviderContext = useMenubarProviderContext; exports.MenubarContextProvider = MenubarContextProvider; exports.MenubarScopedContextProvider = MenubarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
