// @ariakit/react-core@0.4.15/cjs/__chunks/HS75GVU3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _JLMLKTUBcjs = require('./JLMLKTUB.cjs');




var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');



var _L5A2ID7Scjs = require('./L5A2ID7S.cjs');



var _7EQBAZ46cjs = require('./7EQBAZ46.cjs');

// src/heading/heading.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "h1";
var useHeading = _6TI3NXX6cjs.createHook.call(void 0, 
  function useHeading2(props) {
    const ref = _react.useRef.call(void 0, null);
    const level = _react.useContext.call(void 0, _JLMLKTUBcjs.HeadingContext) || 1;
    const Element = `h${level}`;
    const tagName = _L5A2ID7Scjs.useTagName.call(void 0, ref, Element);
    const isNativeHeading = _react.useMemo.call(void 0, 
      () => !!tagName && /^h\d$/.test(tagName),
      [tagName]
    );
    props = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {
      render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Element, {}),
      role: !isNativeHeading ? "heading" : void 0,
      "aria-level": !isNativeHeading ? level : void 0
    }, props), {
      ref: _L5A2ID7Scjs.useMergeRefs.call(void 0, ref, props.ref)
    });
    return props;
  }
);
var Heading = _6TI3NXX6cjs.forwardRef.call(void 0, function Heading2(props) {
  const htmlProps = useHeading(props);
  return _6TI3NXX6cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useHeading = useHeading; exports.Heading = Heading;
