// @ariakit/react-core@0.4.15/cjs/__chunks/JLUMKCNR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});





var _HLFPESUHcjs = require('./HLFPESUH.cjs');



var _CWQU4R75cjs = require('./CWQU4R75.cjs');



var _KPH4LMZOcjs = require('./KPH4LMZO.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/menu/menu-context.tsx
var _react = require('react');
var menu = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_CWQU4R75cjs.CompositeContextProvider, _KPH4LMZOcjs.HovercardContextProvider],
  [_CWQU4R75cjs.CompositeScopedContextProvider, _KPH4LMZOcjs.HovercardScopedContextProvider]
);
var useMenuContext = menu.useContext;
var useMenuScopedContext = menu.useScopedContext;
var useMenuProviderContext = menu.useProviderContext;
var MenuContextProvider = menu.ContextProvider;
var MenuScopedContextProvider = menu.ScopedContextProvider;
var useMenuBarContext = _HLFPESUHcjs.useMenubarContext;
var useMenuBarScopedContext = _HLFPESUHcjs.useMenubarScopedContext;
var useMenuBarProviderContext = _HLFPESUHcjs.useMenubarProviderContext;
var MenuBarContextProvider = _HLFPESUHcjs.MenubarContextProvider;
var MenuBarScopedContextProvider = _HLFPESUHcjs.MenubarScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);













exports.useMenuContext = useMenuContext; exports.useMenuScopedContext = useMenuScopedContext; exports.useMenuProviderContext = useMenuProviderContext; exports.MenuContextProvider = MenuContextProvider; exports.MenuScopedContextProvider = MenuScopedContextProvider; exports.useMenuBarContext = useMenuBarContext; exports.useMenuBarScopedContext = useMenuBarScopedContext; exports.useMenuBarProviderContext = useMenuBarProviderContext; exports.MenuBarContextProvider = MenuBarContextProvider; exports.MenuBarScopedContextProvider = MenuBarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
