// @ariakit/react-core@0.4.15/cjs/__chunks/KJ3PVFDW.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _CWQU4R75cjs = require('./CWQU4R75.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/tag/tag-context.tsx
var _react = require('react');
var TagValueContext = _react.createContext.call(void 0, null);
var TagRemoveIdContext = _react.createContext.call(void 0, 
  null
);
var ctx = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_CWQU4R75cjs.CompositeContextProvider],
  [_CWQU4R75cjs.CompositeScopedContextProvider]
);
var useTagContext = ctx.useContext;
var useTagScopedContext = ctx.useScopedContext;
var useTagProviderContext = ctx.useProviderContext;
var TagContextProvider = ctx.ContextProvider;
var TagScopedContextProvider = ctx.ScopedContextProvider;









exports.TagValueContext = TagValueContext; exports.TagRemoveIdContext = TagRemoveIdContext; exports.useTagContext = useTagContext; exports.useTagScopedContext = useTagScopedContext; exports.useTagProviderContext = useTagProviderContext; exports.TagContextProvider = TagContextProvider; exports.TagScopedContextProvider = TagScopedContextProvider;
