// @ariakit/react-core@0.4.15/cjs/__chunks/M6LCPEQR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _Y35J4OLVcjs = require('./Y35J4OLV.cjs');




var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');



var _7EQBAZ46cjs = require('./7EQBAZ46.cjs');

// src/focus-trap/focus-trap.tsx
var TagName = "span";
var useFocusTrap = _6TI3NXX6cjs.createHook.call(void 0, 
  function useFocusTrap2(props) {
    props = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {
      "data-focus-trap": "",
      tabIndex: 0,
      "aria-hidden": true
    }, props), {
      style: _7EQBAZ46cjs.__spreadValues.call(void 0, {
        // Prevents unintended scroll jumps.
        position: "fixed",
        top: 0,
        left: 0
      }, props.style)
    });
    props = _Y35J4OLVcjs.useVisuallyHidden.call(void 0, props);
    return props;
  }
);
var FocusTrap = _6TI3NXX6cjs.forwardRef.call(void 0, function FocusTrap2(props) {
  const htmlProps = useFocusTrap(props);
  return _6TI3NXX6cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useFocusTrap = useFocusTrap; exports.FocusTrap = FocusTrap;
