// @ariakit/react-core@0.4.15/cjs/__chunks/MUURFTXX.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});




var _SQHULPZ2cjs = require('./SQHULPZ2.cjs');


var _53NM5U3Acjs = require('./53NM5U3A.cjs');




var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');






var _L5A2ID7Scjs = require('./L5A2ID7S.cjs');




var _7EQBAZ46cjs = require('./7EQBAZ46.cjs');

// src/combobox/combobox-list.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useComboboxList = _6TI3NXX6cjs.createHook.call(void 0, 
  function useComboboxList2(_a) {
    var _b = _a, { store, alwaysVisible } = _b, props = _7EQBAZ46cjs.__objRest.call(void 0, _b, ["store", "alwaysVisible"]);
    const scopedContext = _SQHULPZ2cjs.useComboboxScopedContext.call(void 0, true);
    const context = _SQHULPZ2cjs.useComboboxContext.call(void 0, );
    store = store || context;
    const scopedContextSameStore = !!store && store === scopedContext;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "ComboboxList must receive a `store` prop or be wrapped in a ComboboxProvider component."
    );
    const ref = _react.useRef.call(void 0, null);
    const id = _L5A2ID7Scjs.useId.call(void 0, props.id);
    const mounted = store.useState("mounted");
    const hidden = _53NM5U3Acjs.isHidden.call(void 0, mounted, props.hidden, alwaysVisible);
    const style = hidden ? _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {}, props.style), { display: "none" }) : props.style;
    const multiSelectable = store.useState(
      (state) => Array.isArray(state.selectedValue)
    );
    const role = _L5A2ID7Scjs.useAttribute.call(void 0, ref, "role", props.role);
    const isCompositeRole = role === "listbox" || role === "tree" || role === "grid";
    const ariaMultiSelectable = isCompositeRole ? multiSelectable || void 0 : void 0;
    const [hasListboxInside, setHasListboxInside] = _react.useState.call(void 0, false);
    const contentElement = store.useState("contentElement");
    _L5A2ID7Scjs.useSafeLayoutEffect.call(void 0, () => {
      if (!mounted) return;
      const element = ref.current;
      if (!element) return;
      if (contentElement !== element) return;
      const callback = () => {
        setHasListboxInside(!!element.querySelector("[role='listbox']"));
      };
      const observer = new MutationObserver(callback);
      observer.observe(element, {
        subtree: true,
        childList: true,
        attributeFilter: ["role"]
      });
      callback();
      return () => observer.disconnect();
    }, [mounted, contentElement]);
    if (!hasListboxInside) {
      props = _7EQBAZ46cjs.__spreadValues.call(void 0, {
        role: "listbox",
        "aria-multiselectable": ariaMultiSelectable
      }, props);
    }
    props = _L5A2ID7Scjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _SQHULPZ2cjs.ComboboxScopedContextProvider, { value: store, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _SQHULPZ2cjs.ComboboxListRoleContext.Provider, { value: role, children: element }) }),
      [store, role]
    );
    const setContentElement = id && (!scopedContext || !scopedContextSameStore) ? store.setContentElement : null;
    props = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {
      id,
      hidden
    }, props), {
      ref: _L5A2ID7Scjs.useMergeRefs.call(void 0, setContentElement, ref, props.ref),
      style
    });
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var ComboboxList = _6TI3NXX6cjs.forwardRef.call(void 0, function ComboboxList2(props) {
  const htmlProps = useComboboxList(props);
  return _6TI3NXX6cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useComboboxList = useComboboxList; exports.ComboboxList = ComboboxList;
