// @ariakit/react-core@0.4.15/cjs/__chunks/QZEXWZZZ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _CWQU4R75cjs = require('./CWQU4R75.cjs');



var _ZVS6SBRRcjs = require('./ZVS6SBRR.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/select/select-context.tsx
var _react = require('react');
var ctx = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_ZVS6SBRRcjs.PopoverContextProvider, _CWQU4R75cjs.CompositeContextProvider],
  [_ZVS6SBRRcjs.PopoverScopedContextProvider, _CWQU4R75cjs.CompositeScopedContextProvider]
);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
