// @ariakit/react-core@0.4.15/cjs/__chunks/SQHULPZ2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _CWQU4R75cjs = require('./CWQU4R75.cjs');



var _ZVS6SBRRcjs = require('./ZVS6SBRR.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/combobox/combobox-context.tsx
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, 
  void 0
);
var ctx = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_ZVS6SBRRcjs.PopoverContextProvider, _CWQU4R75cjs.CompositeContextProvider],
  [_ZVS6SBRRcjs.PopoverScopedContextProvider, _CWQU4R75cjs.CompositeScopedContextProvider]
);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, 
  void 0
);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
