// @ariakit/react-core@0.4.15/cjs/__chunks/VF7H2WBX.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _EJVJIL4Scjs = require('./EJVJIL4S.cjs');


var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');

// src/dialog/dialog-context.tsx
var _react = require('react');
var ctx = _6TI3NXX6cjs.createStoreContext.call(void 0, 
  [_EJVJIL4Scjs.DisclosureContextProvider],
  [_EJVJIL4Scjs.DisclosureScopedContextProvider]
);
var useDialogContext = ctx.useContext;
var useDialogScopedContext = ctx.useScopedContext;
var useDialogProviderContext = ctx.useProviderContext;
var DialogContextProvider = ctx.ContextProvider;
var DialogScopedContextProvider = ctx.ScopedContextProvider;
var DialogHeadingContext = _react.createContext.call(void 0, void 0);
var DialogDescriptionContext = _react.createContext.call(void 0, void 0);









exports.useDialogContext = useDialogContext; exports.useDialogScopedContext = useDialogScopedContext; exports.useDialogProviderContext = useDialogProviderContext; exports.DialogContextProvider = DialogContextProvider; exports.DialogScopedContextProvider = DialogScopedContextProvider; exports.DialogHeadingContext = DialogHeadingContext; exports.DialogDescriptionContext = DialogDescriptionContext;
