// @ariakit/react-core@0.4.15/cjs/__chunks/Y35J4OLV.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



var _6TI3NXX6cjs = require('./6TI3NXX6.cjs');



var _7EQBAZ46cjs = require('./7EQBAZ46.cjs');

// src/visually-hidden/visually-hidden.tsx
var TagName = "span";
var useVisuallyHidden = _6TI3NXX6cjs.createHook.call(void 0, 
  function useVisuallyHidden2(props) {
    props = _7EQBAZ46cjs.__spreadProps.call(void 0, _7EQBAZ46cjs.__spreadValues.call(void 0, {}, props), {
      style: _7EQBAZ46cjs.__spreadValues.call(void 0, {
        border: 0,
        clip: "rect(0 0 0 0)",
        height: "1px",
        margin: "-1px",
        overflow: "hidden",
        padding: 0,
        position: "absolute",
        whiteSpace: "nowrap",
        width: "1px"
      }, props.style)
    });
    return props;
  }
);
var VisuallyHidden = _6TI3NXX6cjs.forwardRef.call(void 0, function VisuallyHidden2(props) {
  const htmlProps = useVisuallyHidden(props);
  return _6TI3NXX6cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useVisuallyHidden = useVisuallyHidden; exports.VisuallyHidden = VisuallyHidden;
